// @platformatic/kafka/dist/clients/admin/admin.js
import { createPromisifiedCallback, kCallbackPromise, runConcurrentCallbacks } from "../../apis/callbacks.js";
import { ConfigResourceTypes, FetchIsolationLevels, FindCoordinatorKeyTypes } from "../../apis/enumerations.js";
import { adminAclsChannel, adminClientQuotasChannel, adminConfigsChannel, adminConsumerGroupOffsetsChannel, adminGroupsChannel, adminLogDirsChannel, adminOffsetsChannel, adminTopicsChannel, createDiagnosticContext } from "../../diagnostic.js";
import { MultipleErrors } from "../../errors.js";
import { Reader } from "../../protocol/reader.js";
import { Base, kAfterCreate, kCheckNotClosed, kConnections, kGetApi, kGetBootstrapConnection, kGetConnection, kMetadata, kOptions, kPerformDeduplicated, kPerformWithRetry, kValidateOptions } from "../base/base.js";
import { adminListOffsetsOptionsValidator, alterClientQuotasOptionsValidator, alterConfigsOptionsValidator, alterConsumerGroupOffsetsOptionsValidator, createAclsOptionsValidator, createPartitionsOptionsValidator, createTopicsOptionsValidator, deleteAclsOptionsValidator, deleteConsumerGroupOffsetsOptionsValidator, deleteRecordsOptionsValidator, deleteGroupsOptionsValidator, deleteTopicsOptionsValidator, describeAclsOptionsValidator, describeClientQuotasOptionsValidator, describeConfigsOptionsValidator, describeGroupsOptionsValidator, describeLogDirsOptionsValidator, incrementalAlterConfigsOptionsValidator, listConsumerGroupOffsetsOptionsValidator, listGroupsOptionsValidator, listTopicsOptionsValidator, removeMembersFromConsumerGroupOptionsValidator } from "./options.js";
export class Admin extends Base {
    #controller = null;
    constructor(options) {
        super(options);
        this[kAfterCreate]('admin');
    }
    listTopics(options, callback) {
        if (!callback) {
            callback = createPromisifiedCallback();
        }
        if (this[kCheckNotClosed](callback)) {
            return callback[kCallbackPromise];
        }
        if (!options) {
            options = {};
        }
        const validationError = this[kValidateOptions](options, listTopicsOptionsValidator, '/options', false);
        if (validationError) {
            callback(validationError);
            return callback[kCallbackPromise];
        }
        adminTopicsChannel.traceCallback(this.#listTopics, 1, createDiagnosticContext({ client: this, operation: 'listTopics', options }), this, options, callback);
        return callback[kCallbackPromise];
    }
    createTopics(options, callback) {
        if (!callback) {
            callback = createPromisifiedCallback();
        }
        if (this[kCheckNotClosed](callback)) {
            return callback[kCallbackPromise];
        }
        const validationError = this[kValidateOptions](options, createTopicsOptionsValidator, '/options', false);
        if (validationError) {
            callback(validationError);
            return callback[kCallbackPromise];
        }
        adminTopicsChannel.traceCallback(this.#createTopics, 1, createDiagnosticContext({ client: this, operation: 'createTopics', options }), this, options, callback);
        return callback[kCallbackPromise];
    }
    deleteTopics(options, callback) {
        if (!callback) {
            callback = createPromisifiedCallback();
        }
        if (this[kCheckNotClosed](callback)) {
            return callback[kCallbackPromise];
        }
        const validationError = this[kValidateOptions](options, deleteTopicsOptionsValidator, '/options', false);
        if (validationError) {
            callback(validationError);
            return callback[kCallbackPromise];
        }
        adminTopicsChannel.traceCallback(this.#deleteTopics, 1, createDiagnosticContext({ client: this, operation: 'deleteTopics', options }), this, options, callback);
        return callback[kCallbackPromise];
    }
    createPartitions(options, callback) {
        if (!callback) {
            callback = createPromisifiedCallback();
        }
        if (this[kCheckNotClosed](callback)) {
            return callback[kCallbackPromise];
        }
        const validationError = this[kValidateOptions](options, createPartitionsOptionsValidator, '/options', false);
        if (validationError) {
            callback(validationError, undefined);
            return callback[kCallbackPromise];
        }
        adminTopicsChannel.traceCallback(this.#createPartitions, 1, createDiagnosticContext({ client: this, operation: 'createPartitions', options }), this, options, callback);
        return callback[kCallbackPromise];
    }
    listGroups(options, callback) {
        if (!callback) {
            callback = createPromisifiedCallback();
        }
        if (this[kCheckNotClosed](callback)) {
            return callback[kCallbackPromise];
        }
        if (!options) {
            options = {};
        }
        const validationError = this[kValidateOptions](options, listGroupsOptionsValidator, '/options', false);
        if (validationError) {
            callback(validationError);
            return callback[kCallbackPromise];
        }
        options.types ??= ['classic'];
        adminGroupsChannel.traceCallback(this.#listGroups, 1, createDiagnosticContext({ client: this, operation: 'listGroups', options }), this, options, callback);
        return callback[kCallbackPromise];
    }
    describeGroups(options, callback) {
        if (!callback) {
            callback = createPromisifiedCallback();
        }
        if (this[kCheckNotClosed](callback)) {
            return callback[kCallbackPromise];
        }
        const validationError = this[kValidateOptions](options, describeGroupsOptionsValidator, '/options', false);
        if (validationError) {
            callback(validationError);
            return callback[kCallbackPromise];
        }
        adminGroupsChannel.traceCallback(this.#describeGroups, 1, createDiagnosticContext({ client: this, operation: 'describeGroups', options }), this, options, callback);
        return callback[kCallbackPromise];
    }
    deleteGroups(options, callback) {
        if (!callback) {
            callback = createPromisifiedCallback();
        }
        if (this[kCheckNotClosed](callback)) {
            return callback[kCallbackPromise];
        }
        const validationError = this[kValidateOptions](options, deleteGroupsOptionsValidator, '/options', false);
        if (validationError) {
            callback(validationError);
            return callback[kCallbackPromise];
        }
        adminGroupsChannel.traceCallback(this.#deleteGroups, 1, createDiagnosticContext({ client: this, operation: 'deleteGroups', options }), this, options, callback);
        return callback[kCallbackPromise];
    }
    removeMembersFromConsumerGroup(options, callback) {
        if (!callback) {
            callback = createPromisifiedCallback();
        }
        if (this[kCheckNotClosed](callback)) {
            return callback[kCallbackPromise];
        }
        const validationError = this[kValidateOptions](options, removeMembersFromConsumerGroupOptionsValidator, '/options', false);
        if (validationError) {
            callback(validationError);
            return callback[kCallbackPromise];
        }
        adminGroupsChannel.traceCallback(this.#removeMembersFromConsumerGroup, 1, createDiagnosticContext({ client: this, operation: 'removeMembersFromConsumerGroup', options }), this, options, callback);
        return callback[kCallbackPromise];
    }
    describeClientQuotas(options, callback) {
        if (!callback) {
            callback = createPromisifiedCallback();
        }
        if (this[kCheckNotClosed](callback)) {
            return callback[kCallbackPromise];
        }
        const validationError = this[kValidateOptions](options, describeClientQuotasOptionsValidator, '/options', false);
        if (validationError) {
            callback(validationError);
            return callback[kCallbackPromise];
        }
        adminClientQuotasChannel.traceCallback(this.#describeClientQuotas, 1, createDiagnosticContext({ client: this, operation: 'describeClientQuotas', options }), this, options, callback);
        return callback[kCallbackPromise];
    }
    alterClientQuotas(options, callback) {
        if (!callback) {
            callback = createPromisifiedCallback();
        }
        if (this[kCheckNotClosed](callback)) {
            return callback[kCallbackPromise];
        }
        const validationError = this[kValidateOptions](options, alterClientQuotasOptionsValidator, '/options', false);
        if (validationError) {
            callback(validationError);
            return callback[kCallbackPromise];
        }
        adminClientQuotasChannel.traceCallback(this.#alterClientQuotas, 1, createDiagnosticContext({ client: this, operation: 'alterClientQuotas', options }), this, options, callback);
        return callback[kCallbackPromise];
    }
    describeLogDirs(options, callback) {
        if (!callback) {
            callback = createPromisifiedCallback();
        }
        /* c8 ignore next 3 - Hard to test */
        if (this[kCheckNotClosed](callback)) {
            return callback[kCallbackPromise];
        }
        const validationError = this[kValidateOptions](options, describeLogDirsOptionsValidator, '/options', false);
        if (validationError) {
            callback(validationError);
            return callback[kCallbackPromise];
        }
        adminLogDirsChannel.traceCallback(this.#describeLogDirs, 1, createDiagnosticContext({ client: this, operation: 'describeLogDirs', options }), this, options, callback);
        return callback[kCallbackPromise];
    }
    listConsumerGroupOffsets(options, callback) {
        if (!callback) {
            callback = createPromisifiedCallback();
        }
        if (this[kCheckNotClosed](callback)) {
            return callback[kCallbackPromise];
        }
        const validationError = this[kValidateOptions](options, listConsumerGroupOffsetsOptionsValidator, '/options', false);
        if (validationError) {
            callback(validationError);
            return callback[kCallbackPromise];
        }
        adminConsumerGroupOffsetsChannel.traceCallback(this.#listConsumerGroupOffsets, 1, createDiagnosticContext({ client: this, operation: 'listConsumerGroupOffsets', options }), this, options, callback);
        return callback[kCallbackPromise];
    }
    alterConsumerGroupOffsets(options, callback) {
        if (!callback) {
            callback = createPromisifiedCallback();
        }
        if (this[kCheckNotClosed](callback)) {
            return callback[kCallbackPromise];
        }
        const validationError = this[kValidateOptions](options, alterConsumerGroupOffsetsOptionsValidator, '/options', false);
        if (validationError) {
            callback(validationError);
            return callback[kCallbackPromise];
        }
        adminConsumerGroupOffsetsChannel.traceCallback(this.#alterConsumerGroupOffsets, 1, createDiagnosticContext({ client: this, operation: 'alterConsumerGroupOffsets', options }), this, options, callback);
        return callback[kCallbackPromise];
    }
    deleteConsumerGroupOffsets(options, callback) {
        if (!callback) {
            callback = createPromisifiedCallback();
        }
        if (this[kCheckNotClosed](callback)) {
            return callback[kCallbackPromise];
        }
        const validationError = this[kValidateOptions](options, deleteConsumerGroupOffsetsOptionsValidator, '/options', false);
        if (validationError) {
            callback(validationError);
            return callback[kCallbackPromise];
        }
        adminConsumerGroupOffsetsChannel.traceCallback(this.#deleteConsumerGroupOffsets, 1, createDiagnosticContext({ client: this, operation: 'deleteConsumerGroupOffsets', options }), this, options, callback);
        return callback[kCallbackPromise];
    }
    describeConfigs(options, callback) {
        if (!callback) {
            callback = createPromisifiedCallback();
        }
        if (this[kCheckNotClosed](callback)) {
            return callback[kCallbackPromise];
        }
        const validationError = this[kValidateOptions](options, describeConfigsOptionsValidator, '/options', false);
        if (validationError) {
            callback(validationError);
            return callback[kCallbackPromise];
        }
        adminConfigsChannel.traceCallback(this.#describeConfigs, 1, createDiagnosticContext({ client: this, operation: 'describeConfigs', options }), this, options, callback);
        return callback[kCallbackPromise];
    }
    alterConfigs(options, callback) {
        if (!callback) {
            callback = createPromisifiedCallback();
        }
        if (this[kCheckNotClosed](callback)) {
            return callback[kCallbackPromise];
        }
        const validationError = this[kValidateOptions](options, alterConfigsOptionsValidator, '/options', false);
        if (validationError) {
            callback(validationError, undefined);
            return callback[kCallbackPromise];
        }
        adminConfigsChannel.traceCallback(this.#alterConfigs, 1, createDiagnosticContext({ client: this, operation: 'alterConfigs', options }), this, options, callback);
        return callback[kCallbackPromise];
    }
    incrementalAlterConfigs(options, callback) {
        if (!callback) {
            callback = createPromisifiedCallback();
        }
        if (this[kCheckNotClosed](callback)) {
            return callback[kCallbackPromise];
        }
        const validationError = this[kValidateOptions](options, incrementalAlterConfigsOptionsValidator, '/options', false);
        if (validationError) {
            callback(validationError, undefined);
            return callback[kCallbackPromise];
        }
        adminConfigsChannel.traceCallback(this.#incrementalAlterConfigs, 1, createDiagnosticContext({ client: this, operation: 'incrementalAlterConfigs', options }), this, options, callback);
        return callback[kCallbackPromise];
    }
    createAcls(options, callback) {
        if (!callback) {
            callback = createPromisifiedCallback();
        }
        if (this[kCheckNotClosed](callback)) {
            return callback[kCallbackPromise];
        }
        const validationError = this[kValidateOptions](options, createAclsOptionsValidator, '/options', false);
        if (validationError) {
            callback(validationError, undefined);
            return callback[kCallbackPromise];
        }
        adminAclsChannel.traceCallback(this.#createAcls, 1, createDiagnosticContext({ client: this, operation: 'createAcls', options }), this, options, callback);
        return callback[kCallbackPromise];
    }
    describeAcls(options, callback) {
        if (!callback) {
            callback = createPromisifiedCallback();
        }
        if (this[kCheckNotClosed](callback)) {
            return callback[kCallbackPromise];
        }
        const validationError = this[kValidateOptions](options, describeAclsOptionsValidator, '/options', false);
        if (validationError) {
            callback(validationError);
            return callback[kCallbackPromise];
        }
        adminAclsChannel.traceCallback(this.#describeAcls, 1, createDiagnosticContext({ client: this, operation: 'describeAcls', options }), this, options, callback);
        return callback[kCallbackPromise];
    }
    deleteAcls(options, callback) {
        if (!callback) {
            callback = createPromisifiedCallback();
        }
        if (this[kCheckNotClosed](callback)) {
            return callback[kCallbackPromise];
        }
        const validationError = this[kValidateOptions](options, deleteAclsOptionsValidator, '/options', false);
        if (validationError) {
            callback(validationError);
            return callback[kCallbackPromise];
        }
        adminAclsChannel.traceCallback(this.#deleteAcls, 1, createDiagnosticContext({ client: this, operation: 'deleteAcls', options }), this, options, callback);
        return callback[kCallbackPromise];
    }
    listOffsets(options, callback) {
        if (!callback) {
            callback = createPromisifiedCallback();
        }
        if (this[kCheckNotClosed](callback)) {
            return callback[kCallbackPromise];
        }
        const validationError = this[kValidateOptions](options, adminListOffsetsOptionsValidator, '/options', false);
        if (validationError) {
            callback(validationError);
            return callback[kCallbackPromise];
        }
        adminOffsetsChannel.traceCallback(this.#listOffsets, 1, createDiagnosticContext({ client: this, operation: 'listOffsets', options }), this, options, callback);
        return callback[kCallbackPromise];
    }
    deleteRecords(options, callback) {
        if (!callback) {
            callback = createPromisifiedCallback();
        }
        if (this[kCheckNotClosed](callback)) {
            return callback[kCallbackPromise];
        }
        const validationError = this[kValidateOptions](options, deleteRecordsOptionsValidator, '/options', false);
        if (validationError) {
            callback(validationError);
            return callback[kCallbackPromise];
        }
        adminOffsetsChannel.traceCallback(this.#deleteRecords, 1, createDiagnosticContext({ client: this, operation: 'deleteRecords', options }), this, options, callback);
        return callback[kCallbackPromise];
    }
    #getControllerConnection(callback) {
        if (this.#controller) {
            this[kConnections].get(this.#controller, callback);
        }
        else {
            this[kGetBootstrapConnection](callback);
        }
    }
    #handleNotControllerError(error, value, callback) {
        if (error && error?.findBy?.('apiCode', 41)) {
            this.metadata({ topics: [] }, (metadataError, metadata) => {
                /* c8 ignore next 4 - Hard to test */
                if (metadataError) {
                    callback(metadataError);
                    return;
                }
                this.#controller = metadata.brokers.get(metadata.controllerId);
                callback(error);
            });
        }
        else {
            callback(error, value);
        }
    }
    #listTopics(options, callback) {
        const includeInternals = options.includeInternals ?? false;
        this[kPerformDeduplicated]('metadata', deduplicateCallback => {
            this[kPerformWithRetry]('metadata', retryCallback => {
                this[kGetBootstrapConnection]((error, connection) => {
                    if (error) {
                        retryCallback(error);
                        return;
                    }
                    this[kGetApi]('Metadata', (error, api) => {
                        if (error) {
                            retryCallback(error);
                            return;
                        }
                        api(connection, null, false, false, retryCallback);
                    });
                });
            }, (error, metadata) => {
                if (error) {
                    deduplicateCallback(error);
                    return;
                }
                const topics = new Set();
                for (const { name, isInternal } of metadata.topics) {
                    /* c8 ignore next 3 - Sometimes internal topics might be returned by Kafka */
                    if (isInternal && !includeInternals) {
                        continue;
                    }
                    topics.add(name);
                }
                deduplicateCallback(null, Array.from(topics).sort());
            }, 0);
        }, callback);
    }
    #createTopics(options, callback) {
        // -1 is required if manual assignments are used. If no manual assignments are used, -1 will default to broker settings.
        const numPartitions = options.partitions ?? -1;
        const replicationFactor = options.replicas ?? -1;
        const assignments = [];
        const configs = options.configs ?? [];
        for (const { partition, brokers } of options.assignments ?? []) {
            assignments.push({ partitionIndex: partition, brokerIds: brokers });
        }
        const requests = [];
        for (const topic of options.topics) {
            requests.push({
                name: topic,
                numPartitions,
                replicationFactor,
                assignments,
                configs
            });
        }
        this[kPerformDeduplicated](`createTopics-${options.topics.join(',')}`, deduplicateCallback => {
            this[kPerformWithRetry]('createTopics', retryCallback => {
                this.#getControllerConnection((error, connection) => {
                    if (error) {
                        retryCallback(error);
                        return;
                    }
                    this[kGetApi]('CreateTopics', (error, api) => {
                        if (error) {
                            retryCallback(error);
                            return;
                        }
                        api(connection, requests, this[kOptions].timeout, false, (error, response) => {
                            this.#handleNotControllerError(error, response, retryCallback);
                        });
                    });
                });
            }, (error, response) => {
                if (error) {
                    deduplicateCallback(error);
                    return;
                }
                const created = [];
                for (const { name, topicId: id, numPartitions: partitions, replicationFactor: replicas, configs } of response.topics) {
                    const configuration = {};
                    for (const { name, value } of configs) {
                        configuration[name] = value;
                    }
                    created.push({ id, name, partitions, replicas, configuration });
                }
                deduplicateCallback(null, created);
            }, 0);
        }, callback);
    }
    #deleteTopics(options, callback) {
        this[kPerformDeduplicated](`deleteTopics-${options.topics.join(',')}`, deduplicateCallback => {
            this[kPerformWithRetry]('deleteTopics', retryCallback => {
                this.#getControllerConnection((error, connection) => {
                    if (error) {
                        retryCallback(error, undefined);
                        return;
                    }
                    const requests = [];
                    for (const topic of options.topics) {
                        requests.push({ name: topic });
                    }
                    this[kGetApi]('DeleteTopics', (error, api) => {
                        if (error) {
                            retryCallback(error);
                            return;
                        }
                        api(connection, requests, this[kOptions].timeout, (error, response) => {
                            this.#handleNotControllerError(error, response, retryCallback);
                        });
                    });
                });
            }, deduplicateCallback, 0);
        }, error => callback(error));
    }
    #createPartitions(options, callback) {
        this[kPerformDeduplicated]('createPartitions', deduplicateCallback => {
            this[kPerformWithRetry]('createPartitions', retryCallback => {
                this.#getControllerConnection((error, connection) => {
                    if (error) {
                        retryCallback(error);
                        return;
                    }
                    this[kGetApi]('CreatePartitions', (error, api) => {
                        if (error) {
                            retryCallback(error);
                            return;
                        }
                        api(connection, options.topics, this[kOptions].timeout, options.validateOnly ?? false, (error, response) => {
                            this.#handleNotControllerError(error, response, retryCallback);
                        });
                    });
                });
            }, deduplicateCallback, 0);
        }, error => {
            if (error) {
                callback(new MultipleErrors('Creating partitions failed.', [error]));
            }
            else {
                callback(null);
            }
        });
    }
    #listGroups(options, callback) {
        // Find all the brokers in the cluster
        this[kMetadata]({ topics: [] }, (error, metadata) => {
            if (error) {
                callback(error);
                return;
            }
            runConcurrentCallbacks('Listing groups failed.', metadata.brokers, ([, broker], concurrentCallback) => {
                this[kGetConnection](broker, (error, connection) => {
                    if (error) {
                        concurrentCallback(error);
                        return;
                    }
                    this[kPerformWithRetry]('listGroups', retryCallback => {
                        this[kGetApi]('ListGroups', (error, api) => {
                            if (error) {
                                retryCallback(error);
                                return;
                            }
                            /* c8 ignore next 5 */
                            api(connection, options.states ?? [], options.types, retryCallback);
                        });
                    }, concurrentCallback, 0);
                });
            }, (error, results) => {
                if (error) {
                    callback(error);
                    return;
                }
                const groups = new Map();
                for (const result of results) {
                    for (const raw of result.groups) {
                        groups.set(raw.groupId, {
                            id: raw.groupId,
                            state: raw.groupState.toUpperCase(),
                            groupType: raw.groupType,
                            protocolType: raw.protocolType
                        });
                    }
                }
                callback(null, groups);
            });
        });
    }
    #describeGroups(options, callback) {
        this[kMetadata]({ topics: [] }, (error, metadata) => {
            if (error) {
                callback(error);
                return;
            }
            this.#findGroupCoordinator(options.groups, (error, response) => {
                if (error) {
                    callback(error);
                    return;
                }
                // Group the groups by coordinator
                const coordinators = new Map();
                for (const { key: group, nodeId: node } of response.coordinators) {
                    let coordinator = coordinators.get(node);
                    if (!coordinator) {
                        coordinator = [];
                        coordinators.set(node, coordinator);
                    }
                    coordinator.push(group);
                }
                runConcurrentCallbacks('Describing groups failed.', coordinators, ([node, groups], concurrentCallback) => {
                    this[kGetConnection](metadata.brokers.get(node), (error, connection) => {
                        if (error) {
                            concurrentCallback(error);
                            return;
                        }
                        this[kPerformWithRetry]('describeGroups', retryCallback => {
                            this[kGetApi]('DescribeGroups', (error, api) => {
                                if (error) {
                                    retryCallback(error);
                                    return;
                                }
                                api(connection, groups, options.includeAuthorizedOperations ?? false, retryCallback);
                            });
                        }, concurrentCallback, 0);
                    });
                }, (error, results) => {
                    if (error) {
                        callback(error);
                        return;
                    }
                    const groups = new Map();
                    for (const result of results) {
                        for (const raw of result.groups) {
                            const group = {
                                id: raw.groupId,
                                state: raw.groupState.toUpperCase(),
                                protocolType: raw.protocolType,
                                protocol: raw.protocolData,
                                members: new Map(),
                                authorizedOperations: raw.authorizedOperations
                            };
                            for (const member of raw.members) {
                                const reader = Reader.from(member.memberMetadata);
                                let memberMetadata;
                                let memberAssignments;
                                if (reader.remaining > 0) {
                                    memberMetadata = {
                                        version: reader.readInt16(),
                                        topics: reader.readArray(r => r.readString(false), false, false),
                                        metadata: reader.readBytes(false)
                                    };
                                    reader.reset(member.memberAssignment);
                                    reader.skip(2); // Ignore Version information
                                    memberAssignments = reader.readMap(r => {
                                        const topic = r.readString(false);
                                        return [topic, { topic, partitions: reader.readArray(r => r.readInt32(), false, false) }];
                                    }, false, false);
                                    reader.readBytes(); // Ignore the user data
                                }
                                group.members.set(member.memberId, {
                                    id: member.memberId,
                                    groupInstanceId: member.groupInstanceId,
                                    clientId: member.clientId,
                                    clientHost: member.clientHost,
                                    metadata: memberMetadata,
                                    assignments: memberAssignments
                                });
                            }
                            groups.set(group.id, group);
                        }
                    }
                    callback(null, groups);
                });
            });
        });
    }
    #deleteGroups(options, callback) {
        this[kMetadata]({ topics: [] }, (error, metadata) => {
            if (error) {
                callback(error);
                return;
            }
            this.#findGroupCoordinator(options.groups, (error, response) => {
                if (error) {
                    callback(error);
                    return;
                }
                // Group the groups by coordinator
                const coordinators = new Map();
                for (const { key: group, nodeId: node } of response.coordinators) {
                    let coordinator = coordinators.get(node);
                    if (!coordinator) {
                        coordinator = [];
                        coordinators.set(node, coordinator);
                    }
                    coordinator.push(group);
                }
                runConcurrentCallbacks('Deleting groups failed.', coordinators, ([node, groups], concurrentCallback) => {
                    this[kGetConnection](metadata.brokers.get(node), (error, connection) => {
                        if (error) {
                            concurrentCallback(error);
                            return;
                        }
                        this[kPerformWithRetry]('deleteGroups', retryCallback => {
                            this[kGetApi]('DeleteGroups', (error, api) => {
                                if (error) {
                                    retryCallback(error);
                                    return;
                                }
                                api(connection, groups, retryCallback);
                            });
                        }, concurrentCallback, 0);
                    });
                }, error => callback(error));
            });
        });
    }
    #removeMembersFromConsumerGroup(options, callback) {
        if (!options.members || options.members.length === 0) {
            this.#describeGroups({ groups: [options.groupId] }, (error, groupsMap) => {
                if (error) {
                    callback(new MultipleErrors('Removing members from consumer group failed.', [error]));
                    return;
                }
                const group = groupsMap.get(options.groupId);
                /* c8 ignore next 4 - Hard to test */
                if (!group) {
                    callback(new MultipleErrors('Removing members from consumer group failed.', []));
                    return;
                }
                const allMemberIds = Array.from(group.members.keys());
                /* c8 ignore next 4 - Hard to test */
                if (allMemberIds.length === 0) {
                    callback(null);
                    return;
                }
                this.#removeMembersFromConsumerGroup({ ...options, members: allMemberIds.map(memberId => ({ memberId })) }, callback);
            });
            return;
        }
        this[kMetadata]({ topics: [] }, (error, metadata) => {
            if (error) {
                callback(error, undefined);
                return;
            }
            this.#findGroupCoordinator([options.groupId], (error, response) => {
                if (error) {
                    callback(new MultipleErrors('Removing members from consumer group failed.', [error]));
                    return;
                }
                const coordinator = response.coordinators.find(c => c.key === options.groupId);
                /* c8 ignore next 8 - Hard to test */
                if (!coordinator) {
                    callback(new MultipleErrors('Removing members from consumer group failed.', [
                        new Error(`No coordinator found for group ${options.groupId}`)
                    ]));
                    return;
                }
                const broker = metadata.brokers.get(coordinator.nodeId);
                /* c8 ignore next 8 - Hard to test */
                if (!broker) {
                    callback(new MultipleErrors('Removing members from consumer group failed.', [
                        new Error(`Broker ${coordinator.nodeId} not found`)
                    ]));
                    return;
                }
                this[kGetConnection](broker, (error, connection) => {
                    if (error) {
                        callback(new MultipleErrors('Removing members from consumer group failed.', [error]));
                        return;
                    }
                    this[kPerformWithRetry]('removeMembersFromConsumerGroup', retryCallback => {
                        this[kGetApi]('LeaveGroup', (error, api) => {
                            if (error) {
                                retryCallback(error);
                                return;
                            }
                            /* c8 ignore next 5 - Hard to test */
                            const members = options.members.map(member => ({
                                memberId: typeof member === 'string' ? member : member.memberId,
                                groupInstanceId: null,
                                reason: typeof member === 'string' ? 'Not specified' : member.reason
                            }));
                            api(connection, options.groupId, members, retryCallback);
                        });
                    }, error => {
                        if (error) {
                            callback(new MultipleErrors('Removing members from consumer group failed.', [error]));
                            return;
                        }
                        callback(null);
                    }, 0);
                });
            });
        });
    }
    #findGroupCoordinator(groups, callback) {
        this[kPerformWithRetry]('findGroupCoordinator', retryCallback => {
            this[kGetBootstrapConnection]((error, connection) => {
                if (error) {
                    retryCallback(error);
                    return;
                }
                this[kGetApi]('FindCoordinator', (error, api) => {
                    if (error) {
                        retryCallback(error);
                        return;
                    }
                    api(connection, FindCoordinatorKeyTypes.GROUP, groups, retryCallback);
                });
            });
        }, (error, response) => {
            if (error) {
                callback(error);
                return;
            }
            callback(null, response);
        }, 0);
    }
    #describeClientQuotas(options, callback) {
        this[kPerformWithRetry]('describeClientQuotas', retryCallback => {
            this[kGetBootstrapConnection]((error, connection) => {
                if (error) {
                    retryCallback(error);
                    return;
                }
                this[kGetApi]('DescribeClientQuotas', (error, api) => {
                    if (error) {
                        retryCallback(error);
                        return;
                    }
                    api(connection, options.components, options.strict ?? false, retryCallback);
                });
            });
        }, (error, response) => {
            if (error) {
                callback(new MultipleErrors('Describing client quotas failed.', [error]));
                return;
            }
            callback(null, response.entries);
        }, 0);
    }
    #alterClientQuotas(options, callback) {
        this[kPerformWithRetry]('alterClientQuotas', retryCallback => {
            this[kGetBootstrapConnection]((error, connection) => {
                if (error) {
                    retryCallback(error);
                    return;
                }
                this[kGetApi]('AlterClientQuotas', (error, api) => {
                    if (error) {
                        retryCallback(error);
                        return;
                    }
                    api(connection, options.entries, options.validateOnly ?? false, retryCallback);
                });
            });
        }, (error, response) => {
            if (error) {
                callback(new MultipleErrors('Altering client quotas failed.', [error]));
                return;
            }
            callback(null, response.entries);
        }, 0);
    }
    #describeLogDirs(options, callback) {
        this[kMetadata]({ topics: [] }, (error, metadata) => {
            /* c8 ignore next 4 - Hard to test */
            if (error) {
                callback(error);
                return;
            }
            runConcurrentCallbacks('Describing log dirs failed.', metadata.brokers, ([id, broker], concurrentCallback) => {
                this[kGetConnection](broker, (error, connection) => {
                    if (error) {
                        concurrentCallback(error);
                        return;
                    }
                    this[kPerformWithRetry]('describeLogDirs', retryCallback => {
                        this[kGetApi]('DescribeLogDirs', (error, api) => {
                            if (error) {
                                retryCallback(error);
                                return;
                            }
                            api(connection, options.topics, retryCallback);
                        });
                    }, (error, response) => {
                        if (error) {
                            concurrentCallback(error);
                            return;
                        }
                        concurrentCallback(null, {
                            broker: id,
                            throttleTimeMs: response.throttleTimeMs,
                            results: response.results.map(result => ({
                                logDir: result.logDir,
                                topics: result.topics,
                                totalBytes: result.totalBytes,
                                usableBytes: result.usableBytes
                            }))
                        });
                    }, 0);
                });
            }, callback);
        });
    }
    #listConsumerGroupOffsets(options, callback) {
        this[kMetadata]({ topics: [] }, (error, metadata) => {
            if (error) {
                callback(error);
                return;
            }
            /* c8 ignore next - Hard to test */
            const groupIds = options.groups.map(group => (typeof group === 'string' ? group : group.groupId));
            this.#findGroupCoordinator(groupIds, (error, response) => {
                if (error) {
                    callback(new MultipleErrors('Listing consumer group offsets failed.', [error]));
                    return;
                }
                const coordinators = new Map();
                for (const { key: groupId, nodeId: node } of response.coordinators) {
                    const groupRequest = {
                        groupId,
                        memberId: null,
                        memberEpoch: -1
                    };
                    for (const group of options.groups) {
                        if (typeof group !== 'string' && group.groupId === groupId && !!group.topics && group.topics.length > 0) {
                            groupRequest.topics = group.topics;
                            break;
                        }
                    }
                    let coordinator = coordinators.get(node);
                    if (!coordinator) {
                        coordinator = [];
                        coordinators.set(node, coordinator);
                    }
                    coordinator.push(groupRequest);
                }
                runConcurrentCallbacks('Listing consumer group offsets failed.', coordinators, ([node, groups], concurrentCallback) => {
                    this[kGetConnection](metadata.brokers.get(node), (error, connection) => {
                        if (error) {
                            concurrentCallback(error);
                            return;
                        }
                        this[kPerformWithRetry]('offsetFetch', retryCallback => {
                            this[kGetApi]('OffsetFetch', (error, api) => {
                                if (error) {
                                    retryCallback(error);
                                    return;
                                }
                                /* c8 ignore next - Hard to test */
                                api(connection, groups, options.requireStable ?? false, retryCallback);
                            });
                        }, concurrentCallback, 0);
                    });
                }, (error, responses) => {
                    if (error) {
                        callback(error);
                        return;
                    }
                    callback(null, responses.flatMap(r => r.groups.map(group => ({
                        groupId: group.groupId,
                        topics: group.topics.map(topic => ({
                            name: topic.name,
                            partitions: topic.partitions.map(p => ({
                                partitionIndex: p.partitionIndex,
                                committedOffset: p.committedOffset,
                                committedLeaderEpoch: p.committedLeaderEpoch,
                                metadata: p.metadata
                            }))
                        }))
                    }))));
                });
            });
        });
    }
    #alterConsumerGroupOffsets(options, callback) {
        this[kMetadata]({ topics: [] }, (error, metadata) => {
            if (error) {
                callback(error, undefined);
                return;
            }
            this.#findGroupCoordinator([options.groupId], (error, response) => {
                if (error) {
                    callback(new MultipleErrors('Altering consumer group offsets failed.', [error]));
                    return;
                }
                const coordinator = response.coordinators.find(c => c.key === options.groupId);
                /* c8 ignore next 9 - Hard to test */
                if (!coordinator) {
                    callback(new MultipleErrors('Altering consumer group offsets failed.', [
                        new Error(`No coordinator found for group ${options.groupId}`)
                    ]));
                    return;
                }
                const broker = metadata.brokers.get(coordinator.nodeId);
                /* c8 ignore next 9 - Hard to test */
                if (!broker) {
                    callback(new MultipleErrors('Altering consumer group offsets failed.', [
                        new Error(`Broker ${coordinator.nodeId} not found`)
                    ]));
                    return;
                }
                this[kGetConnection](broker, (error, connection) => {
                    if (error) {
                        callback(new MultipleErrors('Altering consumer group offsets failed.', [error]));
                        return;
                    }
                    this[kPerformWithRetry]('alterConsumerGroupOffsets', retryCallback => {
                        this[kGetApi]('OffsetCommit', (error, api) => {
                            if (error) {
                                retryCallback(error);
                                return;
                            }
                            const topics = [];
                            for (const topic of options.topics) {
                                const partitions = topic.partitionOffsets.map(p => ({
                                    partitionIndex: p.partition,
                                    committedOffset: p.offset,
                                    committedLeaderEpoch: -1,
                                    committedMetadata: null
                                }));
                                topics.push({ name: topic.name, partitions });
                            }
                            api(connection, options.groupId, -1, '', null, topics, retryCallback);
                        });
                    }, error => {
                        if (error) {
                            callback(new MultipleErrors('Altering consumer group offsets failed.', [error]));
                            return;
                        }
                        callback(null);
                    }, 0);
                });
            });
        });
    }
    #deleteConsumerGroupOffsets(options, callback) {
        this[kMetadata]({ topics: [] }, (error, metadata) => {
            if (error) {
                callback(error);
                return;
            }
            this.#findGroupCoordinator([options.groupId], (error, response) => {
                if (error) {
                    callback(new MultipleErrors('Deleting consumer group offsets failed.', [error]));
                    return;
                }
                const coordinator = response.coordinators.find(c => c.key === options.groupId);
                /* c8 ignore next 9 - Hard to test */
                if (!coordinator) {
                    callback(new MultipleErrors('Deleting consumer group offsets failed.', [
                        new Error(`No coordinator found for group ${options.groupId}`)
                    ]));
                    return;
                }
                const broker = metadata.brokers.get(coordinator.nodeId);
                /* c8 ignore next 9 - Hard to test */
                if (!broker) {
                    callback(new MultipleErrors('Deleting consumer group offsets failed.', [
                        new Error(`Broker ${coordinator.nodeId} not found`)
                    ]));
                    return;
                }
                this[kGetConnection](broker, (error, connection) => {
                    if (error) {
                        callback(new MultipleErrors('Deleting consumer group offsets failed.', [error]));
                        return;
                    }
                    this[kPerformWithRetry]('deleteOffset', retryCallback => {
                        this[kGetApi]('OffsetDelete', (error, api) => {
                            if (error) {
                                retryCallback(error);
                                return;
                            }
                            api(connection, options.groupId, options.topics.map(t => ({
                                name: t.name,
                                partitions: t.partitionIndexes.map(p => ({ partitionIndex: p }))
                            })), retryCallback);
                        });
                    }, (error, response) => {
                        if (error) {
                            callback(new MultipleErrors('Deleting consumer group offsets failed.', [error]));
                            return;
                        }
                        callback(null, response.topics.map(topic => ({
                            name: topic.name,
                            partitionIndexes: topic.partitions.map(p => p.partitionIndex)
                        })));
                    }, 0);
                });
            });
        });
    }
    #getConfigRequestsDistributedToBrokers(resources) {
        const brokerConfigResourceMap = new Map();
        for (const resource of resources) {
            if (resource.resourceType === ConfigResourceTypes.BROKER ||
                resource.resourceType === ConfigResourceTypes.BROKER_LOGGER) {
                if (!brokerConfigResourceMap.has(Number(resource.resourceName))) {
                    brokerConfigResourceMap.set(Number(resource.resourceName), []);
                }
                brokerConfigResourceMap.get(Number(resource.resourceName)).push(resource);
            }
            else {
                if (!brokerConfigResourceMap.has(-1)) {
                    brokerConfigResourceMap.set(-1, []);
                }
                brokerConfigResourceMap.get(-1).push(resource);
            }
        }
        return brokerConfigResourceMap;
    }
    #getAnyOrSpecificBrokerConnection(brokerId, callback) {
        if (brokerId === -1) {
            this[kGetBootstrapConnection](callback);
            return;
        }
        this[kMetadata]({ topics: [] }, (error, metadata) => {
            /* c8 ignore next 4 - Hard to test */
            if (error) {
                callback(error);
                return;
            }
            const brokerInstance = metadata.brokers.get(brokerId);
            /* c8 ignore next 4 - Hard to test */
            if (!brokerInstance) {
                callback(new Error(`Broker with id ${brokerId} not found in cluster.`));
                return;
            }
            this[kGetConnection](brokerInstance, callback);
        });
    }
    #describeConfigs(options, callback) {
        runConcurrentCallbacks('Describing configs failed.', this.#getConfigRequestsDistributedToBrokers(options.resources), ([brokerId, resources], concurrentCallback) => {
            this.#describeConfigsOnBroker({ ...options, resources }, brokerId, (error, response) => {
                if (error) {
                    concurrentCallback(error);
                    return;
                }
                concurrentCallback(null, response);
            });
        }, (error, results) => {
            callback(error, results?.flat());
        });
    }
    #describeConfigsOnBroker(options, broker, callback) {
        if (typeof broker === 'number') {
            this.#getAnyOrSpecificBrokerConnection(broker, (error, connection) => {
                /* c8 ignore next 4 - Hard to test */
                if (error) {
                    callback(error);
                    return;
                }
                this.#describeConfigsOnBroker(options, connection, callback);
            });
            return;
        }
        this[kPerformWithRetry]('describeConfigs', retryCallback => {
            this[kGetApi]('DescribeConfigs', (error, api) => {
                if (error) {
                    retryCallback(error);
                    return;
                }
                api(broker, options.resources, options.includeSynonyms ?? false, options.includeDocumentation ?? false, retryCallback);
            });
        }, (error, response) => {
            if (error) {
                callback(error);
                return;
            }
            const resultsWithoutErrors = response.results.map(result => ({
                resourceType: result.resourceType,
                resourceName: result.resourceName,
                configs: result.configs
            }));
            callback(null, resultsWithoutErrors);
        }, 0);
    }
    #alterConfigs(options, callback) {
        runConcurrentCallbacks('Altering configs failed.', this.#getConfigRequestsDistributedToBrokers(options.resources), ([brokerId, resources], concurrentCallback) => {
            this.#alterConfigsOnBroker({ ...options, resources }, brokerId, concurrentCallback);
        }, error => {
            callback(error);
        });
    }
    #alterConfigsOnBroker(options, broker, callback) {
        if (typeof broker === 'number') {
            this.#getAnyOrSpecificBrokerConnection(broker, (error, connection) => {
                /* c8 ignore next 4 - Hard to test */
                if (error) {
                    callback(error);
                    return;
                }
                this.#alterConfigsOnBroker(options, connection, callback);
            });
            return;
        }
        this[kPerformWithRetry]('alterConfigs', retryCallback => {
            this[kGetApi]('AlterConfigs', (error, api) => {
                if (error) {
                    retryCallback(error);
                    return;
                }
                api(broker, options.resources, options.validateOnly ?? false, retryCallback);
            });
        }, callback, 0);
    }
    #incrementalAlterConfigs(options, callback) {
        runConcurrentCallbacks('Incrementally altering configs failed.', this.#getConfigRequestsDistributedToBrokers(options.resources), ([brokerId, resources], concurrentCallback) => {
            this.#incrementalAlterConfigsOnBroker({ ...options, resources }, brokerId, concurrentCallback);
        }, error => {
            callback(error);
        });
    }
    #incrementalAlterConfigsOnBroker(options, broker, callback) {
        if (typeof broker === 'number') {
            this.#getAnyOrSpecificBrokerConnection(broker, (error, connection) => {
                /* c8 ignore next 4 - Hard to test */
                if (error) {
                    callback(error);
                    return;
                }
                this.#incrementalAlterConfigsOnBroker(options, connection, callback);
            });
            return;
        }
        this[kPerformWithRetry]('incrementalAlterConfigs', retryCallback => {
            this[kGetApi]('IncrementalAlterConfigs', (error, api) => {
                if (error) {
                    retryCallback(error);
                    return;
                }
                api(broker, options.resources, options.validateOnly ?? false, retryCallback);
            });
        }, callback, 0);
    }
    #createAcls(options, callback) {
        this[kPerformWithRetry]('createAcls', retryCallback => {
            this[kGetBootstrapConnection]((error, connection) => {
                if (error) {
                    retryCallback(error, undefined);
                    return;
                }
                this[kGetApi]('CreateAcls', (error, api) => {
                    if (error) {
                        retryCallback(error, undefined);
                        return;
                    }
                    api(connection, options.creations, retryCallback);
                });
            });
        }, error => {
            if (error) {
                callback(new MultipleErrors('Creating ACLs failed.', [error]), undefined);
                return;
            }
            callback(null, undefined);
        }, 0);
    }
    #describeAcls(options, callback) {
        this[kPerformWithRetry]('describeAcls', retryCallback => {
            this[kGetBootstrapConnection]((error, connection) => {
                if (error) {
                    retryCallback(error);
                    return;
                }
                this[kGetApi]('DescribeAcls', (error, api) => {
                    if (error) {
                        retryCallback(error);
                        return;
                    }
                    api(connection, options.filter, retryCallback);
                });
            });
        }, (error, response) => {
            if (error) {
                callback(new MultipleErrors('Describing ACLs failed.', [error]));
                return;
            }
            callback(null, response.resources);
        }, 0);
    }
    #deleteAcls(options, callback) {
        this[kPerformWithRetry]('deleteAcls', retryCallback => {
            this[kGetBootstrapConnection]((error, connection) => {
                if (error) {
                    retryCallback(error);
                    return;
                }
                this[kGetApi]('DeleteAcls', (error, api) => {
                    if (error) {
                        retryCallback(error);
                        return;
                    }
                    api(connection, options.filters, retryCallback);
                });
            });
        }, (error, response) => {
            if (error) {
                callback(new MultipleErrors('Deleting ACLs failed.', [error]));
                return;
            }
            callback(null, response.filterResults.flatMap(results => results.matchingAcls.map(acl => {
                return {
                    resourceType: acl.resourceType,
                    resourceName: acl.resourceName,
                    resourcePatternType: acl.resourcePatternType,
                    principal: acl.principal,
                    host: acl.host,
                    operation: acl.operation,
                    permissionType: acl.permissionType
                };
            })));
        }, 0);
    }
    #deleteRecords(options, callback) {
        this[kMetadata]({ topics: options.topics.map(topic => topic.name) }, (error, metadata) => {
            if (error) {
                callback(error);
                return;
            }
            const requestsByLeader = new Map();
            for (const topic of options.topics) {
                for (const partition of topic.partitions) {
                    const { leader } = metadata.topics.get(topic.name).partitions[partition.partition];
                    let leaderRequests = requestsByLeader.get(leader);
                    if (!leaderRequests) {
                        leaderRequests = new Map();
                        requestsByLeader.set(leader, leaderRequests);
                    }
                    let topicRequest = leaderRequests.get(topic.name);
                    if (!topicRequest) {
                        topicRequest = { name: topic.name, partitions: [] };
                        leaderRequests.set(topic.name, topicRequest);
                    }
                    topicRequest.partitions.push({
                        partitionIndex: partition.partition,
                        offset: partition.offset
                    });
                }
            }
            const requests = new Map();
            for (const [leader, topics] of requestsByLeader) {
                requests.set(leader, Array.from(topics.values()));
            }
            runConcurrentCallbacks('Deleting records failed.', requests, ([leader, requests], concurrentCallback) => {
                this[kGetConnection](metadata.brokers.get(leader), (error, connection) => {
                    if (error) {
                        concurrentCallback(error);
                        return;
                    }
                    this[kPerformWithRetry]('deleteRecords', retryCallback => {
                        this[kGetApi]('DeleteRecords', (error, api) => {
                            if (error) {
                                retryCallback(error);
                                return;
                            }
                            api(connection, Array.from(requests.values()), this[kOptions].timeout, retryCallback);
                        });
                    }, concurrentCallback, 0);
                });
            }, (error, responses) => {
                if (error) {
                    callback(error);
                    return;
                }
                const deletedRecordsByTopic = new Map();
                for (const response of responses) {
                    for (const topic of response.topics) {
                        let topicDeletedRecords = deletedRecordsByTopic.get(topic.name);
                        if (!topicDeletedRecords) {
                            topicDeletedRecords = { name: topic.name, partitions: [] };
                            deletedRecordsByTopic.set(topic.name, topicDeletedRecords);
                        }
                        for (const partition of topic.partitions) {
                            topicDeletedRecords.partitions.push({
                                partition: partition.partitionIndex,
                                lowWatermark: partition.lowWatermark
                            });
                        }
                    }
                }
                callback(null, Array.from(deletedRecordsByTopic.values()));
            });
        });
    }
    #listOffsets(options, callback) {
        this[kMetadata]({ topics: options.topics.map(topic => topic.name) }, (error, metadata) => {
            if (error) {
                callback(error);
                return;
            }
            const requests = new Map();
            for (const topic of options.topics) {
                for (const partition of topic.partitions) {
                    const { leader, leaderEpoch } = metadata.topics.get(topic.name).partitions[partition.partitionIndex];
                    let leaderRequests = requests.get(leader);
                    if (!leaderRequests) {
                        leaderRequests = [];
                        requests.set(leader, leaderRequests);
                    }
                    let topicRequest = leaderRequests.find(t => t.name === topic.name);
                    if (!topicRequest) {
                        topicRequest = { name: topic.name, partitions: [] };
                        leaderRequests.push(topicRequest);
                    }
                    topicRequest.partitions.push({
                        partitionIndex: partition.partitionIndex,
                        currentLeaderEpoch: leaderEpoch,
                        /* c8 ignore next - Hard to test */
                        timestamp: partition.timestamp ?? -1n
                    });
                }
            }
            runConcurrentCallbacks('Listing offsets failed.', requests, ([leader, requests], concurrentCallback) => {
                this[kGetConnection](metadata.brokers.get(leader), (error, connection) => {
                    if (error) {
                        concurrentCallback(error);
                        return;
                    }
                    this[kPerformWithRetry]('listOffsets', retryCallback => {
                        this[kGetApi]('ListOffsets', (error, api) => {
                            if (error) {
                                retryCallback(error);
                                return;
                            }
                            api(connection, -1, options.isolationLevel ?? FetchIsolationLevels.READ_UNCOMMITTED, Array.from(requests.values()), retryCallback);
                        });
                    }, concurrentCallback, 0);
                });
            }, (error, responses) => {
                if (error) {
                    callback(error);
                    return;
                }
                const ret = [];
                for (const response of responses) {
                    for (const topic of response.topics) {
                        let topicOffsets = ret.find(t => t.name === topic.name);
                        if (!topicOffsets) {
                            topicOffsets = { name: topic.name, partitions: [] };
                            ret.push(topicOffsets);
                        }
                        for (const partition of topic.partitions) {
                            topicOffsets.partitions.push({
                                offset: partition.offset,
                                timestamp: partition.timestamp,
                                partitionIndex: partition.partitionIndex,
                                leaderEpoch: partition.leaderEpoch
                            });
                        }
                    }
                }
                callback(null, ret);
            });
        });
    }
}
