// effect/dist/testing/TestSchema.js
/**
 * Testing utilities for Schema validation and assertions.
 *
 * @since 4.0.0
 */
import * as assert from "node:assert";
import * as Effect from "../Effect.js";
import * as Record from "../Record.js";
import * as Result from "../Result.js";
import * as Schema from "../Schema.js";
import * as AST from "../SchemaAST.js";
import * as Parser from "../SchemaParser.js";
import * as FastCheck from "../testing/FastCheck.js";
/**
 * The `Asserts` class provides a comprehensive testing framework for schema validation,
 * encoding/decoding operations, and property-based testing.
 *
 * @since 4.0.0
 */
export class Asserts {
  static ast = {
    /** Asserts that two struct field definitions are equal by comparing their AST representations. */
    fields: {
      equals: (a, b) => {
        assert.deepStrictEqual(Record.map(a, AST.getAST), Record.map(b, AST.getAST));
      }
    },
    /** Asserts that two tuple element definitions are equal by comparing their AST representations. */
    elements: {
      equals: (a, b) => {
        assert.deepStrictEqual(a.map(AST.getAST), b.map(AST.getAST));
      }
    }
  };
  schema;
  constructor(schema) {
    this.schema = schema;
  }
  /**
   * Provides make operation testing utilities.
   */
  make(options) {
    const makeEffect = Parser.makeEffect(this.schema);
    async function succeed(input, expected) {
      const r = await Effect.runPromise(makeEffect(input, options).pipe(Effect.mapError(issue => issue.toString()), Effect.result));
      expected = arguments.length === 1 ? input : expected;
      assert.deepStrictEqual(r, Result.succeed(expected));
    }
    return {
      succeed,
      /**
       * Asserts that make operation fails with the expected error message.
       */
      async fail(input, message) {
        const r = await Effect.runPromise(makeEffect(input, options).pipe(Effect.mapError(issue => issue.toString()), Effect.result));
        assert.deepStrictEqual(r, Result.fail(message));
      }
    };
  }
  /**
   * Verifies that encoding values produces the same type after decoding (round-trip).
   */
  verifyLosslessTransformation(options) {
    const decodeUnknownEffect = Parser.decodeUnknownEffect(this.schema);
    const encodeEffect = Parser.encodeEffect(this.schema);
    const arbitrary = Schema.toArbitrary(this.schema);
    return FastCheck.assert(FastCheck.asyncProperty(arbitrary, async t => {
      const r = await Effect.runPromise(encodeEffect(t).pipe(Effect.flatMap(e => decodeUnknownEffect(e)), Effect.mapError(issue => issue.toString()), Effect.result));
      assert.deepStrictEqual(r, Result.succeed(t));
    }), options?.params);
  }
  /**
   * Provides decoding testing utilities.
   */
  decoding(options) {
    return new Decoding(this.schema, options);
  }
  /**
   * Provides encoding testing utilities.
   */
  encoding(options) {
    return new Encoding(this.schema, options);
  }
  /**
   * Provides property-based testing utilities for schema validation.
   */
  arbitrary() {
    const schema = this.schema;
    return {
      /**
       * Verifies that the schema generates valid arbitrary values that satisfy
       * the schema constraints.
       */
      verifyGeneration(options) {
        const params = options?.params;
        const is = Schema.is(schema);
        const arb = Schema.toArbitrary(schema);
        FastCheck.assert(FastCheck.property(arb, a => is(a)), {
          numRuns: 20,
          ...params
        });
      }
    };
  }
}
/**
 * @since 4.0.0
 */
export class Decoding {
  schema;
  decodeUnknownEffect;
  options;
  constructor(schema, options) {
    this.schema = schema;
    this.decodeUnknownEffect = Parser.decodeUnknownEffect(schema);
    this.options = options;
  }
  async succeed(input, expected) {
    const r = await Effect.runPromise(this.decodeUnknownEffect(input, this.options?.parseOptions).pipe(Effect.mapError(issue => issue.toString()), Effect.result));
    expected = arguments.length === 1 ? input : expected;
    assert.deepStrictEqual(r, Result.succeed(expected));
  }
  /**
   * Asserts that decoding fails with the expected error message.
   */
  async fail(input, message) {
    const r = await Effect.runPromise(this.decodeUnknownEffect(input, this.options?.parseOptions).pipe(Effect.mapError(issue => issue.toString()), Effect.result));
    assert.deepStrictEqual(r, Result.fail(message));
  }
  /**
   * Creates a new Decoding instance with the provided service available in the
   * decoding context.
   */
  provide(service, implementation) {
    return new Decoding(this.schema.pipe(Schema.middlewareDecoding(Effect.provideService(service, implementation))), this.options);
  }
}
/**
 * @since 4.0.0
 */
class Encoding {
  schema;
  encodeUnknownEffect;
  options;
  constructor(schema, options) {
    this.schema = schema;
    this.encodeUnknownEffect = Parser.encodeUnknownEffect(schema);
    this.options = options;
  }
  async succeed(input, expected) {
    const r = await Effect.runPromise(this.encodeUnknownEffect(input, this.options?.parseOptions).pipe(Effect.mapError(issue => issue.toString()), Effect.result));
    expected = arguments.length === 1 ? input : expected;
    assert.deepStrictEqual(r, Result.succeed(expected));
  }
  /**
   * Asserts that encoding fails with the expected error message.
   */
  async fail(input, message) {
    const r = await Effect.runPromise(this.encodeUnknownEffect(input, this.options?.parseOptions).pipe(Effect.mapError(issue => issue.toString()), Effect.result));
    assert.deepStrictEqual(r, Result.fail(message));
  }
  /**
   * Creates a new Encoding instance with the provided service available in the
   * encoding context.
   */
  provide(service, implementation) {
    return new Encoding(this.schema.pipe(Schema.middlewareEncoding(Effect.provideService(service, implementation))), this.options);
  }
}
