// @platformatic/kafka/dist/protocol/compression.js
import { DynamicBuffer } from '@platformatic/dynamic-buffer';
import { lz4Compress, lz4Decompress, snappyCompress, snappyDecompress } from '@platformatic/wasm-utils';
import zlib from 'node:zlib';
import { UnsupportedCompressionError } from "../errors.js";
const { zstdCompressSync, zstdDecompressSync, gzipSync, gunzipSync } = zlib;
export const CompressionAlgorithms = {
    NONE: 'none',
    GZIP: 'gzip',
    SNAPPY: 'snappy',
    LZ4: 'lz4',
    ZSTD: 'zstd'
};
export const allowedCompressionsAlgorithms = Object.values(CompressionAlgorithms);
function ensureBuffer(data) {
    return DynamicBuffer.isDynamicBuffer(data) ? data.buffer : data;
}
const snappyCompressSync = snappyCompress;
const snappyDecompressSync = snappyDecompress;
const lz4CompressFrameSync = lz4Compress;
const lz4DecompressFrameSync = lz4Decompress;
export const compressionsAlgorithms = {
    /* c8 ignore next 8 - 'none' is actually never used but this is to please Typescript */
    none: {
        compressSync(data) {
            return ensureBuffer(data);
        },
        decompressSync(data) {
            return ensureBuffer(data);
        },
        bitmask: 0,
        available: true
    },
    gzip: {
        compressSync(data) {
            return gzipSync(ensureBuffer(data));
        },
        decompressSync(data) {
            return gunzipSync(ensureBuffer(data));
        },
        bitmask: 1,
        available: true
    },
    snappy: {
        compressSync(data) {
            return snappyCompressSync(ensureBuffer(data));
        },
        decompressSync(data) {
            return snappyDecompressSync(ensureBuffer(data));
        },
        bitmask: 2,
        available: true
    },
    lz4: {
        compressSync(data) {
            return lz4CompressFrameSync(ensureBuffer(data));
        },
        decompressSync(data) {
            return lz4DecompressFrameSync(ensureBuffer(data));
        },
        bitmask: 3,
        available: true
    },
    zstd: {
        /* c8 ignore next 7 - Tests are only run on Node.js versions that support zstd */
        compressSync(data) {
            if (!zstdCompressSync) {
                throw new UnsupportedCompressionError('zstd is not supported in the current Node.js version');
            }
            return zstdCompressSync(ensureBuffer(data));
        },
        /* c8 ignore next 7 - Tests are only run on Node.js versions that support zstd */
        decompressSync(data) {
            if (!zstdCompressSync) {
                throw new UnsupportedCompressionError('zstd is not supported in the current Node.js version');
            }
            return zstdDecompressSync(ensureBuffer(data));
        },
        bitmask: 4,
        available: typeof zstdCompressSync === 'function'
    }
};
export const compressionsAlgorithmsByBitmask = Object.fromEntries(Object.values(compressionsAlgorithms).map(a => [a.bitmask, a]));
