// @platformatic/kafka/dist/registries/abstract.js
export function runAsyncSeries(operation, collection, index, callback) {
    operation(collection[index], error => {
        if (error) {
            callback(error);
            return;
        }
        else if (index === collection.length - 1) {
            callback(null);
            return;
        }
        runAsyncSeries(operation, collection, index + 1, callback);
    });
}
/* c8 ignore start */
export class AbstractSchemaRegistry {
    get(_) {
        throw new Error('AbstractSchemaRegistry.get() should be implemented in subclasses.');
    }
    fetch(_i, _c) {
        throw new Error('AbstractSchemaRegistry.fetch() should be implemented in subclasses.');
    }
    getSchemaId(_p, _t) {
        throw new Error('AbstractSchemaRegistry.getSchemaId() should be implemented in subclasses.');
    }
    getSerializers() {
        throw new Error('AbstractSchemaRegistry.getSerializers() should be implemented in subclasses.');
    }
    getDeserializers() {
        throw new Error('AbstractSchemaRegistry.getDeserializers() should be implemented in subclasses.');
    }
    getBeforeSerializationHook() {
        throw new Error('AbstractSchemaRegistry.getBeforeSerializationHook() should be implemented in subclasses.');
    }
    getBeforeDeserializationHook() {
        throw new Error('AbstractSchemaRegistry.getBeforeDeserializationHook() should be implemented in subclasses.');
    }
}
/* c8 ignore end */
