// @platformatic/kafka/dist/registries/confluent-schema-registry.js
import avro from 'avsc';
import { createRequire } from 'node:module';
import { jsonDeserializer, jsonSerializer, stringDeserializer, stringSerializer } from "../clients/serde.js";
import { ajv, EMPTY_BUFFER, UnsupportedFormatError, UserError } from "../index.js";
import { getCredential } from "../protocol/sasl/utils.js";
import { AbstractSchemaRegistry } from "./abstract.js";
const require = createRequire(import.meta.url);
/* c8 ignore next 8 */
export function defaultProtobufTypeMapper(_, type, context) {
    // Confluent Schema Registry convention
    return `${context.topic}-${type}`;
}
// TODO(ShogunPanda): Authentication support
export class ConfluentSchemaRegistry extends AbstractSchemaRegistry {
    #url;
    #schemas;
    #protobufParse;
    #protobufTypeMapper;
    #jsonValidateSend;
    #auth;
    constructor(options) {
        super();
        this.#url = options.url;
        this.#schemas = new Map();
        this.#protobufTypeMapper = options.protobufTypeMapper ?? defaultProtobufTypeMapper;
        this.#jsonValidateSend = options.jsonValidateSend ?? false;
        this.#auth = options.auth;
    }
    getSchemaId(message, type) {
        if (Buffer.isBuffer(message)) {
            if (type !== 'value') {
                return undefined;
            }
            return message.readInt32BE(1);
        }
        return message.metadata?.schemas?.[type];
    }
    get(id) {
        return this.#schemas.get(id);
    }
    async fetchSchema(id, callback) {
        try {
            const requestInit = {};
            if (this.#auth) {
                if (this.#auth.token) {
                    const token = await getCredential('token', this.#auth.token);
                    requestInit.headers = {
                        Authorization: `Bearer ${token}`
                    };
                }
                else {
                    const username = await getCredential('username', this.#auth.username);
                    const password = await getCredential('password', this.#auth.password);
                    requestInit.headers = {
                        Authorization: `Basic ${Buffer.from(`${username}:${password}`).toString('base64')}`
                    };
                }
            }
            const response = await fetch(`${this.#url}/schemas/ids/${id}`, requestInit);
            if (!response.ok) {
                throw new UserError(`Failed to fetch a schema: [HTTP ${response.status}]`, { response: await response.text() });
            }
            const responseBody = await response.json();
            const { schema, schemaType } = responseBody;
            switch (schemaType) {
                case 'AVRO':
                    this.#schemas.set(id, { id, type: 'avro', schema: avro.Type.forSchema(JSON.parse(schema)) });
                    break;
                case 'PROTOBUF':
                    this.#protobufParse ??= this.#loadProtobuf();
                    this.#schemas.set(id, { id, type: 'protobuf', schema: this.#protobufParse(schema).root });
                    break;
                case 'JSON':
                    this.#schemas.set(id, { id, type: 'json', schema: ajv.compile(JSON.parse(schema)) });
                    break;
            }
            process.nextTick(callback);
        }
        catch (err) {
            process.nextTick(() => callback(err));
        }
    }
    getSerializers() {
        return {
            key: this.#schemaSerializer.bind(this, 'key', stringSerializer),
            value: this.#schemaSerializer.bind(this, 'value', jsonSerializer),
            headerKey: this.#schemaSerializer.bind(this, 'headerKey', stringSerializer),
            headerValue: this.#schemaSerializer.bind(this, 'headerValue', jsonSerializer)
        };
    }
    getDeserializers() {
        return {
            key: this.#schemaDeserializer.bind(this, 'key', stringDeserializer),
            value: this.#schemaDeserializer.bind(this, 'value', jsonDeserializer),
            headerKey: this.#schemaDeserializer.bind(this, 'headerKey', stringDeserializer),
            headerValue: this.#schemaDeserializer.bind(this, 'headerValue', jsonDeserializer)
        };
    }
    getBeforeSerializationHook() {
        const registry = this;
        return function beforeSerialization(_, type, message, callback) {
            // Extract the schema ID from the message metadata
            const schemaId = registry.getSchemaId(message, type);
            // When no schema ID is found, nothing to do
            if (!schemaId) {
                callback(null);
                return;
            }
            // The schema is already fetch
            if (registry.get(schemaId)) {
                callback(null);
                return;
            }
            registry.fetchSchema(schemaId, callback);
        };
    }
    getBeforeDeserializationHook() {
        const registry = this;
        return function beforeDeserialization(payload, type, _message, callback) {
            // Extract the schema ID from the message metadata
            const schemaId = registry.getSchemaId(payload, type);
            // When no schema ID is found, nothing to do
            if (!schemaId) {
                callback(null);
                return;
            }
            // The schema is already fetch
            if (registry.get(schemaId)) {
                callback(null);
                return;
            }
            registry.fetchSchema(schemaId, callback);
        };
    }
    #schemaSerializer(type, fallbackSerializer, data, headers, message) {
        /* c8 ignore next 3 - Hard to test */
        if (typeof data === 'undefined') {
            return EMPTY_BUFFER;
        }
        if (type === 'headerKey' || type === 'headerValue') {
            message = headers;
        }
        const schemaId = message?.metadata?.schemas?.[type];
        if (!schemaId) {
            return fallbackSerializer(data);
        }
        const schema = this.#schemas.get(schemaId);
        if (!schema) {
            throw new UserError(`Schema with ID ${schemaId} not found.`, { missingSchema: schemaId });
        }
        let encodedMessage;
        switch (schema.type) {
            case 'avro':
                encodedMessage = schema.schema.toBuffer(data);
                break;
            case 'protobuf':
                {
                    const typeName = this.#protobufTypeMapper(schemaId, type, message);
                    const Type = schema.schema.lookupType(typeName);
                    encodedMessage = Buffer.from(Type.encode(Type.create(data)).finish());
                }
                break;
            case 'json':
                if (this.#jsonValidateSend) {
                    const validate = schema.schema;
                    const valid = validate(data);
                    if (!valid) {
                        throw new UserError(`JSON Schema validation failed before serialization: ${ajv.errorsText(validate.errors)}`, { type, data, headers, validationErrors: validate.errors });
                    }
                }
                encodedMessage = Buffer.from(JSON.stringify(data));
                break;
        }
        const header = Buffer.alloc(5);
        header.writeInt32BE(schemaId, 1);
        return Buffer.concat([header, encodedMessage]);
    }
    #schemaDeserializer(type, fallbackDeserializer, data, headers, message) {
        /* c8 ignore next 3 - Hard to test */
        if (typeof data === 'undefined' || data.length === 0) {
            return EMPTY_BUFFER;
        }
        if (type === 'headerKey' || type === 'headerValue') {
            message = headers;
        }
        const schemaId = this.getSchemaId(data, type);
        if (!schemaId) {
            return fallbackDeserializer(data);
        }
        const schema = this.#schemas.get(schemaId);
        if (!schema) {
            throw new UserError(`Schema with ID ${schemaId} not found.`, { missingSchema: schemaId });
        }
        switch (schema.type) {
            case 'avro':
                return schema.schema.fromBuffer(data.subarray(5));
            case 'protobuf': {
                const typeName = this.#protobufTypeMapper(schemaId, type, message);
                const Type = schema.schema.lookupType(typeName);
                return Type.decode(data.subarray(5));
            }
            case 'json': {
                const parsed = JSON.parse(data.subarray(5).toString('utf-8'));
                const validate = schema.schema;
                const valid = validate(parsed);
                if (!valid) {
                    throw new UserError(`JSON Schema validation failed before deserialization: ${ajv.errorsText(validate.errors)}`, { type, data: parsed, headers, validationErrors: validate.errors });
                }
                return parsed;
            }
        }
    }
    #loadProtobuf() {
        try {
            return require('protobufjs').parse;
            /* c8 ignore next 5 - In tests protobufjs is always available */
        }
        catch (e) {
            throw new UnsupportedFormatError('Cannot load protobufjs module, which is an optionalDependency. Please check your local installation.');
        }
    }
}
