// @platformatic/wasm-utils/dist/index.js
import { DynamicBuffer } from '@platformatic/dynamic-buffer';
import { readFileSync } from 'node:fs';
const FastBuffer = Buffer[Symbol.species];
const wasm = readFileSync(new URL('../dist/native.wasm', import.meta.url));
const wasmModule = new WebAssembly.Module(wasm);
const instance = new WebAssembly.Instance(wasmModule);
const { alloc, crc32c: _crc32c, dealloc, lz4_compress: _lz4Compress, lz4_decompress: _lz4Decompress, memory, snappy_compress: _snappyCompress, snappy_decompress: _snappyDecompress } = instance.exports;
let currentBufferLength = 4096;
let currentBufferOffset = alloc(currentBufferLength);
let currentBuffer = new Uint8Array(memory.buffer, currentBufferOffset, currentBufferLength);
export function prepareInput(data) {
    const input = DynamicBuffer.isDynamicBuffer(data) ? data.buffer : data;
    if (currentBuffer.length < data.length) {
        dealloc(currentBufferOffset, currentBufferLength);
        currentBufferLength = data.length;
        currentBufferOffset = alloc(currentBufferLength);
        currentBuffer = new Uint8Array(memory.buffer, currentBufferOffset, currentBufferLength);
    }
    currentBuffer.set(input);
}
export function prepareOutput(combined) {
    const len = Number(BigInt.asUintN(32, combined));
    const ptr = Number(combined >> 32n);
    const output = Buffer.from(new FastBuffer(memory.buffer, ptr, len));
    dealloc(ptr, len);
    return output;
}
export function crc32c(data) {
    prepareInput(data);
    return _crc32c(currentBufferOffset, data.length) >>> 0;
}
export function lz4Compress(data) {
    prepareInput(data);
    return prepareOutput(_lz4Compress(currentBufferOffset, data.length));
}
export function lz4Decompress(data) {
    prepareInput(data);
    return prepareOutput(_lz4Decompress(currentBufferOffset, data.length));
}
export function snappyCompress(data) {
    prepareInput(data);
    return prepareOutput(_snappyCompress(currentBufferOffset, data.length));
}
export function snappyDecompress(data) {
    prepareInput(data);
    return prepareOutput(_snappyDecompress(currentBufferOffset, data.length));
}
