// @platformatic/kafka/dist/apis/enumerations.js
// SASL Authentication
export const SASLMechanisms = {
    PLAIN: 'PLAIN',
    SCRAM_SHA_256: 'SCRAM-SHA-256',
    SCRAM_SHA_512: 'SCRAM-SHA-512',
    OAUTHBEARER: 'OAUTHBEARER',
    GSSAPI: 'GSSAPI'
};
export const allowedSASLMechanisms = Object.values(SASLMechanisms);
// Metadata API
// ./metadata/find-coordinator.ts
export const FindCoordinatorKeyTypes = { GROUP: 0, TRANSACTION: 1, SHARE: 2 };
export const allowedFindCoordinatorKeyTypes = Object.values(FindCoordinatorKeyTypes);
// Producer API
export const ProduceAcks = {
    ALL: -1,
    NO_RESPONSE: 0,
    LEADER: 1
};
export const allowedProduceAcks = Object.values(ProduceAcks);
// Consumer API
export const GroupProtocols = { CLASSIC: 'classic', CONSUMER: 'consumer' };
export const allowedGroupProtocols = Object.values(GroupProtocols);
// ./consumer/fetch.ts
export const FetchIsolationLevels = { READ_UNCOMMITTED: 0, READ_COMMITTED: 1 };
export const allowedFetchIsolationLevels = Object.values(FetchIsolationLevels);
export const ListOffsetTimestamps = { LATEST: -1n, EARLIEST: -2n };
export const allowedListOffsetTimestamps = Object.values(ListOffsetTimestamps);
// Admin API
// ./admin/*-acls.ts - See: https://cwiki.apache.org/confluence/display/KAFKA/KIP-140%3A+Add+administrative+RPCs+for+adding%2C+deleting%2C+and+listing+ACLs
export const ResourceTypes = {
    UNKNOWN: 0,
    ANY: 1,
    TOPIC: 2,
    GROUP: 3,
    CLUSTER: 4,
    TRANSACTIONAL_ID: 5,
    DELEGATION_TOKEN: 6,
    USER: 7
};
export const allowedResourceTypes = Object.values(ResourceTypes);
export const ResourcePatternTypes = { UNKNOWN: 0, ANY: 1, MATCH: 2, LITERAL: 3, PREFIXED: 4 };
export const allowedResourcePatternTypes = Object.values(ResourcePatternTypes);
export const AclOperations = {
    UNKNOWN: 0,
    ANY: 1,
    ALL: 2,
    READ: 3,
    WRITE: 4,
    CREATE: 5,
    DELETE: 6,
    ALTER: 7,
    DESCRIBE: 8,
    CLUSTER_ACTION: 9,
    DESCRIBE_CONFIGS: 10,
    ALTER_CONFIGS: 11,
    IDEMPOTENT_WRITE: 12,
    CREATE_TOKENS: 13,
    DESCRIBE_TOKENS: 14,
    TWO_PHASE_COMMIT: 15
};
export const allowedAclOperations = Object.values(AclOperations);
export const AclPermissionTypes = { UNKNOWN: 0, ANY: 1, DENY: 2, ALLOW: 3 };
export const allowedAclPermissionTypes = Object.values(AclPermissionTypes);
// ./admin/*-configs.ts
export const ConfigSources = {
    UNKNOWN: 0,
    TOPIC_CONFIG: 1,
    DYNAMIC_BROKER_CONFIG: 2,
    DYNAMIC_DEFAULT_BROKER_CONFIG: 3,
    STATIC_BROKER_CONFIG: 4,
    DEFAULT_CONFIG: 5,
    DYNAMIC_BROKER_LOGGER_CONFIG: 6,
    CLIENT_METRICS_CONFIG: 7,
    GROUP_CONFIG: 8
};
export const allowedConfigSources = Object.values(ConfigSources);
export const ConfigResourceTypes = {
    UNKNOWN: 0,
    TOPIC: 2,
    BROKER: 4,
    BROKER_LOGGER: 8,
    CLIENT_METRICS: 16,
    GROUP: 32
};
export const allowedConfigResourceTypes = Object.values(ConfigResourceTypes);
export const ConfigTypes = {
    UNKNOWN: 0,
    BOOLEAN: 1,
    STRING: 2,
    INT: 3,
    SHORT: 4,
    LONG: 5,
    DOUBLE: 6,
    LIST: 7,
    CLASS: 8,
    PASSWORD: 9
};
export const allowedConfigTypes = Object.values(ConfigTypes);
export const IncrementalAlterConfigOperationTypes = { SET: 0, DELETE: 1, APPEND: 2, SUBTRACT: 3 };
export const allowedIncrementalAlterConfigOperationTypes = Object.values(IncrementalAlterConfigOperationTypes);
// ./admin/*-client-quotas.ts
export const ClientQuotaMatchTypes = { EXACT: 0, DEFAULT: 1, ANY: 2 };
export const allowedClientQuotaMatchTypes = Object.values(ClientQuotaMatchTypes);
export const ClientQuotaEntityTypes = { CLIENT_ID: 'client-id', USER: 'user' };
export const allowedClientQuotaEntityTypes = Object.values(ClientQuotaEntityTypes);
export const ClientQuotaKeys = {
    PRODUCER_BYTE_RATE: 'producer_byte_rate',
    CONSUMER_BYTE_RATE: 'consumer_byte_rate',
    REQUEST_PERCENTAGE: 'request_percentage'
};
export const allowedClientQuotaKeys = Object.values(ClientQuotaKeys);
// ./admin/*-scram-credentials.ts
export const ScramMechanisms = { UNKNOWN: 0, SCRAM_SHA_256: 1, SCRAM_SHA_512: 2 };
export const allowedScramMechanisms = Object.values(ScramMechanisms);
// ./admin/describe-cluster.ts
export const DescribeClusterEndpointTypes = { BROKERS: 1, CONTROLLERS: 2 };
export const allowedDescribeClusterEndpointTypes = Object.values(DescribeClusterEndpointTypes);
// ./admin/list-groups.ts
export const ConsumerGroupStates = ['PREPARING_REBALANCE', 'COMPLETING_REBALANCE', 'STABLE', 'DEAD', 'EMPTY'];
// ./admin/list-transactions.ts
export const TransactionStates = [
    'EMPTY',
    'ONGOING',
    'PREPARE_ABORT',
    'COMMITTING',
    'ABORTING',
    'COMPLETE_COMMIT',
    'COMPLETE_ABORT'
];
// ./admin/update-features.ts
export const FeatureUpgradeTypes = { UPGRADE: 1, SAFE_DOWNGRADE: 2, UNSAFE_DOWNGRADE: 3 };
export const allowedFeatureUpgradeTypes = Object.values(FeatureUpgradeTypes);
