// effect/dist/unstable/http/HttpServerRequest.js
import * as Channel from "../../Channel.js";
import * as Effect from "../../Effect.js";
import * as Inspectable from "../../Inspectable.js";
import * as Result from "../../Result.js";
import * as Schema from "../../Schema.js";
import * as ServiceMap from "../../ServiceMap.js";
import * as Stream from "../../Stream.js";
import * as Socket from "../socket/Socket.js";
import * as Cookies from "./Cookies.js";
import * as Headers from "./Headers.js";
import * as HttpIncomingMessage from "./HttpIncomingMessage.js";
import { hasBody } from "./HttpMethod.js";
import { HttpServerError, RequestParseError } from "./HttpServerError.js";
import * as Multipart from "./Multipart.js";
import * as UrlParams from "./UrlParams.js";
export {
/**
 * @since 4.0.0
 * @category fiber refs
 */
MaxBodySize } from "./HttpIncomingMessage.js";
/**
 * @since 4.0.0
 * @category Type IDs
 */
export const TypeId = "~effect/http/HttpServerRequest";
/**
 * @since 4.0.0
 * @category context
 */
export const HttpServerRequest = /*#__PURE__*/ServiceMap.Service("effect/http/HttpServerRequest");
/**
 * @since 4.0.0
 * @category search params
 */
export class ParsedSearchParams extends /*#__PURE__*/ServiceMap.Service()("effect/http/ParsedSearchParams") {}
/**
 * @since 4.0.0
 * @category search params
 */
export const searchParamsFromURL = url => {
  const out = {};
  for (const [key, value] of url.searchParams.entries()) {
    const entry = out[key];
    if (entry !== undefined) {
      if (Array.isArray(entry)) {
        entry.push(value);
      } else {
        out[key] = [entry, value];
      }
    } else {
      out[key] = value;
    }
  }
  return out;
};
/**
 * @since 4.0.0
 * @category accessors
 */
export const upgradeChannel = () => HttpServerRequest.asEffect().pipe(Effect.flatMap(_ => _.upgrade), Effect.map(Socket.toChannelWith()), Channel.unwrap);
/**
 * @since 4.0.0
 * @category schema
 */
export const schemaCookies = (schema, options) => {
  const parse = Schema.decodeUnknownEffect(schema);
  return Effect.flatMap(HttpServerRequest.asEffect(), req => parse(req.cookies, options));
};
/**
 * @since 4.0.0
 * @category schema
 */
export const schemaHeaders = (schema, options) => {
  const parse = Schema.decodeUnknownEffect(schema);
  return Effect.flatMap(HttpServerRequest.asEffect(), req => parse(req.headers, options));
};
/**
 * @since 4.0.0
 * @category schema
 */
export const schemaSearchParams = (schema, options) => {
  const parse = Schema.decodeUnknownEffect(schema);
  return Effect.flatMap(ParsedSearchParams.asEffect(), params => parse(params, options));
};
/**
 * @since 4.0.0
 * @category schema
 */
export const schemaBodyJson = (schema, options) => {
  const parse = HttpIncomingMessage.schemaBodyJson(schema, options);
  return Effect.flatMap(HttpServerRequest.asEffect(), parse);
};
const isMultipart = request => request.headers["content-type"]?.toLowerCase().includes("multipart/form-data");
/**
 * @since 4.0.0
 * @category schema
 */
export const schemaBodyForm = (schema, options) => {
  const parseMultipart = Multipart.schemaPersisted(schema);
  const parseUrlParams = HttpIncomingMessage.schemaBodyUrlParams(schema, options);
  return Effect.flatMap(HttpServerRequest.asEffect(), request => {
    if (isMultipart(request)) {
      return Effect.flatMap(request.multipart, _ => parseMultipart(_, options));
    }
    return parseUrlParams(request);
  });
};
/**
 * @since 4.0.0
 * @category schema
 */
export const schemaBodyUrlParams = (schema, options) => {
  const parse = HttpIncomingMessage.schemaBodyUrlParams(schema, options);
  return Effect.flatMap(HttpServerRequest.asEffect(), parse);
};
/**
 * @since 4.0.0
 * @category schema
 */
export const schemaBodyMultipart = (schema, options) => {
  const parse = Multipart.schemaPersisted(schema);
  return HttpServerRequest.asEffect().pipe(Effect.flatMap(_ => _.multipart), Effect.flatMap(_ => parse(_, options)));
};
/**
 * @since 4.0.0
 * @category schema
 */
export const schemaBodyFormJson = (schema, options) => {
  const parseMultipart = Multipart.schemaJson(schema, options);
  return field => {
    const parseUrlParams = UrlParams.schemaJsonField(field).pipe(Schema.decodeTo(schema), Schema.decodeEffect);
    return Effect.flatMap(HttpServerRequest.asEffect(), request => {
      if (isMultipart(request)) {
        return Effect.flatMap(Effect.mapError(request.multipart, cause => new HttpServerError({
          reason: new RequestParseError({
            request,
            cause
          })
        })), parseMultipart(field));
      }
      return Effect.flatMap(request.urlParamsBody, _ => parseUrlParams(_, options));
    });
  };
};
/**
 * @since 4.0.0
 * @category conversions
 */
export const fromWeb = request => new ServerRequestImpl(request, removeHost(request.url));
const removeHost = url => {
  if (url[0] === "/") {
    return url;
  }
  const index = url.indexOf("/", url.indexOf("//") + 2);
  return index === -1 ? "/" : url.slice(index);
};
class ServerRequestImpl extends Inspectable.Class {
  [TypeId];
  [HttpIncomingMessage.TypeId];
  source;
  url;
  headersOverride;
  remoteAddressOverride;
  constructor(source, url, headersOverride, remoteAddressOverride) {
    super();
    this[TypeId] = TypeId;
    this[HttpIncomingMessage.TypeId] = HttpIncomingMessage.TypeId;
    this.source = source;
    this.url = url;
    this.headersOverride = headersOverride;
    this.remoteAddressOverride = remoteAddressOverride;
  }
  toJSON() {
    return HttpIncomingMessage.inspect(this, {
      _id: "HttpServerRequest",
      method: this.method,
      url: this.originalUrl
    });
  }
  modify(options) {
    return new ServerRequestImpl(this.source, options.url ?? this.url, options.headers ?? this.headersOverride, options.remoteAddress ?? this.remoteAddressOverride);
  }
  get method() {
    return this.source.method.toUpperCase();
  }
  get originalUrl() {
    return this.source.url;
  }
  get remoteAddress() {
    return this.remoteAddressOverride ? this.remoteAddressOverride : undefined;
  }
  get headers() {
    this.headersOverride ??= Headers.fromInput(this.source.headers);
    return this.headersOverride;
  }
  cachedCookies;
  get cookies() {
    if (this.cachedCookies) {
      return this.cachedCookies;
    }
    return this.cachedCookies = Cookies.parseHeader(this.headers.cookie ?? "");
  }
  get stream() {
    return this.source.body ? Stream.fromReadableStream({
      evaluate: () => this.source.body,
      onError: cause => new HttpServerError({
        reason: new RequestParseError({
          request: this,
          cause
        })
      })
    }) : Stream.fail(new HttpServerError({
      reason: new RequestParseError({
        request: this,
        description: "can not create stream from empty body"
      })
    }));
  }
  textEffect;
  get text() {
    if (this.textEffect) {
      return this.textEffect;
    }
    this.textEffect = Effect.runSync(Effect.cached(Effect.tryPromise({
      try: () => this.source.text(),
      catch: cause => new HttpServerError({
        reason: new RequestParseError({
          request: this,
          cause
        })
      })
    })));
    return this.textEffect;
  }
  get json() {
    return Effect.flatMap(this.text, text => Effect.try({
      try: () => JSON.parse(text),
      catch: cause => new HttpServerError({
        reason: new RequestParseError({
          request: this,
          cause
        })
      })
    }));
  }
  get urlParamsBody() {
    return Effect.flatMap(this.text, _ => Effect.try({
      try: () => UrlParams.fromInput(new URLSearchParams(_)),
      catch: cause => new HttpServerError({
        reason: new RequestParseError({
          request: this,
          cause
        })
      })
    }));
  }
  multipartEffect;
  get multipart() {
    if (this.multipartEffect) {
      return this.multipartEffect;
    }
    this.multipartEffect = Effect.runSync(Effect.cached(Multipart.toPersisted(this.multipartStream)));
    return this.multipartEffect;
  }
  get multipartStream() {
    return Stream.pipeThroughChannel(Stream.mapError(this.stream, cause => Multipart.MultipartError.fromReason("InternalError", cause)), Multipart.makeChannel(this.headers));
  }
  arrayBufferEffect;
  get arrayBuffer() {
    if (this.arrayBufferEffect) {
      return this.arrayBufferEffect;
    }
    this.arrayBufferEffect = Effect.runSync(Effect.cached(Effect.tryPromise({
      try: () => this.source.arrayBuffer(),
      catch: cause => new HttpServerError({
        reason: new RequestParseError({
          request: this,
          cause
        })
      })
    })));
    return this.arrayBufferEffect;
  }
  get upgrade() {
    return Effect.fail(new HttpServerError({
      reason: new RequestParseError({
        request: this,
        description: "Not an upgradeable ServerRequest"
      })
    }));
  }
}
/**
 * @since 4.0.0
 * @category conversions
 */
export const toURL = self => {
  const host = self.headers.host ?? "localhost";
  const protocol = self.headers["x-forwarded-proto"] === "https" ? "https" : "http";
  try {
    return new URL(self.url, `${protocol}://${host}`);
  } catch {
    return undefined;
  }
};
/**
 * @since 4.0.0
 * @category conversions
 */
export const toWebResult = (self, options) => {
  if (self.source instanceof Request) {
    return Result.succeed(self.source);
  }
  const url = toURL(self);
  if (url === undefined) {
    return Result.fail(new RequestParseError({
      request: self,
      description: "Invalid URL"
    }));
  }
  const requestInit = {
    method: self.method,
    headers: self.headers
  };
  if (options?.signal) {
    requestInit.signal = options.signal;
  }
  if (hasBody(self.method)) {
    requestInit.body = Stream.toReadableStreamWith(self.stream, options?.services ?? ServiceMap.empty());
    requestInit.duplex = "half";
  }
  return Result.succeed(new Request(url, requestInit));
};
/**
 * @since 4.0.0
 * @category conversions
 */
export const toWeb = (self, options) => Effect.servicesWith(services => toWebResult(self, {
  services,
  signal: options?.signal
}).asEffect());
