// effect/dist/unstable/http/HttpMiddleware.js
/**
 * @since 4.0.0
 */
import { Clock } from "../../Clock.js";
import * as Effect from "../../Effect.js";
import { constant, constFalse } from "../../Function.js";
import * as internalEffect from "../../internal/effect.js";
import * as Layer from "../../Layer.js";
import { TracerEnabled } from "../../References.js";
import * as ServiceMap from "../../ServiceMap.js";
import { ParentSpan } from "../../Tracer.js";
import * as Headers from "./Headers.js";
import { causeResponseStripped, exitResponse } from "./HttpServerError.js";
import { HttpServerRequest } from "./HttpServerRequest.js";
import * as Request from "./HttpServerRequest.js";
import * as Response from "./HttpServerResponse.js";
import * as TraceContext from "./HttpTraceContext.js";
import { appendPreResponseHandlerUnsafe } from "./internal/preResponseHandler.js";
/**
 * @since 4.0.0
 * @category constructors
 */
export const make = middleware => middleware;
const loggerDisabledRequests = /*#__PURE__*/new WeakSet();
/**
 * @since 4.0.0
 * @category Logger
 */
export const withLoggerDisabled = self => Effect.withFiber(fiber => {
  const request = ServiceMap.getUnsafe(fiber.services, HttpServerRequest);
  loggerDisabledRequests.add(request.source);
  return self;
});
/**
 * @since 4.0.0
 * @category Tracer
 */
export const TracerDisabledWhen = /*#__PURE__*/ServiceMap.Reference("effect/http/HttpMiddleware/TracerDisabledWhen", {
  defaultValue: () => constFalse
});
/**
 * @since 4.0.0
 * @category Tracer
 */
export const layerTracerDisabledForUrls = urls => Layer.succeed(TracerDisabledWhen)(req => urls.includes(req.url));
/**
 * @since 4.0.0
 * @category Tracer
 */
export const SpanNameGenerator = /*#__PURE__*/ServiceMap.Reference("@effect/platform/HttpMiddleware/SpanNameGenerator", {
  defaultValue: () => request => `http.server ${request.method}`
});
/**
 * @since 4.0.0
 * @category Logger
 */
export const logger = /*#__PURE__*/make(httpApp => {
  let counter = 0;
  return Effect.withFiber(fiber => {
    const request = ServiceMap.getUnsafe(fiber.services, HttpServerRequest);
    return Effect.withLogSpan(Effect.flatMap(Effect.exit(httpApp), exit => {
      if (loggerDisabledRequests.has(request.source)) {
        return exit;
      } else if (exit._tag === "Failure") {
        const [response, cause] = causeResponseStripped(exit.cause);
        return Effect.andThen(Effect.annotateLogs(Effect.log(cause ?? "Sent HTTP Response"), {
          "http.method": request.method,
          "http.url": request.url,
          "http.status": response.status
        }), exit);
      }
      return Effect.andThen(Effect.annotateLogs(Effect.log("Sent HTTP response"), {
        "http.method": request.method,
        "http.url": request.url,
        "http.status": exit.value.status
      }), exit);
    }), `http.span.${++counter}`);
  });
});
/**
 * @since 4.0.0
 * @category Tracer
 */
export const tracer = /*#__PURE__*/make(httpApp => Effect.withFiber(fiber => {
  const request = ServiceMap.getUnsafe(fiber.services, HttpServerRequest);
  const disabled = !fiber.getRef(TracerEnabled) || fiber.getRef(TracerDisabledWhen)(request);
  if (disabled) {
    return httpApp;
  }
  const nameGenerator = fiber.getRef(SpanNameGenerator);
  const span = internalEffect.makeSpanUnsafe(fiber, nameGenerator(request), {
    parent: TraceContext.fromHeaders(request.headers),
    kind: "server"
  });
  const prevServices = fiber.services;
  fiber.setServices(ServiceMap.add(fiber.services, ParentSpan, span));
  return Effect.onExitPrimitive(httpApp, exit => {
    fiber.setServices(prevServices);
    const endTime = fiber.getRef(Clock).currentTimeNanosUnsafe();
    fiber.currentScheduler.scheduleTask(() => {
      const url = Request.toURL(request);
      if (url !== undefined && (url.username !== "" || url.password !== "")) {
        url.username = "REDACTED";
        url.password = "REDACTED";
      }
      const redactedHeaderNames = fiber.getRef(Headers.CurrentRedactedNames);
      const requestHeaders = Headers.redact(request.headers, redactedHeaderNames);
      span.attribute("http.request.method", request.method);
      if (url !== undefined) {
        span.attribute("url.full", url.toString());
        span.attribute("url.path", url.pathname);
        const query = url.search.slice(1);
        if (query !== "") {
          span.attribute("url.query", url.search.slice(1));
        }
        span.attribute("url.scheme", url.protocol.slice(0, -1));
      }
      if (request.headers["user-agent"] !== undefined) {
        span.attribute("user_agent.original", request.headers["user-agent"]);
      }
      for (const name in requestHeaders) {
        span.attribute(`http.request.header.${name}`, String(requestHeaders[name]));
      }
      if (request.remoteAddress !== undefined) {
        span.attribute("client.address", request.remoteAddress);
      }
      const response = exitResponse(exit);
      span.attribute("http.response.status_code", response.status);
      const responseHeaders = Headers.redact(response.headers, redactedHeaderNames);
      for (const name in responseHeaders) {
        span.attribute(`http.response.header.${name}`, String(responseHeaders[name]));
      }
      span.end(endTime, exit);
    }, 0);
    return undefined;
  }, true);
}));
/**
 * @since 4.0.0
 * @category Proxying
 */
export const xForwardedHeaders = /*#__PURE__*/make(httpApp => Effect.updateService(httpApp, HttpServerRequest, request => request.headers["x-forwarded-host"] ? request.modify({
  headers: Headers.set(request.headers, "host", request.headers["x-forwarded-host"]),
  remoteAddress: request.headers["x-forwarded-for"]?.split(",")[0].trim()
}) : request));
/**
 * @since 4.0.0
 * @category Search params
 */
export const searchParamsParser = httpApp => Effect.withFiber(fiber => {
  const services = fiber.services;
  const request = ServiceMap.getUnsafe(services, HttpServerRequest);
  const params = Request.searchParamsFromURL(new URL(request.originalUrl));
  return Effect.provideService(httpApp, Request.ParsedSearchParams, params);
});
/**
 * @since 4.0.0
 * @category CORS
 */
export const cors = options => {
  const opts = {
    allowedOrigins: options?.allowedOrigins ?? [],
    allowedMethods: options?.allowedMethods ?? ["GET", "HEAD", "PUT", "PATCH", "POST", "DELETE"],
    allowedHeaders: options?.allowedHeaders ?? [],
    exposedHeaders: options?.exposedHeaders ?? [],
    credentials: options?.credentials ?? false,
    maxAge: options?.maxAge
  };
  const isAllowedOrigin = typeof opts.allowedOrigins === "function" ? opts.allowedOrigins : origin => opts.allowedOrigins.includes(origin);
  const allowOrigin = typeof opts.allowedOrigins === "function" || opts.allowedOrigins.length > 1 ? originHeader => {
    if (!isAllowedOrigin(originHeader)) return undefined;
    return {
      "access-control-allow-origin": originHeader,
      vary: "Origin"
    };
  } : opts.allowedOrigins.length === 0 ? constant({
    "access-control-allow-origin": "*"
  }) : constant({
    "access-control-allow-origin": opts.allowedOrigins[0],
    vary: "Origin"
  });
  const allowMethods = opts.allowedMethods.length > 0 ? {
    "access-control-allow-methods": opts.allowedMethods.join(", ")
  } : undefined;
  const allowCredentials = opts.credentials ? {
    "access-control-allow-credentials": "true"
  } : undefined;
  const allowHeaders = accessControlRequestHeaders => {
    if (opts.allowedHeaders.length === 0 && accessControlRequestHeaders) {
      return {
        vary: "Access-Control-Request-Headers",
        "access-control-allow-headers": accessControlRequestHeaders
      };
    }
    if (opts.allowedHeaders) {
      return {
        "access-control-allow-headers": opts.allowedHeaders.join(",")
      };
    }
    return undefined;
  };
  const exposeHeaders = opts.exposedHeaders.length > 0 ? {
    "access-control-expose-headers": opts.exposedHeaders.join(",")
  } : undefined;
  const maxAge = opts.maxAge ? {
    "access-control-max-age": opts.maxAge.toString()
  } : undefined;
  const headersFromRequest = request => {
    const origin = request.headers["origin"];
    return Headers.fromRecordUnsafe({
      ...allowOrigin(origin),
      ...allowCredentials,
      ...exposeHeaders
    });
  };
  const headersFromRequestOptions = request => {
    const origin = request.headers["origin"];
    const accessControlRequestHeaders = request.headers["access-control-request-headers"];
    return Headers.fromRecordUnsafe({
      ...allowOrigin(origin),
      ...allowCredentials,
      ...exposeHeaders,
      ...allowMethods,
      ...allowHeaders(accessControlRequestHeaders),
      ...maxAge
    });
  };
  const preResponseHandler = (request, response) => Effect.succeed(Response.setHeaders(response, headersFromRequest(request)));
  return httpApp => Effect.withFiber(fiber => {
    const request = ServiceMap.getUnsafe(fiber.services, HttpServerRequest);
    if (request.method === "OPTIONS") {
      return Effect.succeed(Response.empty({
        status: 204,
        headers: headersFromRequestOptions(request)
      }));
    }
    appendPreResponseHandlerUnsafe(request, preResponseHandler);
    return httpApp;
  });
};
