// effect/dist/unstable/cluster/ClusterSchema.js
/**
 * @since 4.0.0
 */
import { constFalse, constTrue } from "../../Function.js";
import * as ServiceMap from "../../ServiceMap.js";
/**
 * @since 4.0.0
 * @category Annotations
 */
export const Persisted = /*#__PURE__*/ServiceMap.Reference("effect/cluster/ClusterSchema/Persisted", {
  defaultValue: constFalse
});
/**
 * @since 4.0.0
 * @category Annotations
 */
export const Uninterruptible = /*#__PURE__*/ServiceMap.Reference("effect/cluster/ClusterSchema/Uninterruptible", {
  defaultValue: constFalse
});
/**
 * @since 4.0.0
 * @category Annotations
 */
export const isUninterruptibleForServer = context => {
  const value = ServiceMap.get(context, Uninterruptible);
  return value === true || value === "server";
};
/**
 * @since 4.0.0
 * @category Annotations
 */
export const isUninterruptibleForClient = context => {
  const value = ServiceMap.get(context, Uninterruptible);
  return value === true || value === "client";
};
/**
 * @since 4.0.0
 * @category Annotations
 */
export const ShardGroup = /*#__PURE__*/ServiceMap.Reference("effect/cluster/ClusterSchema/ShardGroup", {
  defaultValue: () => _ => "default"
});
/**
 * @since 4.0.0
 * @category Annotations
 */
export const ClientTracingEnabled = /*#__PURE__*/ServiceMap.Reference("effect/cluster/ClusterSchema/ClientTracingEnabled", {
  defaultValue: constTrue
});
