// effect/dist/unstable/cluster/ShardId.js
/**
 * @since 4.0.0
 */
import * as Equal from "../../Equal.js";
import * as Hash from "../../Hash.js";
import * as PrimaryKey from "../../PrimaryKey.js";
import * as S from "../../Schema.js";
const TypeId = "~effect/cluster/ShardId";
const constDisableValidation = {
  disableValidation: true
};
/**
 * @since 4.0.0
 * @category Constructors
 */
export const make = (group, id) => {
  const key = `${group}:${id}`;
  let shardId = shardIdCache.get(key);
  if (!shardId) {
    shardId = new ShardId({
      group,
      id
    }, constDisableValidation);
    shardIdCache.set(key, shardId);
  }
  return shardId;
};
const shardIdCache = /*#__PURE__*/new Map();
/**
 * @since 4.0.0
 * @category Models
 */
export class ShardId extends /*#__PURE__*/S.Class(TypeId)({
  group: S.String,
  id: S.Int
}) {
  /**
   * @since 4.0.0
   */
  static make = make;
  /**
   * @since 4.0.0
   */
  [TypeId] = TypeId;
  /**
   * @since 4.0.0
   */
  [Equal.symbol](that) {
    return this.group === that.group && this.id === that.id;
  }
  /**
   * @since 4.0.0
   */
  [Hash.symbol]() {
    return Hash.string(this.toString());
  }
  /**
   * @since 4.0.0
   */
  [PrimaryKey.symbol]() {
    return this.toString();
  }
  /**
   * @since 4.0.0
   */
  toString() {
    return `${this.group}:${this.id}`;
  }
  /**
   * @since 4.0.0
   */
  static toString(shardId) {
    return `${shardId.group}:${shardId.id}`;
  }
  /**
   * @since 4.0.0
   */
  static fromStringEncoded(s) {
    const index = s.lastIndexOf(":");
    if (index === -1) {
      throw new Error(`Invalid ShardId format`);
    }
    const group = s.substring(0, index);
    const id = Number(s.substring(index + 1));
    if (isNaN(id)) {
      throw new Error(`ShardId id must be a number`);
    }
    return {
      group,
      id
    };
  }
  /**
   * @since 4.0.0
   */
  static fromString(s) {
    const encoded = ShardId.fromStringEncoded(s);
    return make(encoded.group, encoded.id);
  }
}
