// effect/dist/ManagedRuntime.js
/**
 * @since 2.0.0
 */
import * as Effect from "./Effect.js";
import * as Exit from "./Exit.js";
import * as Fiber from "./Fiber.js";
import * as Layer from "./Layer.js";
import { hasProperty } from "./Predicate.js";
import * as Scope from "./Scope.js";
const TypeId = "~effect/ManagedRuntime";
/**
 * Checks if the provided argument is a `ManagedRuntime`.
 *
 * @since 3.9.0
 * @category guards
 */
export const isManagedRuntime = input => hasProperty(input, TypeId);
/**
 * Convert a Layer into an ManagedRuntime, that can be used to run Effect's using
 * your services.
 *
 * @since 2.0.0
 * @category runtime class
 * @example
 * ```ts
 * import { Console, Effect, Layer, ManagedRuntime, ServiceMap } from "effect"
 *
 * class Notifications extends ServiceMap.Service<Notifications, {
 *   readonly notify: (message: string) => Effect.Effect<void>
 * }>()("Notifications") {
 *   static readonly layer = Layer.succeed(this)({
 *     notify: Effect.fn("Notifications.notify")((message) => Console.log(message))
 *   })
 * }
 *
 * async function main() {
 *   const runtime = ManagedRuntime.make(Notifications.layer)
 *   await runtime.runPromise(Effect.flatMap(
 *     Notifications.asEffect(),
 *     (_) => _.notify("Hello, world!")
 *   ))
 *   await runtime.dispose()
 * }
 *
 * main()
 * ```
 */
export const make = (layer, options) => {
  const memoMap = options?.memoMap ?? Layer.makeMemoMapUnsafe();
  const scope = Scope.makeUnsafe();
  let buildFiber;
  const servicesEffect = Effect.withFiber(fiber => {
    if (!buildFiber) {
      buildFiber = Fiber.runIn(Effect.runFork(Effect.tap(Layer.buildWithMemoMap(layer, memoMap, scope), services => Effect.sync(() => {
        self.cachedServices = services;
      })), {
        scheduler: fiber.currentScheduler
      }), scope);
    }
    return Effect.flatten(Fiber.await(buildFiber));
  });
  const self = {
    [TypeId]: TypeId,
    memoMap,
    scope,
    servicesEffect,
    cachedServices: undefined,
    services() {
      return self.cachedServices === undefined ? Effect.runPromise(self.servicesEffect) : Promise.resolve(self.cachedServices);
    },
    dispose() {
      return Effect.runPromise(self.disposeEffect);
    },
    disposeEffect: Effect.suspend(() => {
      ;
      self.servicesEffect = Effect.die("ManagedRuntime disposed");
      self.cachedServices = undefined;
      return Scope.close(self.scope, Exit.void);
    }),
    runFork(effect, options) {
      return self.cachedServices === undefined ? Effect.runFork(provide(self, effect), options) : Effect.runForkWith(self.cachedServices)(effect, options);
    },
    runCallback(effect, options) {
      return self.cachedServices === undefined ? Effect.runCallback(provide(self, effect), options) : Effect.runCallbackWith(self.cachedServices)(effect, options);
    },
    runSyncExit(effect) {
      return self.cachedServices === undefined ? Effect.runSyncExit(provide(self, effect)) : Effect.runSyncExitWith(self.cachedServices)(effect);
    },
    runSync(effect) {
      return self.cachedServices === undefined ? Effect.runSync(provide(self, effect)) : Effect.runSyncWith(self.cachedServices)(effect);
    },
    runPromiseExit(effect, options) {
      return self.cachedServices === undefined ? Effect.runPromiseExit(provide(self, effect), options) : Effect.runPromiseExitWith(self.cachedServices)(effect, options);
    },
    runPromise(effect, options) {
      return self.cachedServices === undefined ? Effect.runPromise(provide(self, effect), options) : Effect.runPromiseWith(self.cachedServices)(effect, options);
    }
  };
  return self;
};
function provide(managed, effect) {
  return Effect.flatMap(managed.servicesEffect, services => Effect.provideServices(effect, services));
}
