// effect/dist/Redactable.js
/**
 * @since 4.0.0
 */
import { pipeArguments } from "./Pipeable.js";
import { hasProperty } from "./Predicate.js";
/**
 * Symbol used to identify objects that implement redaction capabilities.
 *
 * @since 4.0.0
 * @category symbol
 */
export const symbolRedactable = /*#__PURE__*/Symbol.for("~effect/Inspectable/redactable");
/**
 * Checks if a value implements the `Redactable` interface.
 *
 * This function determines whether a given value has redaction capabilities,
 * meaning it can provide alternative representations based on context.
 *
 * @param u - The value to check
 *
 * @example
 * ```ts
 * import { Redactable } from "effect"
 *
 * class RedactableSecret {
 *   [Redactable.symbolRedactable]() {
 *     return "[REDACTED]"
 *   }
 * }
 *
 * const secret = new RedactableSecret()
 * const normal = { value: 42 }
 *
 * console.log(Redactable.isRedactable(secret)) // true
 * console.log(Redactable.isRedactable(normal)) // false
 * console.log(Redactable.isRedactable("string")) // false
 * ```
 *
 * @since 4.0.0
 * @category redactable
 */
export const isRedactable = u => hasProperty(u, symbolRedactable);
/**
 * Applies redaction to a value if it implements the Redactable interface.
 *
 * This function checks if the value is redactable and applies the redaction
 * transformation if a current fiber context is available. Otherwise, it returns
 * the value unchanged.
 *
 * @param u - The value to potentially redact
 *
 * @example
 * ```ts
 * import { Redactable } from "effect"
 *
 * class CreditCard {
 *   constructor(private number: string) {}
 *
 *   [Redactable.symbolRedactable]() {
 *     return {
 *       number: this.number.slice(0, 4) + "****"
 *     }
 *   }
 * }
 *
 * const card = new CreditCard("1234567890123456")
 * console.log(Redactable.redact(card)) // { number: "1234****" }
 *
 * // Non-redactable values are returned unchanged
 * console.log(Redactable.redact("normal string")) // "normal string"
 * console.log(Redactable.redact({ id: 123 })) // { id: 123 }
 * ```
 *
 * @since 4.0.0
 */
export function redact(u) {
  if (isRedactable(u)) return getRedacted(u);
  return u;
}
/**
 * @since 4.0.0
 */
export function getRedacted(redactable) {
  return redactable[symbolRedactable](globalThis[currentFiberTypeId]?.services ?? emptyServiceMap);
}
/** @internal */
export const currentFiberTypeId = "~effect/Fiber/currentFiber";
const emptyServiceMap = {
  "~effect/ServiceMap": {},
  mapUnsafe: /*#__PURE__*/new Map(),
  pipe() {
    return pipeArguments(this, arguments);
  }
};
