// effect/dist/Schema.js
/**
 * @since 4.0.0
 */
import * as Arr from "./Array.js";
import * as BigDecimal_ from "./BigDecimal.js";
import * as Cause_ from "./Cause.js";
import * as Chunk_ from "./Chunk.js";
import * as Data from "./Data.js";
import * as DateTime from "./DateTime.js";
import * as Duration_ from "./Duration.js";
import * as Effect from "./Effect.js";
import * as Encoding from "./Encoding.js";
import * as Equal from "./Equal.js";
import * as Equivalence from "./Equivalence.js";
import * as Exit_ from "./Exit.js";
import { format, formatDate, formatPropertyKey } from "./Formatter.js";
import { identity } from "./Function.js";
import * as HashMap_ from "./HashMap.js";
import * as HashSet_ from "./HashSet.js";
import * as core from "./internal/core.js";
import * as InternalAnnotations from "./internal/schema/annotations.js";
import * as InternalArbitrary from "./internal/schema/arbitrary.js";
import * as InternalEquivalence from "./internal/schema/equivalence.js";
import * as InternalStandard from "./internal/schema/representation.js";
import * as InternalSchema from "./internal/schema/schema.js";
import * as InternalToCodec from "./internal/schema/to-codec.js";
import * as JsonPatch from "./JsonPatch.js";
import * as JsonSchema from "./JsonSchema.js";
import { remainder } from "./Number.js";
import * as Optic_ from "./Optic.js";
import * as Option_ from "./Option.js";
import * as Order from "./Order.js";
import * as Pipeable from "./Pipeable.js";
import * as Predicate from "./Predicate.js";
import * as Record_ from "./Record.js";
import * as Redacted_ from "./Redacted.js";
import * as Result_ from "./Result.js";
import * as Scheduler from "./Scheduler.js";
import * as AST from "./SchemaAST.js";
import * as Getter from "./SchemaGetter.js";
import * as Issue from "./SchemaIssue.js";
import * as Parser from "./SchemaParser.js";
import * as Transformation from "./SchemaTransformation.js";
import * as Struct_ from "./Struct.js";
import * as FastCheck from "./testing/FastCheck.js";
const TypeId = InternalSchema.TypeId;
/**
 * An API for creating schemas for parametric types.
 *
 * @see {@link declare} for creating schemas for non parametric types.
 *
 * @category Constructors
 * @since 4.0.0
 */
export function declareConstructor() {
  return (typeParameters, run, annotations) => {
    return make(new AST.Declaration(typeParameters.map(AST.getAST), typeParameters => run(typeParameters.map(ast => make(ast))), annotations));
  };
}
/**
 * An API for creating schemas for non parametric types.
 *
 * @see {@link declareConstructor} for creating schemas for parametric types.
 *
 * @since 4.0.0
 */
export function declare(is, annotations) {
  return declareConstructor()([], () => (input, ast) => is(input) ? Effect.succeed(input) : Effect.fail(new Issue.InvalidType(ast, Option_.some(input))), annotations);
}
/**
 * Reveals the complete Bottom interface type of a schema, exposing all 14 type
 * parameters.
 *
 * @since 4.0.0
 */
export function revealBottom(bottom) {
  return bottom;
}
/**
 * Adds metadata annotations to a schema without changing its runtime behavior.
 * Annotations are used to provide additional context for documentation,
 * JSON schema generation, error formatting, and other tooling.
 *
 * @category Annotations
 * @since 4.0.0
 */
export function annotate(annotations) {
  return self => {
    return self.annotate(annotations);
  };
}
/**
 * Adds key-specific annotations to a schema field. This is useful for providing
 * custom error messages and documentation for individual fields within
 * structures.
 *
 * @category Annotations
 * @since 4.0.0
 */
export function annotateKey(annotations) {
  return self => {
    return self.rebuild(AST.annotateKey(self.ast, annotations));
  };
}
/**
 * @since 4.0.0
 */
export function revealCodec(codec) {
  return codec;
}
const SchemaErrorTypeId = "~effect/Schema/SchemaError";
/**
 * A `SchemaError` is returned when schema decoding or encoding fails.
 *
 * This error extends `Data.TaggedError` and contains detailed information about
 * what went wrong during schema processing. The error includes an `issue` field
 * that provides comprehensive details about the validation failure, including
 * the path to the problematic data, expected types, and actual values.
 *
 * @since 4.0.0
 */
export class SchemaError {
  [SchemaErrorTypeId] = SchemaErrorTypeId;
  _tag = "SchemaError";
  name = "SchemaError";
  issue;
  constructor(issue) {
    this.issue = issue;
  }
  get message() {
    return this.issue.toString();
  }
  toString() {
    return `SchemaError(${this.message})`;
  }
}
/**
 * @since 4.0.0
 */
export function isSchemaError(u) {
  return Predicate.hasProperty(u, SchemaErrorTypeId);
}
function makeStandardResult(exit) {
  return Exit_.isSuccess(exit) ? exit.value : {
    issues: [{
      message: Cause_.pretty(exit.cause)
    }]
  };
}
/**
 * Returns a "Standard Schema" object conforming to the [Standard Schema
 * v1](https://standardschema.dev/) specification.
 *
 * This function creates a schema whose `validate` method attempts to decode and
 * validate the provided input synchronously. If the underlying `Schema`
 * includes any asynchronous components (e.g., asynchronous message resolutions
 * or checks), then validation will necessarily return a `Promise` instead.
 *
 * **Example** (Creating a standard schema from a regular schema)
 *
 * ```ts
 * import { Schema } from "effect"
 *
 * // Define custom hook functions for error formatting
 * const leafHook = (issue: any) => {
 *   switch (issue._tag) {
 *     case "InvalidType":
 *       return "Expected different type"
 *     case "InvalidValue":
 *       return "Invalid value provided"
 *     case "MissingKey":
 *       return "Required property missing"
 *     case "UnexpectedKey":
 *       return "Unexpected property found"
 *     case "Forbidden":
 *       return "Operation not allowed"
 *     case "OneOf":
 *       return "Multiple valid options available"
 *     default:
 *       return "Validation error"
 *   }
 * }
 *
 * // Create a standard schema from a regular schema
 * const PersonSchema = Schema.Struct({
 *   name: Schema.NonEmptyString,
 *   age: Schema.Number.check(Schema.isBetween({ minimum: 0, maximum: 150 }))
 * })
 *
 * const standardSchema = Schema.toStandardSchemaV1(PersonSchema, {
 *   leafHook
 * })
 *
 * // The standard schema can be used with any Standard Schema v1 compatible library
 * const validResult = standardSchema["~standard"].validate({
 *   name: "Alice",
 *   age: 30
 * })
 * console.log(validResult) // { value: { name: "Alice", age: 30 } }
 *
 * const invalidResult = standardSchema["~standard"].validate({
 *   name: "",
 *   age: 200
 * })
 * console.log(invalidResult) // { issues: [{ path: ["name"], message: "..." }, { path: ["age"], message: "..." }] }
 * ```
 *
 * @category Standard Schema
 * @since 4.0.0
 */
export function toStandardSchemaV1(self, options) {
  const decodeUnknownEffect = Parser.decodeUnknownEffect(self);
  const parseOptions = {
    errors: "all",
    ...options?.parseOptions
  };
  const formatter = Issue.makeFormatterStandardSchemaV1(options);
  const validate = value => {
    const scheduler = new Scheduler.MixedScheduler();
    const fiber = Effect.runFork(Effect.match(decodeUnknownEffect(value, parseOptions), {
      onFailure: formatter,
      onSuccess: value => ({
        value
      })
    }), {
      scheduler
    });
    scheduler.flush();
    const exit = fiber.pollUnsafe();
    if (exit) {
      return makeStandardResult(exit);
    }
    return new Promise(resolve => {
      fiber.addObserver(exit => {
        resolve(makeStandardResult(exit));
      });
    });
  };
  if ("~standard" in self) {
    const out = self;
    if ("validate" in out["~standard"]) return out;
    Object.assign(out["~standard"], {
      validate
    });
    return out;
  } else {
    return Object.assign(self, {
      "~standard": {
        version: 1,
        vendor: "effect",
        validate
      }
    });
  }
}
function toBaseStandardJSONSchemaV1(self, target) {
  const doc2020_12 = toJsonSchemaDocument(self);
  if (target === "draft-2020-12") {
    const schema = doc2020_12.schema;
    if (Object.keys(doc2020_12.definitions).length > 0) {
      schema.$defs = doc2020_12.definitions;
    }
    return schema;
  } else if (target === "draft-07") {
    const doc07 = JsonSchema.toDocumentDraft07(doc2020_12);
    const schema = doc07.schema;
    if (Object.keys(doc07.definitions).length > 0) {
      schema.definitions = doc07.definitions;
    }
    return schema;
  }
  throw new globalThis.Error(`Unsupported target: ${target}`);
}
/**
 * Experimental support for converting a schema to a Standard JSON Schema V1.
 *
 * https://github.com/standard-schema/standard-schema/pull/134
 *
 * @category Standard Schema
 * @since 4.0.0
 * @experimental
 */
export function toStandardJSONSchemaV1(self) {
  const jsonSchema = {
    input(options) {
      return toBaseStandardJSONSchemaV1(self, options.target);
    },
    output(options) {
      return toBaseStandardJSONSchemaV1(toType(self), options.target);
    }
  };
  if ("~standard" in self) {
    const out = self;
    if ("jsonSchema" in out["~standard"]) return out;
    Object.assign(out["~standard"], {
      jsonSchema
    });
    return out;
  } else {
    return Object.assign(self, {
      "~standard": {
        version: 1,
        vendor: "effect",
        jsonSchema
      }
    });
  }
}
/**
 * Creates a type guard function that checks if a value conforms to a given
 * schema.
 *
 * This function returns a predicate that performs a type-safe check, narrowing
 * the type of the input value if the check passes. It's particularly useful for
 * runtime type validation and TypeScript type narrowing.
 *
 * **Example** (Basic Type Guard)
 *
 * ```ts
 * import { Schema } from "effect"
 *
 * const isString = Schema.is(Schema.String)
 *
 * console.log(isString("hello")) // true
 * console.log(isString(42)) // false
 *
 * // Type narrowing in action
 * const value: unknown = "hello"
 * if (isString(value)) {
 *   // value is now typed as string
 *   console.log(value.toUpperCase()) // "HELLO"
 * }
 * ```
 *
 * @category Asserting
 * @since 4.0.0
 */
export const is = Parser.is;
/**
 * Creates an assertion function that throws an error if the input doesn't match
 * the schema.
 *
 * This function is useful for runtime type checking with TypeScript's `asserts`
 * type guard. It narrows the type of the input if the assertion succeeds, or
 * throws an error if it fails.
 *
 * **Example** (Basic Usage)
 *
 * ```ts
 * import { Schema } from "effect"
 *
 * const assertString: (u: unknown) => asserts u is string = Schema.asserts(
 *   Schema.String
 * )
 *
 * // This will pass silently (no return value)
 * try {
 *   assertString("hello")
 *   console.log("String assertion passed")
 * } catch (error) {
 *   console.log("String assertion failed")
 * }
 *
 * // This will throw an error
 * try {
 *   assertString(123)
 * } catch (error) {
 *   console.log("Non-string assertion failed as expected")
 * }
 * ```
 *
 * @category Asserting
 * @since 4.0.0
 */
export const asserts = Parser.asserts;
/**
 * @category Decoding
 * @since 4.0.0
 */
export function decodeUnknownEffect(schema) {
  const parser = Parser.decodeUnknownEffect(schema);
  return (input, options) => {
    return Effect.mapErrorEager(parser(input, options), issue => new SchemaError(issue));
  };
}
/**
 * @category Decoding
 * @since 4.0.0
 */
export const decodeEffect = decodeUnknownEffect;
/**
 * @category Decoding
 * @since 4.0.0
 */
export function decodeUnknownExit(schema) {
  const parser = Parser.decodeUnknownExit(schema);
  return (input, options) => {
    return Exit_.mapError(parser(input, options), issue => new SchemaError(issue));
  };
}
/**
 * @category Decoding
 * @since 4.0.0
 */
export const decodeExit = decodeUnknownExit;
/**
 * @category Decoding
 * @since 4.0.0
 */
export const decodeUnknownOption = Parser.decodeUnknownOption;
/**
 * @category Decoding
 * @since 4.0.0
 */
export const decodeOption = Parser.decodeOption;
/**
 * @category Decoding
 * @since 4.0.0
 */
export const decodeUnknownPromise = Parser.decodeUnknownPromise;
/**
 * @category Decoding
 * @since 4.0.0
 */
export const decodePromise = Parser.decodePromise;
/**
 * @category Decoding
 * @since 4.0.0
 */
export const decodeUnknownSync = Parser.decodeUnknownSync;
/**
 * @category Decoding
 * @since 4.0.0
 */
export const decodeSync = Parser.decodeSync;
/**
 * @category Encoding
 * @since 4.0.0
 */
export function encodeUnknownEffect(schema) {
  const parser = Parser.encodeUnknownEffect(schema);
  return (input, options) => {
    return Effect.mapErrorEager(parser(input, options), issue => new SchemaError(issue));
  };
}
/**
 * @category Encoding
 * @since 4.0.0
 */
export const encodeEffect = encodeUnknownEffect;
/**
 * @category Encoding
 * @since 4.0.0
 */
export function encodeUnknownExit(schema) {
  const parser = Parser.encodeUnknownExit(schema);
  return (input, options) => {
    return Exit_.mapError(parser(input, options), issue => new SchemaError(issue));
  };
}
/**
 * @category Encoding
 * @since 4.0.0
 */
export const encodeExit = encodeUnknownExit;
/**
 * @category Encoding
 * @since 4.0.0
 */
export const encodeUnknownOption = Parser.encodeUnknownOption;
/**
 * @category Encoding
 * @since 4.0.0
 */
export const encodeOption = Parser.encodeOption;
/**
 * @category Encoding
 * @since 4.0.0
 */
export const encodeUnknownPromise = Parser.encodeUnknownPromise;
/**
 * @category Encoding
 * @since 4.0.0
 */
export const encodePromise = Parser.encodePromise;
/**
 * @category Encoding
 * @since 4.0.0
 */
export const encodeUnknownSync = Parser.encodeUnknownSync;
/**
 * @category Encoding
 * @since 4.0.0
 */
export const encodeSync = Parser.encodeSync;
/**
 * Creates a schema from an AST (Abstract Syntax Tree) node.
 *
 * This is the fundamental constructor for all schemas in the Effect Schema
 * library. It takes an AST node and wraps it in a fully-typed schema that
 * preserves all type information and provides the complete schema API.
 *
 * The `make` function is used internally to create all primitive schemas like
 * `String`, `Number`, `Boolean`, etc., as well as more complex schemas. It's
 * the bridge between the untyped AST representation and the strongly-typed
 * schema.
 *
 * @category Constructors
 * @since 4.0.0
 */
export const make = InternalSchema.make;
/**
 * Tests if a value is a `Schema`.
 *
 * @category Guards
 * @since 4.0.0
 */
export function isSchema(u) {
  return Predicate.hasProperty(u, TypeId) && u[TypeId] === TypeId;
}
/**
 * Creates an exact optional key schema for struct fields. Unlike `optional`,
 * this creates exact optional properties (not `| undefined`) that can be
 * completely omitted from the object.
 *
 * **Example** (Creating a struct with optional key)
 *
 * ```ts
 * import { Schema } from "effect"
 *
 * const schema = Schema.Struct({
 *   name: Schema.String,
 *   age: Schema.optionalKey(Schema.Number)
 * })
 *
 * // Type: { readonly name: string; readonly age?: number }
 * type Person = typeof schema["Type"]
 * ```
 *
 * @since 4.0.0
 */
export const optionalKey = /*#__PURE__*/Struct_.lambda(schema => make(AST.optionalKey(schema.ast), {
  schema
}));
/**
 * @since 4.0.0
 */
export const requiredKey = /*#__PURE__*/Struct_.lambda(self => self.schema);
/**
 * Creates an optional schema field that allows both the specified type and
 * `undefined`.
 *
 * This is equivalent to `optionalKey(UndefinedOr(schema))`, creating a field
 * that:
 * - Can be omitted from the object entirely
 * - Can be explicitly set to `undefined`
 * - Can contain the specified schema type
 *
 * **Example** (Creating a struct with optional)
 *
 * ```ts
 * import { Schema } from "effect"
 *
 * const schema = Schema.Struct({
 *   name: Schema.String,
 *   age: Schema.optionalKey(Schema.Number)
 * })
 *
 * // Type: { readonly name: string; readonly age?: number | undefined }
 * type Person = typeof schema["Type"]
 * ```
 *
 * @since 4.0.0
 */
export const optional = /*#__PURE__*/Struct_.lambda(self => optionalKey(UndefinedOr(self)));
/**
 * @since 4.0.0
 */
export const required = /*#__PURE__*/Struct_.lambda(self => self.schema.members[0]);
/**
 * @since 4.0.0
 */
export const mutableKey = /*#__PURE__*/Struct_.lambda(schema => make(AST.mutableKey(schema.ast), {
  schema
}));
/**
 * @since 4.0.0
 */
export const readonlyKey = /*#__PURE__*/Struct_.lambda(self => self.schema);
/**
 * @since 4.0.0
 */
export const toType = /*#__PURE__*/Struct_.lambda(schema => make(AST.toType(schema.ast), {
  schema
}));
/**
 * @since 4.0.0
 */
export const toEncoded = /*#__PURE__*/Struct_.lambda(schema => make(AST.toEncoded(schema.ast), {
  schema
}));
const FlipTypeId = "~effect/Schema/flip";
function isFlip$(schema) {
  return Predicate.hasProperty(schema, FlipTypeId) && schema[FlipTypeId] === FlipTypeId;
}
export function flip(schema) {
  if (isFlip$(schema)) {
    return schema.schema.rebuild(AST.flip(schema.ast));
  }
  return make(AST.flip(schema.ast), {
    [FlipTypeId]: FlipTypeId,
    schema
  });
}
/**
 * @see {@link Literals} for a schema that represents a union of literals.
 * @see {@link tag} for a schema that represents a literal value that can be
 * used as a discriminator field in tagged unions and has a constructor default.
 * @since 4.0.0
 */
export function Literal(literal) {
  const out = make(new AST.Literal(literal), {
    literal,
    transform(to) {
      return out.pipe(decodeTo(Literal(to), {
        decode: Getter.transform(() => to),
        encode: Getter.transform(() => literal)
      }));
    }
  });
  return out;
}
function templateLiteralFromParts(parts) {
  return new AST.TemplateLiteral(parts.map(part => isSchema(part) ? part.ast : new AST.Literal(part)));
}
/**
 * @since 4.0.0
 */
export function TemplateLiteral(parts) {
  return make(templateLiteralFromParts(parts), {
    parts
  });
}
/**
 * @since 4.0.0
 */
export function TemplateLiteralParser(parts) {
  return make(templateLiteralFromParts(parts).asTemplateLiteralParser(), {
    parts: [...parts]
  });
}
/**
 * @since 4.0.0
 */
export function Enum(enums) {
  return make(new AST.Enum(Object.keys(enums).filter(key => typeof enums[enums[key]] !== "number").map(key => [key, enums[key]])), {
    enums
  });
}
/**
 * @since 4.0.0
 */
export const Never = /*#__PURE__*/make(AST.never);
/**
 * @since 4.0.0
 */
export const Any = /*#__PURE__*/make(AST.any);
/**
 * @since 4.0.0
 */
export const Unknown = /*#__PURE__*/make(AST.unknown);
/**
 * @since 4.0.0
 */
export const Null = /*#__PURE__*/make(AST.null);
/**
 * @since 4.0.0
 */
export const Undefined = /*#__PURE__*/make(AST.undefined);
/**
 * A schema for all strings.
 *
 * @since 4.0.0
 */
export const String = /*#__PURE__*/make(AST.string);
/**
 * A schema for all numbers, including `NaN`, `Infinity`, and `-Infinity`.
 *
 * **Default Json Serializer**
 *
 * - If the number is finite, it is serialized as a number.
 * - Otherwise, it is serialized as a string ("NaN", "Infinity", or "-Infinity").
 *
 * @since 4.0.0
 */
export const Number = /*#__PURE__*/make(AST.number);
/**
 * A schema for all booleans.
 *
 * @category Boolean
 * @since 4.0.0
 */
export const Boolean = /*#__PURE__*/make(AST.boolean);
/**
 * A schema for all symbols.
 *
 * @since 4.0.0
 */
export const Symbol = /*#__PURE__*/make(AST.symbol);
/**
 * A schema for all bigints.
 *
 * @since 4.0.0
 */
export const BigInt = /*#__PURE__*/make(AST.bigInt);
/**
 * A schema for the `void` type.
 *
 * @since 4.0.0
 */
export const Void = /*#__PURE__*/make(AST.void);
/**
 * A schema for the `object` type.
 *
 * @since 4.0.0
 */
export const ObjectKeyword = /*#__PURE__*/make(AST.objectKeyword);
/**
 * A schema for unique symbols.
 *
 * **Example**
 *
 * ```ts
 * import { Schema } from "effect"
 *
 * const a = Symbol.for("a")
 * const schema = Schema.UniqueSymbol(a)
 * ```
 * @since 4.0.0
 */
export function UniqueSymbol(symbol) {
  return make(new AST.UniqueSymbol(symbol));
}
function makeStruct(ast, fields) {
  return make(ast, {
    fields,
    mapFields(f, options) {
      const fields = f(this.fields);
      return makeStruct(AST.struct(fields, options?.unsafePreserveChecks ? this.ast.checks : undefined), fields);
    }
  });
}
/**
 * @since 4.0.0
 */
export function Struct(fields) {
  return makeStruct(AST.struct(fields, undefined), fields);
}
/**
 * A shortcut for `MyStruct.mapFields(Struct.assign(fields))`. This is useful
 * when you want to add new fields to an existing struct or a union of structs.
 *
 * **Example** (Adding fields to a union of structs)
 *
 * ```ts
 * import { Schema, Tuple } from "effect"
 *
 * // Add a new field to all members of a union of structs
 * const schema = Schema.Union([
 *   Schema.Struct({ a: Schema.String }),
 *   Schema.Struct({ b: Schema.Number })
 * ]).mapMembers(Tuple.map(Schema.fieldsAssign({ c: Schema.Number })))
 * ```
 *
 * @since 4.0.0
 */
export function fieldsAssign(fields) {
  return Struct_.lambda(struct => struct.mapFields(Struct_.assign(fields)));
}
/**
 * @category Struct transformations
 * @since 4.0.0
 */
export function encodeKeys(mapping) {
  return function (self) {
    const fields = {};
    const reverseMapping = {};
    for (const k in self.fields) {
      const encoded = toEncoded(self.fields[k]);
      if (Object.hasOwn(mapping, k)) {
        fields[mapping[k]] = encoded;
        reverseMapping[mapping[k]] = k;
      } else {
        fields[k] = encoded;
      }
    }
    return Struct(fields).pipe(decodeTo(self, Transformation.transform({
      decode: Struct_.renameKeys(reverseMapping),
      encode: Struct_.renameKeys(mapping)
    })));
  };
}
/**
 * @since 4.0.0
 * @experimental
 */
export function extendTo(/** The new fields to add */
fields, /** A function per field to derive its value from the original input */
derive) {
  return self => {
    const f = Record_.map(self.fields, toType);
    const to = Struct({
      ...f,
      ...fields
    });
    return self.pipe(decodeTo(to, Transformation.transform({
      decode: input => {
        const out = {
          ...input
        };
        for (const k in fields) {
          const f = derive[k];
          const o = f(input);
          if (Option_.isSome(o)) {
            out[k] = o.value;
          }
        }
        return out;
      },
      encode: input => {
        const out = {
          ...input
        };
        for (const k in fields) {
          delete out[k];
        }
        return out;
      }
    })));
  };
}
/**
 * @since 4.0.0
 */
export function Record(key, value, options) {
  const keyValueCombiner = options?.keyValueCombiner?.decode || options?.keyValueCombiner?.encode ? new AST.KeyValueCombiner(options.keyValueCombiner.decode, options.keyValueCombiner.encode) : undefined;
  return make(AST.record(key.ast, value.ast, keyValueCombiner), {
    key,
    value
  });
}
/**
 * @since 4.0.0
 */
export function StructWithRest(schema, records) {
  return make(AST.structWithRest(schema.ast, records.map(AST.getAST)), {
    schema,
    records
  });
}
function makeTuple(ast, elements) {
  return make(ast, {
    elements,
    mapElements(f, options) {
      const elements = f(this.elements);
      return makeTuple(AST.tuple(elements, options?.unsafePreserveChecks ? this.ast.checks : undefined), elements);
    }
  });
}
/**
 * @category Constructors
 * @since 4.0.0
 */
export function Tuple(elements) {
  return makeTuple(AST.tuple(elements), elements);
}
/**
 * @category Constructors
 * @since 4.0.0
 */
export function TupleWithRest(schema, rest) {
  return make(AST.tupleWithRest(schema.ast, rest.map(AST.getAST)), {
    schema,
    rest
  });
}
/**
 * @category Constructors
 * @since 4.0.0
 */
export const Array = /*#__PURE__*/Struct_.lambda(schema => make(new AST.Arrays(false, [], [schema.ast]), {
  schema
}));
/**
 * @category Constructors
 * @since 4.0.0
 */
export const NonEmptyArray = /*#__PURE__*/Struct_.lambda(schema => make(new AST.Arrays(false, [schema.ast], [schema.ast]), {
  schema
}));
/**
 * Returns a new array schema that ensures all elements are unique.
 *
 * The equivalence used to determine uniqueness is the one provided by
 * `Schema.toEquivalence(item)`.
 *
 * @category Constructors
 * @since 4.0.0
 */
export function UniqueArray(item) {
  return Array(item).check(isUnique());
}
/**
 * Makes arrays or tuples mutable.
 *
 * @since 4.0.0
 */
export const mutable = /*#__PURE__*/Struct_.lambda(schema => {
  return make(new AST.Arrays(true, schema.ast.elements, schema.ast.rest), {
    schema
  });
});
function makeUnion(ast, members) {
  return make(ast, {
    members,
    mapMembers(f, options) {
      const members = f(this.members);
      return makeUnion(AST.union(members, this.ast.mode, options?.unsafePreserveChecks ? this.ast.checks : undefined), members);
    }
  });
}
/**
 * Creates a schema that represents a union of multiple schemas. Members are checked in order, and the first match is returned.
 *
 * Optionally, you can specify the `mode` to be `"anyOf"` or `"oneOf"`.
 *
 * - `"anyOf"` - The union matches if any member matches.
 * - `"oneOf"` - The union matches if exactly one member matches.
 *
 * @category Constructors
 * @since 4.0.0
 */
export function Union(members, options) {
  return makeUnion(AST.union(members, options?.mode ?? "anyOf", undefined), members);
}
/**
 * @see {@link Literal} for a schema that represents a single literal.
 * @category Constructors
 * @since 4.0.0
 */
export function Literals(literals) {
  const members = literals.map(Literal);
  return make(AST.union(members, "anyOf", undefined), {
    literals,
    members,
    mapMembers(f) {
      return Union(f(this.members));
    },
    pick(literals) {
      return Literals(literals);
    },
    transform(to) {
      return Union(members.map((member, index) => member.transform(to[index])));
    }
  });
}
/**
 * @category Constructors
 * @since 4.0.0
 */
export const NullOr = /*#__PURE__*/Struct_.lambda(self => Union([self, Null]));
/**
 * @category Constructors
 * @since 4.0.0
 */
export const UndefinedOr = /*#__PURE__*/Struct_.lambda(self => Union([self, Undefined]));
/**
 * @category Constructors
 * @since 4.0.0
 */
export const NullishOr = /*#__PURE__*/Struct_.lambda(self => Union([self, Null, Undefined]));
/**
 * Creates a suspended schema that defers evaluation until needed. This is
 * essential for creating recursive schemas where a schema references itself,
 * preventing infinite recursion during schema definition.
 *
 * @category Constructors
 * @since 4.0.0
 */
export function suspend(f) {
  return make(new AST.Suspend(() => f().ast));
}
/**
 * @category Filtering
 * @since 4.0.0
 */
export function check(...checks) {
  return self => self.check(...checks);
}
/**
 * @category Filtering
 * @since 4.0.0
 */
export function refine(refinement, annotations) {
  return schema => make(AST.appendChecks(schema.ast, [AST.makeFilterByGuard(refinement, annotations)]), {
    schema
  });
}
/**
 * Adds a brand to a schema.
 *
 * @category Branding
 * @since 4.0.0
 */
export function brand(identifier) {
  return schema => make(AST.brand(schema.ast, identifier), {
    schema,
    identifier
  });
}
/**
 * @category Constructors
 * @since 4.0.0
 */
export function fromBrand(identifier, ctor) {
  return self => {
    return (ctor.checks ? self.check(...ctor.checks) : self).pipe(brand(identifier));
  };
}
/**
 * @since 4.0.0
 */
export function middlewareDecoding(decode) {
  return schema => make(AST.middlewareDecoding(schema.ast, new Transformation.Middleware(decode, identity)), {
    schema
  });
}
/**
 * @since 4.0.0
 */
export function middlewareEncoding(encode) {
  return schema => make(AST.middlewareEncoding(schema.ast, new Transformation.Middleware(identity, encode)), {
    schema
  });
}
/**
 * @since 4.0.0
 */
export function catchDecoding(f) {
  return catchDecodingWithContext(f);
}
/**
 * @since 4.0.0
 */
export function catchDecodingWithContext(f) {
  return self => self.pipe(middlewareDecoding(Effect.catchEager(f)));
}
/**
 * @since 4.0.0
 */
export function catchEncoding(f) {
  return catchEncodingWithContext(f);
}
/**
 * @since 4.0.0
 */
export function catchEncodingWithContext(f) {
  return self => self.pipe(middlewareEncoding(Effect.catchEager(f)));
}
export function decodeTo(to, transformation) {
  return from => {
    return make(AST.decodeTo(from.ast, to.ast, transformation ? Transformation.make(transformation) : Transformation.passthrough()), {
      from,
      to
    });
  };
}
/**
 * Applies a transformation to a schema, creating a new schema with the same type but transformed encoding/decoding.
 *
 * This is a curried function: call it with a transformation object, then call the returned function with a schema.
 * The resulting schema has `Type = S["Type"]` and `Encoded = S["Encoded"]`, with the transformation applied during
 * encoding and decoding operations.
 *
 * **Key guarantees:**
 * - Resulting schema has `Type = S["Type"]` and `Encoded = S["Encoded"]`
 * - Uses `toType(self)` as the target schema internally (creates a schema where both Type and Encoded are `S["Type"]`)
 * - Combines decoding/encoding services from the source schema and transformation
 * - Transformation `decode` maps `S["Type"]` → `S["Type"]` (used during encoding)
 * - Transformation `encode` maps `S["Type"]` → `S["Type"]` (used during decoding)
 *
 * **AI note - Common mistakes:**
 * - **Currying**: This is curried - must use pipe: `schema.pipe(Schema.decode(transformation))`
 * - **Transformation direction**: `decode` and `encode` both operate on `S["Type"]` (same type, different values)
 * - **Service dependencies**: Resulting schema requires services from the source schema and transformation; use `Schema.provideService` if needed
 *
 * **Example** (Trimming string values during encoding/decoding)
 *
 * ```ts
 * import { Schema, SchemaGetter } from "effect"
 *
 * const Trimmed = Schema.String.pipe(
 *   Schema.decode({
 *     decode: SchemaGetter.transform((s) => s.trim()),
 *     encode: SchemaGetter.transform((s) => s.trim())
 *   })
 * )
 *
 * const result = Schema.decodeUnknownSync(Trimmed)("  hello  ")
 * // result: "hello"
 * ```
 *
 * @since 4.0.0
 */
export function decode(transformation) {
  return self => {
    return self.pipe(decodeTo(toType(self), transformation));
  };
}
export function encodeTo(to, transformation) {
  return from => {
    return transformation ? to.pipe(decodeTo(from, transformation)) : to.pipe(decodeTo(from));
  };
}
/**
 * @since 4.0.0
 */
export function encode(transformation) {
  return self => {
    return toEncoded(self).pipe(decodeTo(self, transformation));
  };
}
/**
 * @since 4.0.0
 */
export function withConstructorDefault(defaultValue) {
  return schema => {
    return make(AST.withConstructorDefault(schema.ast, defaultValue), {
      schema
    });
  };
}
/**
 * **Options**
 *
 * - `encodingStrategy`: The strategy to use when encoding.
 *   - `passthrough`: (default) Pass the default value through to the output.
 *   - `omit`: Omit the value from the output.
 *
 * @since 4.0.0
 */
export function withDecodingDefaultKey(defaultValue, options) {
  const encode = options?.encodingStrategy === "omit" ? Getter.omit() : Getter.passthrough();
  return self => {
    return optionalKey(toEncoded(self)).pipe(decodeTo(self, {
      decode: Getter.withDefault(defaultValue),
      encode
    }));
  };
}
/**
 * **Options**
 *
 * - `encodingStrategy`: The strategy to use when encoding.
 *   - `passthrough`: (default) Pass the default value through to the output.
 *   - `omit`: Omit the value from the output.
 *
 * @since 4.0.0
 */
export function withDecodingDefault(defaultValue, options) {
  const encode = options?.encodingStrategy === "omit" ? Getter.omit() : Getter.passthrough();
  return self => {
    return optional(toEncoded(self)).pipe(decodeTo(self, {
      decode: Getter.withDefault(defaultValue),
      encode
    }));
  };
}
/**
 * Creates a schema for a literal value and automatically provides itself as a
 * default.
 *
 * The `tag` function combines a literal schema with a constructor default,
 * making it perfect for discriminated unions and tagged data structures. The
 * tag value is automatically provided when the field is missing during
 * construction.
 *
 * @since 4.0.0
 */
export function tag(literal) {
  return Literal(literal).pipe(withConstructorDefault(() => Option_.some(literal)));
}
/**
 * Similar to `tag`, but provides itself as a default when decoding and omits
 * the value from the output when encoding.
 *
 * @since 4.0.0
 */
export function tagDefaultOmit(literal) {
  return tag(literal).pipe(withDecodingDefaultKey(() => literal, {
    encodingStrategy: "omit"
  }));
}
/**
 * A tagged struct is a struct that includes a `_tag` field. This field is used
 * to identify the specific variant of the object, which is especially useful
 * when working with union types.
 *
 * When using the `makeUnsafe` method, the `_tag` field is optional and will be
 * added automatically. However, when decoding or encoding, the `_tag` field
 * must be present in the input.
 *
 * **Example** (Tagged struct as a shorthand for a struct with a `_tag` field)
 *
 * ```ts
 * import { Schema } from "effect"
 *
 * // Defines a struct with a fixed `_tag` field
 * const tagged = Schema.TaggedStruct("A", {
 *   a: Schema.String
 * })
 *
 * // This is the same as writing:
 * const equivalent = Schema.Struct({
 *   _tag: Schema.tag("A"),
 *   a: Schema.String
 * })
 * ```
 *
 * **Example** (Accessing the literal value of the tag)
 *
 * ```ts
 * import { Schema } from "effect"
 *
 * const tagged = Schema.TaggedStruct("A", {
 *   a: Schema.String
 * })
 *
 * // literal: "A"
 * const literal = tagged.fields._tag.schema.literal
 * ```
 *
 * @category Constructors
 * @since 4.0.0
 */
export function TaggedStruct(value, fields) {
  return Struct({
    _tag: tag(value),
    ...fields
  });
}
/**
 * @since 4.0.0
 * @experimental
 */
export function toTaggedUnion(tag) {
  return self => {
    const cases = {};
    const guards = {};
    const isAnyOf = keys => value => keys.includes(value[tag]);
    walk(self);
    return Object.assign(self, {
      cases,
      isAnyOf,
      guards,
      match
    });
    function walk(schema) {
      const ast = schema.ast;
      if (AST.isUnion(ast) && "members" in schema && globalThis.Array.isArray(schema.members) && schema.members.every(isSchema)) {
        return schema.members.forEach(walk);
      }
      const sentinels = AST.collectSentinels(ast);
      if (sentinels.length > 0) {
        const literal = sentinels.find(s => s.key === tag)?.literal;
        if (Predicate.isPropertyKey(literal)) {
          cases[literal] = schema;
          guards[literal] = is(toType(schema));
          return;
        }
      }
      throw new globalThis.Error("No literal or unique symbol found");
    }
    function match() {
      if (arguments.length === 1) {
        const cases = arguments[0];
        return function (value) {
          return cases[value[tag]](value);
        };
      }
      const value = arguments[0];
      const cases = arguments[1];
      return cases[value[tag]](value);
    }
  };
}
/**
 * @category Constructors
 * @since 4.0.0
 */
export function TaggedUnion(casesByTag) {
  const cases = {};
  const members = [];
  for (const key of Object.keys(casesByTag)) {
    members.push(cases[key] = TaggedStruct(key, casesByTag[key]));
  }
  const union = Union(members);
  const {
    guards,
    isAnyOf,
    match
  } = toTaggedUnion("_tag")(union);
  return make(union.ast, {
    cases,
    isAnyOf,
    guards,
    match
  });
}
/**
 * @since 4.0.0
 */
export function Opaque() {
  return schema => {
    // oxlint-disable-next-line @typescript-eslint/no-extraneous-class
    class Opaque {}
    Object.setPrototypeOf(Opaque, schema);
    return Opaque;
  };
}
/**
 * Creates a schema that validates an instance of a specific class constructor.
 *
 * @category Constructors
 * @since 4.0.0
 */
export function instanceOf(constructor, annotations) {
  return declare(u => u instanceof constructor, annotations);
}
/**
 * @since 4.0.0
 * @experimental
 */
export function link() {
  return (encodeTo, transformation) => {
    return new AST.Link(encodeTo.ast, Transformation.make(transformation));
  };
}
// -----------------------------------------------------------------------------
// Checks
// -----------------------------------------------------------------------------
/**
 * @category Checks Constructors
 * @since 4.0.0
 */
export const makeFilter = AST.makeFilter;
/**
 * @category Checks Constructors
 * @since 4.0.0
 */
export function makeFilterGroup(checks, annotations = undefined) {
  return new AST.FilterGroup(checks, annotations);
}
const TRIMMED_PATTERN = "^\\S[\\s\\S]*\\S$|^\\S$|^$";
/**
 * Validates that a string has no leading or trailing whitespace.
 *
 * **JSON Schema**
 *
 * This check corresponds to a `pattern` constraint in JSON Schema that
 * matches strings without leading or trailing whitespace.
 *
 * **Arbitrary**
 *
 * When generating test data with fast-check, this applies a `patterns`
 * constraint to ensure generated strings match the trimmed pattern.
 *
 * @category String checks
 * @since 4.0.0
 */
export function isTrimmed(annotations) {
  return makeFilter(s => s.trim() === s, {
    expected: "a string with no leading or trailing whitespace",
    meta: {
      _tag: "isTrimmed",
      regExp: new globalThis.RegExp(TRIMMED_PATTERN)
    },
    toArbitraryConstraint: {
      string: {
        patterns: [TRIMMED_PATTERN]
      }
    },
    ...annotations
  });
}
/**
 * Validates that a string matches the specified regular expression pattern.
 *
 * **JSON Schema**
 *
 * This check corresponds to the `pattern` constraint in JSON Schema.
 *
 * **Arbitrary**
 *
 * When generating test data with fast-check, this applies a `patterns`
 * constraint to ensure generated strings match the specified RegExp pattern.
 *
 * @category String checks
 * @since 4.0.0
 */
export const isPattern = AST.isPattern;
/**
 * Validates that a string represents a finite number.
 *
 * **JSON Schema**
 *
 * This check corresponds to a `pattern` constraint in JSON Schema that matches
 * strings representing finite numbers.
 *
 * **Arbitrary**
 *
 * When generating test data with fast-check, this applies a `patterns`
 * constraint to ensure generated strings match the number string pattern.
 *
 * @category String checks
 * @since 4.0.0
 */
export const isStringFinite = AST.isStringFinite;
/**
 * Validates that a string represents a valid BigInt (can be parsed as a BigInt).
 *
 * **JSON Schema**
 *
 * This check corresponds to a `pattern` constraint in JSON Schema that matches
 * strings representing BigInt values.
 *
 * **Arbitrary**
 *
 * When generating test data with fast-check, this applies a `patterns`
 * constraint to ensure generated strings match the BigInt string pattern.
 *
 * @category String checks
 * @since 4.0.0
 */
export const isStringBigInt = AST.isStringBigInt;
/**
 * Validates that a string represents a valid Symbol (can be parsed as a Symbol).
 *
 * **JSON Schema**
 *
 * This check corresponds to a `pattern` constraint in JSON Schema that matches
 * strings representing Symbol values.
 *
 * **Arbitrary**
 *
 * When generating test data with fast-check, this applies a `patterns`
 * constraint to ensure generated strings match the Symbol string pattern.
 *
 * @category String checks
 * @since 4.0.0
 */
export const isStringSymbol = AST.isStringSymbol;
/**
 * Returns a RegExp for validating an RFC 4122 UUID.
 *
 * Optionally specify a version 1-8. If no version is specified (`undefined`), all versions are supported.
 */
const getUUIDRegExp = version => {
  if (version) {
    return new globalThis.RegExp(`^([0-9a-fA-F]{8}-[0-9a-fA-F]{4}-${version}[0-9a-fA-F]{3}-[89abAB][0-9a-fA-F]{3}-[0-9a-fA-F]{12})$`);
  }
  return /^([0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[1-8][0-9a-fA-F]{3}-[89abAB][0-9a-fA-F]{3}-[0-9a-fA-F]{12}|00000000-0000-0000-0000-000000000000)$/;
};
/**
 * Validates that a string is a valid Universally Unique Identifier (UUID).
 * Optionally specify a version (1-8) to validate against a specific UUID version.
 * If no version is specified (`undefined`), all versions are supported.
 *
 * **JSON Schema**
 *
 * This check corresponds to a `pattern` constraint in JSON Schema that matches
 * UUID format, and includes a `format: "uuid"` annotation.
 *
 * **Arbitrary**
 *
 * When generating test data with fast-check, this applies a `patterns`
 * constraint to ensure generated strings match the UUID pattern.
 *
 * @category String checks
 * @since 4.0.0
 */
export function isUUID(version, annotations) {
  const regExp = getUUIDRegExp(version);
  return isPattern(regExp, {
    expected: version ? `a UUID v${version}` : "a UUID",
    meta: {
      _tag: "isUUID",
      regExp,
      version
    },
    ...annotations
  });
}
/**
 * Validates that a string is a valid ULID (Universally Unique Lexicographically
 * Sortable Identifier).
 *
 * **JSON Schema**
 *
 * This check corresponds to a `pattern` constraint in JSON Schema that matches
 * the ULID format.
 *
 * **Arbitrary**
 *
 * When generating test data with fast-check, this applies a `patterns`
 * constraint to ensure generated strings match the ULID pattern.
 *
 * @category String checks
 * @since 4.0.0
 */
export function isULID(annotations) {
  const regExp = /^[0-9A-HJKMNP-TV-Za-hjkmnp-tv-z]{26}$/;
  return isPattern(regExp, {
    meta: {
      _tag: "isULID",
      regExp
    },
    ...annotations
  });
}
/**
 * Validates that a string is valid Base64 encoded data.
 *
 * **JSON Schema**
 *
 * This check corresponds to a `pattern` constraint in JSON Schema that matches
 * Base64 format.
 *
 * **Arbitrary**
 *
 * When generating test data with fast-check, this applies a `patterns`
 * constraint to ensure generated strings match the Base64 pattern.
 *
 * @category String checks
 * @since 4.0.0
 */
export function isBase64(annotations) {
  const regExp = /^([0-9a-zA-Z+/]{4})*(([0-9a-zA-Z+/]{2}==)|([0-9a-zA-Z+/]{3}=))?$/;
  return isPattern(regExp, {
    expected: "a base64 encoded string",
    meta: {
      _tag: "isBase64",
      regExp
    },
    ...annotations
  });
}
/**
 * Validates that a string is valid Base64URL encoded data (Base64 with URL-safe
 * characters).
 *
 * **JSON Schema**
 *
 * This check corresponds to a `pattern` constraint in JSON Schema that matches
 * Base64URL format.
 *
 * **Arbitrary**
 *
 * When generating test data with fast-check, this applies a `patterns`
 * constraint to ensure generated strings match the Base64URL pattern.
 *
 * @category String checks
 * @since 4.0.0
 */
export function isBase64Url(annotations) {
  const regExp = /^([0-9a-zA-Z-_]{4})*(([0-9a-zA-Z-_]{2}(==)?)|([0-9a-zA-Z-_]{3}(=)?))?$/;
  return isPattern(regExp, {
    expected: "a base64url encoded string",
    meta: {
      _tag: "isBase64Url",
      regExp
    },
    ...annotations
  });
}
/**
 * Validates that a string starts with the specified prefix.
 *
 * **JSON Schema**
 *
 * This check corresponds to a `pattern` constraint in JSON Schema that matches
 * strings starting with the specified prefix.
 *
 * **Arbitrary**
 *
 * When generating test data with fast-check, this applies a `patterns`
 * constraint to ensure generated strings start with the required prefix.
 *
 * @category String checks
 * @since 4.0.0
 */
export function isStartsWith(startsWith, annotations) {
  const formatted = JSON.stringify(startsWith);
  return makeFilter(s => s.startsWith(startsWith), {
    expected: `a string starting with ${formatted}`,
    meta: {
      _tag: "isStartsWith",
      startsWith,
      regExp: new globalThis.RegExp(`^${startsWith}`)
    },
    toArbitraryConstraint: {
      string: {
        patterns: [`^${startsWith}`]
      }
    },
    ...annotations
  });
}
/**
 * Validates that a string ends with the specified suffix.
 *
 * **JSON Schema**
 *
 * This check corresponds to a `pattern` constraint in JSON Schema that matches
 * strings ending with the specified suffix.
 *
 * **Arbitrary**
 *
 * When generating test data with fast-check, this applies a `patterns`
 * constraint to ensure generated strings end with the required suffix.
 *
 * @category String checks
 * @since 4.0.0
 */
export function isEndsWith(endsWith, annotations) {
  const formatted = JSON.stringify(endsWith);
  return makeFilter(s => s.endsWith(endsWith), {
    expected: `a string ending with ${formatted}`,
    meta: {
      _tag: "isEndsWith",
      endsWith,
      regExp: new globalThis.RegExp(`${endsWith}$`)
    },
    toArbitraryConstraint: {
      string: {
        patterns: [`${endsWith}$`]
      }
    },
    ...annotations
  });
}
/**
 * Validates that a string contains the specified substring.
 *
 * **JSON Schema**
 *
 * This check corresponds to a `pattern` constraint in JSON Schema that matches
 * strings containing the specified substring.
 *
 * **Arbitrary**
 *
 * When generating test data with fast-check, this applies a `patterns`
 * constraint to ensure generated strings contain the required substring.
 *
 * @category String checks
 * @since 4.0.0
 */
export function isIncludes(includes, annotations) {
  const formatted = JSON.stringify(includes);
  return makeFilter(s => s.includes(includes), {
    expected: `a string including ${formatted}`,
    meta: {
      _tag: "isIncludes",
      includes,
      regExp: new globalThis.RegExp(includes)
    },
    toArbitraryConstraint: {
      string: {
        patterns: [includes]
      }
    },
    ...annotations
  });
}
const UPPERCASED_PATTERN = "^[^a-z]*$";
/**
 * Validates that a string contains only uppercase characters.
 *
 * **JSON Schema**
 *
 * This check corresponds to a `pattern` constraint in JSON Schema that matches
 * strings with only uppercase characters.
 *
 * **Arbitrary**
 *
 * When generating test data with fast-check, this applies a `patterns`
 * constraint to ensure generated strings contain only uppercase characters.
 *
 * @category String checks
 * @since 4.0.0
 */
export function isUppercased(annotations) {
  return makeFilter(s => s.toUpperCase() === s, {
    expected: "a string with all characters in uppercase",
    meta: {
      _tag: "isUppercased",
      regExp: new globalThis.RegExp(UPPERCASED_PATTERN)
    },
    toArbitraryConstraint: {
      string: {
        patterns: [UPPERCASED_PATTERN]
      }
    },
    ...annotations
  });
}
const LOWERCASED_PATTERN = "^[^A-Z]*$";
/**
 * Validates that a string contains only lowercase characters.
 *
 * **JSON Schema**
 *
 * This check corresponds to a `pattern` constraint in JSON Schema that matches
 * strings with only lowercase characters.
 *
 * **Arbitrary**
 *
 * When generating test data with fast-check, this applies a `patterns`
 * constraint to ensure generated strings contain only lowercase characters.
 *
 * @category String checks
 * @since 4.0.0
 */
export function isLowercased(annotations) {
  return makeFilter(s => s.toLowerCase() === s, {
    expected: "a string with all characters in lowercase",
    meta: {
      _tag: "isLowercased",
      regExp: new globalThis.RegExp(LOWERCASED_PATTERN)
    },
    toArbitraryConstraint: {
      string: {
        patterns: [LOWERCASED_PATTERN]
      }
    },
    ...annotations
  });
}
const CAPITALIZED_PATTERN = "^[^a-z]?.*$";
/**
 * Validates that a string has its first character in uppercase.
 *
 * **JSON Schema**
 *
 * This check corresponds to a `pattern` constraint in JSON Schema that matches
 * strings with the first character in uppercase.
 *
 * **Arbitrary**
 *
 * When generating test data with fast-check, this applies a `patterns`
 * constraint to ensure generated strings have the first character in uppercase.
 *
 * @category String checks
 * @since 4.0.0
 */
export function isCapitalized(annotations) {
  return makeFilter(s => s.charAt(0).toUpperCase() === s.charAt(0), {
    expected: "a string with the first character in uppercase",
    meta: {
      _tag: "isCapitalized",
      regExp: new globalThis.RegExp(CAPITALIZED_PATTERN)
    },
    toArbitraryConstraint: {
      string: {
        patterns: [CAPITALIZED_PATTERN]
      }
    },
    ...annotations
  });
}
const UNCAPITALIZED_PATTERN = "^[^A-Z]?.*$";
/**
 * Validates that a string has its first character in lowercase.
 *
 * **JSON Schema**
 *
 * This check corresponds to a `pattern` constraint in JSON Schema that matches
 * strings with the first character in lowercase.
 *
 * **Arbitrary**
 *
 * When generating test data with fast-check, this applies a `patterns`
 * constraint to ensure generated strings have the first character in lowercase.
 *
 * @category String checks
 * @since 4.0.0
 */
export function isUncapitalized(annotations) {
  return makeFilter(s => s.charAt(0).toLowerCase() === s.charAt(0), {
    expected: "a string with the first character in lowercase",
    meta: {
      _tag: "isUncapitalized",
      regExp: new globalThis.RegExp(UNCAPITALIZED_PATTERN)
    },
    toArbitraryConstraint: {
      string: {
        patterns: [UNCAPITALIZED_PATTERN]
      }
    },
    ...annotations
  });
}
/**
 * Validates that a number is finite (not `Infinity`, `-Infinity`, or `NaN`).
 *
 * **JSON Schema**
 *
 * This check does not have a direct JSON Schema equivalent, but ensures the
 * number is valid and finite.
 *
 * **Arbitrary**
 *
 * When generating test data with fast-check, this applies `noDefaultInfinity`
 * and `noNaN` constraints to ensure generated numbers are finite.
 *
 * @category Number checks
 * @since 4.0.0
 */
export function isFinite(annotations) {
  return makeFilter(n => globalThis.Number.isFinite(n), {
    expected: "a finite number",
    meta: {
      _tag: "isFinite"
    },
    toArbitraryConstraint: {
      number: {
        noDefaultInfinity: true,
        noNaN: true
      }
    },
    ...annotations
  });
}
/**
 * @category Order checks
 * @since 4.0.0
 */
export function makeIsGreaterThan(options) {
  const gt = Order.isGreaterThan(options.order);
  const formatter = options.formatter ?? format;
  return (exclusiveMinimum, annotations) => {
    return makeFilter(input => gt(input, exclusiveMinimum), {
      expected: `a value greater than ${formatter(exclusiveMinimum)}`,
      ...options.annotate?.(exclusiveMinimum),
      ...annotations
    });
  };
}
/**
 * @category Order checks
 * @since 4.0.0
 */
export function makeIsGreaterThanOrEqualTo(options) {
  const gte = Order.isGreaterThanOrEqualTo(options.order);
  const formatter = options.formatter ?? format;
  return (minimum, annotations) => {
    return makeFilter(input => gte(input, minimum), {
      expected: `a value greater than or equal to ${formatter(minimum)}`,
      ...options.annotate?.(minimum),
      ...annotations
    });
  };
}
/**
 * @category Order checks
 * @since 4.0.0
 */
export function makeIsLessThan(options) {
  const lt = Order.isLessThan(options.order);
  const formatter = options.formatter ?? format;
  return (exclusiveMaximum, annotations) => {
    return makeFilter(input => lt(input, exclusiveMaximum), {
      expected: `a value less than ${formatter(exclusiveMaximum)}`,
      ...options.annotate?.(exclusiveMaximum),
      ...annotations
    });
  };
}
/**
 * @category Order checks
 * @since 4.0.0
 */
export function makeIsLessThanOrEqualTo(options) {
  const lte = Order.isLessThanOrEqualTo(options.order);
  const formatter = options.formatter ?? format;
  return (maximum, annotations) => {
    return makeFilter(input => lte(input, maximum), {
      expected: `a value less than or equal to ${formatter(maximum)}`,
      ...options.annotate?.(maximum),
      ...annotations
    });
  };
}
/**
 * @category Order checks
 * @since 4.0.0
 */
export function makeIsBetween(deriveOptions) {
  const greaterThanOrEqualTo = Order.isGreaterThanOrEqualTo(deriveOptions.order);
  const greaterThan = Order.isGreaterThan(deriveOptions.order);
  const lessThanOrEqualTo = Order.isLessThanOrEqualTo(deriveOptions.order);
  const lessThan = Order.isLessThan(deriveOptions.order);
  const formatter = deriveOptions.formatter ?? format;
  return (options, annotations) => {
    const gte = options.exclusiveMinimum ? greaterThan : greaterThanOrEqualTo;
    const lte = options.exclusiveMaximum ? lessThan : lessThanOrEqualTo;
    return makeFilter(input => gte(input, options.minimum) && lte(input, options.maximum), {
      expected: `a value between ${formatter(options.minimum)}${options.exclusiveMinimum ? " (excluded)" : ""} and ${formatter(options.maximum)}${options.exclusiveMaximum ? " (excluded)" : ""}`,
      ...deriveOptions.annotate?.(options),
      ...annotations
    });
  };
}
/**
 * @category Numeric checks
 * @since 4.0.0
 */
export function makeIsMultipleOf(options) {
  return (divisor, annotations) => {
    const formatter = options.formatter ?? format;
    return makeFilter(input => options.remainder(input, divisor) === options.zero, {
      expected: `a value that is a multiple of ${formatter(divisor)}`,
      ...options.annotate?.(divisor),
      ...annotations
    });
  };
}
/**
 * Validates that a number is greater than the specified value (exclusive).
 *
 * **JSON Schema**
 *
 * This check corresponds to the `exclusiveMinimum` constraint in JSON Schema.
 *
 * **Arbitrary**
 *
 * When generating test data with fast-check, this applies a `min` constraint
 * with `minExcluded: true` to ensure generated numbers are greater than the
 * specified value.
 *
 * @category Number checks
 * @since 4.0.0
 */
export const isGreaterThan = /*#__PURE__*/makeIsGreaterThan({
  order: Order.Number,
  annotate: exclusiveMinimum => ({
    meta: {
      _tag: "isGreaterThan",
      exclusiveMinimum
    },
    toArbitraryConstraint: {
      number: {
        min: exclusiveMinimum,
        minExcluded: true
      }
    }
  })
});
/**
 * Validates that a number is greater than or equal to the specified value
 * (inclusive).
 *
 * **JSON Schema**
 *
 * This check corresponds to the `minimum` constraint in JSON Schema.
 *
 * **Arbitrary**
 *
 * When generating test data with fast-check, this applies a `min` constraint
 * to ensure generated numbers are greater than or equal to the specified value.
 *
 * @category Number checks
 * @since 4.0.0
 */
export const isGreaterThanOrEqualTo = /*#__PURE__*/makeIsGreaterThanOrEqualTo({
  order: Order.Number,
  annotate: minimum => ({
    meta: {
      _tag: "isGreaterThanOrEqualTo",
      minimum
    },
    toArbitraryConstraint: {
      number: {
        min: minimum
      }
    }
  })
});
/**
 * Validates that a number is less than the specified value (exclusive).
 *
 * **JSON Schema**
 *
 * This check corresponds to the `exclusiveMaximum` constraint in JSON Schema.
 *
 * **Arbitrary**
 *
 * When generating test data with fast-check, this applies a `max` constraint
 * with `maxExcluded: true` to ensure generated numbers are less than the
 * specified value.
 *
 * @category Number checks
 * @since 4.0.0
 */
export const isLessThan = /*#__PURE__*/makeIsLessThan({
  order: Order.Number,
  annotate: exclusiveMaximum => ({
    meta: {
      _tag: "isLessThan",
      exclusiveMaximum
    },
    toArbitraryConstraint: {
      number: {
        max: exclusiveMaximum,
        maxExcluded: true
      }
    }
  })
});
/**
 * Validates that a number is less than or equal to the specified value
 * (inclusive).
 *
 * **JSON Schema**
 *
 * This check corresponds to the `maximum` constraint in JSON Schema.
 *
 * **Arbitrary**
 *
 * When generating test data with fast-check, this applies a `max` constraint
 * to ensure generated numbers are less than or equal to the specified value.
 *
 * @category Number checks
 * @since 4.0.0
 */
export const isLessThanOrEqualTo = /*#__PURE__*/makeIsLessThanOrEqualTo({
  order: Order.Number,
  annotate: maximum => ({
    meta: {
      _tag: "isLessThanOrEqualTo",
      maximum
    },
    toArbitraryConstraint: {
      number: {
        max: maximum
      }
    }
  })
});
/**
 * Validates that a number is within a specified range. The range boundaries can
 * be inclusive or exclusive based on the provided options.
 *
 * **JSON Schema**
 *
 * This check corresponds to `minimum`/`maximum` or `exclusiveMinimum`/`exclusiveMaximum`
 * constraints in JSON Schema, depending on the options provided.
 *
 * **Arbitrary**
 *
 * When generating test data with fast-check, this applies `min` and `max`
 * constraints with optional `minExcluded` and `maxExcluded` flags to ensure
 * generated numbers fall within the specified range.
 *
 * @category Number checks
 * @since 4.0.0
 */
export const isBetween = /*#__PURE__*/makeIsBetween({
  order: Order.Number,
  annotate: options => {
    return {
      meta: {
        _tag: "isBetween",
        ...options
      },
      toArbitraryConstraint: {
        number: {
          min: options.minimum,
          max: options.maximum,
          ...(options.exclusiveMinimum && {
            minExcluded: true
          }),
          ...(options.exclusiveMaximum && {
            maxExcluded: true
          })
        }
      }
    };
  }
});
/**
 * Validates that a number is a multiple of the specified divisor.
 *
 * **JSON Schema**
 *
 * This check corresponds to the `multipleOf` constraint in JSON Schema.
 *
 * **Arbitrary**
 *
 * When generating test data with fast-check, this applies constraints to ensure
 * generated numbers are multiples of the specified divisor.
 *
 * @category Number checks
 * @since 4.0.0
 */
export const isMultipleOf = /*#__PURE__*/makeIsMultipleOf({
  remainder,
  zero: 0,
  annotate: divisor => ({
    expected: `a value that is a multiple of ${divisor}`,
    meta: {
      _tag: "isMultipleOf",
      divisor
    }
  })
});
/**
 * Validates that a number is a safe integer (within the safe integer range
 * that can be exactly represented in JavaScript).
 *
 * **JSON Schema**
 *
 * This check corresponds to the `type: "integer"` constraint in JSON Schema.
 *
 * **Arbitrary**
 *
 * When generating test data with fast-check, this applies an `isInteger: true`
 * constraint to ensure generated numbers are integers.
 *
 * @category Integer checks
 * @since 4.0.0
 */
export function isInt(annotations) {
  return makeFilter(n => globalThis.Number.isSafeInteger(n), {
    expected: "an integer",
    meta: {
      _tag: "isInt"
    },
    toArbitraryConstraint: {
      number: {
        isInteger: true
      }
    },
    ...annotations
  });
}
/**
 * Validates that a number is a 32-bit signed integer (range: -2,147,483,648 to
 * 2,147,483,647).
 *
 * **JSON Schema**
 *
 * This check corresponds to the `format: "int32"` constraint in OpenAPI 3.1,
 * or `minimum`/`maximum` constraints in other JSON Schema targets.
 *
 * **Arbitrary**
 *
 * When generating test data with fast-check, this applies integer and range
 * constraints to ensure generated numbers are 32-bit signed integers.
 *
 * @category Integer checks
 * @since 4.0.0
 */
export function isInt32(annotations) {
  return new AST.FilterGroup([isInt(annotations), isBetween({
    minimum: -2147483648,
    maximum: 2147483647
  })], {
    expected: "a 32-bit integer",
    ...annotations
  });
}
/**
 * Validates that a number is a 32-bit unsigned integer (range: 0 to
 * 4,294,967,295).
 *
 * **JSON Schema**
 *
 * This check corresponds to the `format: "uint32"` constraint in OpenAPI 3.1,
 * or `minimum`/`maximum` constraints in other JSON Schema targets.
 *
 * **Arbitrary**
 *
 * When generating test data with fast-check, this applies integer and range
 * constraints to ensure generated numbers are 32-bit unsigned integers.
 *
 * @category Integer checks
 * @since 4.0.0
 */
export function isUint32(annotations) {
  return new AST.FilterGroup([isInt(), isBetween({
    minimum: 0,
    maximum: 4294967295
  })], {
    expected: "a 32-bit unsigned integer",
    ...annotations
  });
}
/**
 * Validates that a Date object represents a valid date (not an invalid date
 * like `new Date("invalid")`).
 *
 * **JSON Schema**
 *
 * This check does not have a direct JSON Schema equivalent, as JSON Schema
 * validates date strings, not Date objects.
 *
 * **Arbitrary**
 *
 * When generating test data with fast-check, this applies a `noInvalidDate`
 * constraint to ensure generated Date objects are valid.
 *
 * @category Date checks
 * @since 4.0.0
 */
export function isDateValid(annotations) {
  return makeFilter(date => !isNaN(date.getTime()), {
    expected: "a valid date",
    meta: {
      _tag: "isDateValid"
    },
    toArbitraryConstraint: {
      date: {
        noInvalidDate: true
      }
    },
    ...annotations
  });
}
/**
 * Validates that a Date is greater than the specified value (exclusive).
 *
 * **Arbitrary**
 *
 * When generating test data with fast-check, this applies a `min` constraint
 * with `minExcluded: true` to ensure generated Date objects are greater than the
 * specified value.
 *
 * @category Date checks
 * @since 4.0.0
 */
export const isGreaterThanDate = /*#__PURE__*/makeIsGreaterThan({
  order: Order.Date,
  annotate: exclusiveMinimum => ({
    meta: {
      _tag: "isGreaterThanDate",
      exclusiveMinimum
    },
    toArbitraryConstraint: {
      date: {
        min: exclusiveMinimum,
        minExcluded: true
      }
    }
  })
});
/**
 * Validates that a Date is greater than or equal to the specified date
 * (inclusive).
 *
 * **JSON Schema**
 *
 * This check does not have a direct JSON Schema equivalent, as JSON Schema
 * validates date strings, not Date objects.
 *
 * **Arbitrary**
 *
 * When generating test data with fast-check, this applies a `min` constraint
 * to ensure generated Date objects are greater than or equal to the specified
 * date.
 *
 * @category Date checks
 * @since 4.0.0
 */
export const isGreaterThanOrEqualToDate = /*#__PURE__*/makeIsGreaterThanOrEqualTo({
  order: Order.Date,
  annotate: minimum => ({
    meta: {
      _tag: "isGreaterThanOrEqualToDate",
      minimum
    },
    toArbitraryConstraint: {
      date: {
        min: minimum
      }
    }
  })
});
/**
 * Validates that a Date is less than the specified value (exclusive).
 *
 * **Arbitrary**
 *
 * When generating test data with fast-check, this applies a `max` constraint
 * with `maxExcluded: true` to ensure generated Date objects are less than the
 * specified value.
 *
 * @category Date checks
 * @since 4.0.0
 */
export const isLessThanDate = /*#__PURE__*/makeIsLessThan({
  order: Order.Date,
  annotate: exclusiveMaximum => ({
    meta: {
      _tag: "isLessThanDate",
      exclusiveMaximum
    },
    toArbitraryConstraint: {
      date: {
        max: exclusiveMaximum,
        maxExcluded: true
      }
    }
  })
});
/**
 * Validates that a Date is less than or equal to the specified date
 * (inclusive).
 *
 * **JSON Schema**
 *
 * This check does not have a direct JSON Schema equivalent, as JSON Schema
 * validates date strings, not Date objects.
 *
 * **Arbitrary**
 *
 * When generating test data with fast-check, this applies a `max` constraint
 * to ensure generated Date objects are less than or equal to the specified
 * date.
 *
 * @category Date checks
 * @since 4.0.0
 */
export const isLessThanOrEqualToDate = /*#__PURE__*/makeIsLessThanOrEqualTo({
  order: Order.Date,
  annotate: maximum => ({
    meta: {
      _tag: "isLessThanOrEqualToDate",
      maximum
    },
    toArbitraryConstraint: {
      date: {
        max: maximum
      }
    }
  })
});
/**
 * Validates that a Date is within a specified range. The range boundaries can
 * be inclusive or exclusive based on the provided options.
 *
 * **JSON Schema**
 *
 * This check does not have a direct JSON Schema equivalent, as JSON Schema
 * validates date strings, not Date objects.
 *
 * **Arbitrary**
 *
 * When generating test data with fast-check, this applies `min` and `max`
 * constraints to ensure generated Date objects fall within the specified range.
 *
 * @category Date checks
 * @since 4.0.0
 */
export const isBetweenDate = /*#__PURE__*/makeIsBetween({
  order: Order.Date,
  annotate: options => ({
    meta: {
      _tag: "isBetweenDate",
      ...options
    },
    toArbitraryConstraint: {
      date: {
        min: options.minimum,
        max: options.maximum
      }
    }
  })
});
/**
 * Validates that a BigInt is greater than the specified value (exclusive).
 *
 * **Arbitrary**
 *
 * When generating test data with fast-check, this applies a `min` constraint
 * with `minExcluded: true` to ensure generated BigInts are greater than the
 * specified value.
 *
 * @category BigInt checks
 * @since 4.0.0
 */
export const isGreaterThanBigInt = /*#__PURE__*/makeIsGreaterThan({
  order: Order.BigInt,
  annotate: exclusiveMinimum => ({
    meta: {
      _tag: "isGreaterThanBigInt",
      exclusiveMinimum
    },
    toArbitraryConstraint: {
      bigint: {
        min: exclusiveMinimum,
        minExcluded: true
      }
    }
  })
});
/**
 * Validates that a BigInt is greater than or equal to the specified value
 * (inclusive).
 *
 * **Arbitrary**
 *
 * When generating test data with fast-check, this applies a `min` constraint
 * to ensure generated BigInt values are greater than or equal to the specified
 * value.
 *
 * @category BigInt checks
 * @since 4.0.0
 */
export const isGreaterThanOrEqualToBigInt = /*#__PURE__*/makeIsGreaterThanOrEqualTo({
  order: Order.BigInt,
  annotate: minimum => ({
    meta: {
      _tag: "isGreaterThanOrEqualToBigInt",
      minimum
    },
    toArbitraryConstraint: {
      bigint: {
        min: minimum
      }
    }
  })
});
/**
 * Validates that a BigInt is less than the specified value (exclusive).
 *
 * **Arbitrary**
 *
 * When generating test data with fast-check, this applies a `max` constraint
 * with `maxExcluded: true` to ensure generated BigInts are less than the
 * specified value.
 *
 * @category BigInt checks
 * @since 4.0.0
 */
export const isLessThanBigInt = /*#__PURE__*/makeIsLessThan({
  order: Order.BigInt,
  annotate: exclusiveMaximum => ({
    meta: {
      _tag: "isLessThanBigInt",
      exclusiveMaximum
    },
    toArbitraryConstraint: {
      bigint: {
        max: exclusiveMaximum,
        maxExcluded: true
      }
    }
  })
});
/**
 * Validates that a BigInt is less than or equal to the specified value
 * (inclusive).
 *
 * **Arbitrary**
 *
 * When generating test data with fast-check, this applies a `max` constraint
 * to ensure generated BigInt values are less than or equal to the specified
 * value.
 *
 * @category BigInt checks
 * @since 4.0.0
 */
export const isLessThanOrEqualToBigInt = /*#__PURE__*/makeIsLessThanOrEqualTo({
  order: Order.BigInt,
  annotate: maximum => ({
    meta: {
      _tag: "isLessThanOrEqualToBigInt",
      maximum
    },
    toArbitraryConstraint: {
      bigint: {
        max: maximum
      }
    }
  })
});
/**
 * Validates that a BigInt is within a specified range. The range boundaries can
 * be inclusive or exclusive based on the provided options.
 *
 * **Arbitrary**
 *
 * When generating test data with fast-check, this applies `min` and `max`
 * constraints to ensure generated BigInt values fall within the specified
 * range.
 *
 * @category BigInt checks
 * @since 4.0.0
 */
export const isBetweenBigInt = /*#__PURE__*/makeIsBetween({
  order: Order.BigInt,
  annotate: options => ({
    meta: {
      _tag: "isBetweenBigInt",
      ...options
    },
    toArbitraryConstraint: {
      bigint: {
        min: options.minimum,
        max: options.maximum
      }
    }
  })
});
/**
 * Validates that a BigDecimal is greater than the specified value (exclusive).
 *
 * @category BigDecimal checks
 * @since 4.0.0
 */
export const isGreaterThanBigDecimal = /*#__PURE__*/makeIsGreaterThan({
  order: BigDecimal_.Order,
  formatter: bd => BigDecimal_.format(bd)
});
/**
 * Validates that a BigDecimal is greater than or equal to the specified value
 * (inclusive).
 *
 * @category BigDecimal checks
 * @since 4.0.0
 */
export const isGreaterThanOrEqualToBigDecimal = /*#__PURE__*/makeIsGreaterThanOrEqualTo({
  order: BigDecimal_.Order,
  formatter: bd => BigDecimal_.format(bd)
});
/**
 * Validates that a BigDecimal is less than the specified value (exclusive).
 *
 * @category BigDecimal checks
 * @since 4.0.0
 */
export const isLessThanBigDecimal = /*#__PURE__*/makeIsLessThan({
  order: BigDecimal_.Order,
  formatter: bd => BigDecimal_.format(bd)
});
/**
 * Validates that a BigDecimal is less than or equal to the specified value
 * (inclusive).
 *
 * @category BigDecimal checks
 * @since 4.0.0
 */
export const isLessThanOrEqualToBigDecimal = /*#__PURE__*/makeIsLessThanOrEqualTo({
  order: BigDecimal_.Order,
  formatter: bd => BigDecimal_.format(bd)
});
/**
 * Validates that a BigDecimal is within a specified range.
 *
 * @category BigDecimal checks
 * @since 4.0.0
 */
export const isBetweenBigDecimal = /*#__PURE__*/makeIsBetween({
  order: BigDecimal_.Order,
  formatter: bd => BigDecimal_.format(bd)
});
/**
 * Validates that a value has at least the specified length. Works with strings
 * and arrays.
 *
 * **JSON Schema**
 *
 * This check corresponds to the `minLength` constraint for strings or the
 * `minItems` constraint for arrays in JSON Schema.
 *
 * **Arbitrary**
 *
 * When generating test data with fast-check, this applies a `minLength`
 * constraint to ensure generated strings or arrays have at least the required
 * length.
 *
 * @category Length checks
 * @since 4.0.0
 */
export function isMinLength(minLength, annotations) {
  minLength = Math.max(0, Math.floor(minLength));
  return makeFilter(input => input.length >= minLength, {
    expected: `a value with a length of at least ${minLength}`,
    meta: {
      _tag: "isMinLength",
      minLength
    },
    [AST.STRUCTURAL_ANNOTATION_KEY]: true,
    toArbitraryConstraint: {
      string: {
        minLength
      },
      array: {
        minLength
      }
    },
    ...annotations
  });
}
/**
 * Validates that a value has at least one element. Works with strings and arrays.
 * This is equivalent to `isMinLength(1)`.
 *
 * **JSON Schema**
 *
 * This check corresponds to the `minLength: 1` constraint for strings or the
 * `minItems: 1` constraint for arrays in JSON Schema.
 *
 * **Arbitrary**
 *
 * When generating test data with fast-check, this applies a `minLength: 1`
 * constraint to ensure generated strings or arrays are non-empty.
 *
 * @category Length checks
 * @since 4.0.0
 */
export function isNonEmpty(annotations) {
  return isMinLength(1, annotations);
}
/**
 * Validates that a value has at most the specified length. Works with strings
 * and arrays.
 *
 * **JSON Schema**
 *
 * This check corresponds to the `maxLength` constraint for strings or the
 * `maxItems` constraint for arrays in JSON Schema.
 *
 * **Arbitrary**
 *
 * When generating test data with fast-check, this applies a `maxLength`
 * constraint to ensure generated strings or arrays have at most the required
 * length.
 *
 * @category Length checks
 * @since 4.0.0
 */
export function isMaxLength(maxLength, annotations) {
  maxLength = Math.max(0, Math.floor(maxLength));
  return makeFilter(input => input.length <= maxLength, {
    expected: `a value with a length of at most ${maxLength}`,
    meta: {
      _tag: "isMaxLength",
      maxLength
    },
    [AST.STRUCTURAL_ANNOTATION_KEY]: true,
    toArbitraryConstraint: {
      string: {
        maxLength
      },
      array: {
        maxLength
      }
    },
    ...annotations
  });
}
/**
 * Validates that a value's length is within the specified range. Works with
 * strings and arrays.
 *
 * **JSON Schema**
 *
 * This check corresponds to `minLength`/`maxLength` constraints for strings
 * or `minItems`/`maxItems` constraints for arrays in JSON Schema.
 *
 * **Arbitrary**
 *
 * When generating test data with fast-check, this applies `minLength` and
 * `maxLength` constraints to ensure generated strings or arrays have a length
 * within the specified range.
 *
 * @category Length checks
 * @since 4.0.0
 */
export function isLengthBetween(minimum, maximum, annotations) {
  minimum = Math.max(0, Math.floor(minimum));
  maximum = Math.max(0, Math.floor(maximum));
  return makeFilter(input => input.length >= minimum && input.length <= maximum, {
    expected: minimum === maximum ? `a value with a length of ${minimum}` : `a value with a length between ${minimum} and ${maximum}`,
    meta: {
      _tag: "isLengthBetween",
      minimum,
      maximum
    },
    [AST.STRUCTURAL_ANNOTATION_KEY]: true,
    toArbitraryConstraint: {
      string: {
        minLength: minimum,
        maxLength: maximum
      },
      array: {
        minLength: minimum,
        maxLength: maximum
      }
    },
    ...annotations
  });
}
/**
 * Validates that a value has at least the specified size. Works with values
 * that have a `size` property, such as `Set` or `Map`.
 *
 * **JSON Schema**
 *
 * This check does not have a direct JSON Schema equivalent, as it applies to
 * values with a `size` property rather than standard JSON Schema types.
 *
 * **Arbitrary**
 *
 * When generating test data with fast-check, this applies a `minLength`
 * constraint to the array representation to ensure generated values have at
 * least the required size.
 *
 * @category Size checks
 * @since 4.0.0
 */
export function isMinSize(minSize, annotations) {
  minSize = Math.max(0, Math.floor(minSize));
  return makeFilter(input => input.size >= minSize, {
    expected: `a value with a size of at least ${minSize}`,
    meta: {
      _tag: "isMinSize",
      minSize
    },
    [AST.STRUCTURAL_ANNOTATION_KEY]: true,
    toArbitraryConstraint: {
      array: {
        minLength: minSize
      }
    },
    ...annotations
  });
}
/**
 * Validates that a value has at most the specified size. Works with values
 * that have a `size` property, such as `Set` or `Map`.
 *
 * **JSON Schema**
 *
 * This check does not have a direct JSON Schema equivalent, as it applies to
 * values with a `size` property rather than standard JSON Schema types.
 *
 * **Arbitrary**
 *
 * When generating test data with fast-check, this applies a `maxLength`
 * constraint to the array representation to ensure generated values have at
 * most the required size.
 *
 * @category Size checks
 * @since 4.0.0
 */
export function isMaxSize(maxSize, annotations) {
  maxSize = Math.max(0, Math.floor(maxSize));
  return makeFilter(input => input.size <= maxSize, {
    expected: `a value with a size of at most ${maxSize}`,
    meta: {
      _tag: "isMaxSize",
      maxSize
    },
    [AST.STRUCTURAL_ANNOTATION_KEY]: true,
    toArbitraryConstraint: {
      array: {
        maxLength: maxSize
      }
    },
    ...annotations
  });
}
/**
 * Validates that a value's size is within the specified range. Works with
 * values that have a `size` property, such as `Set` or `Map`.
 *
 * **JSON Schema**
 *
 * This check does not have a direct JSON Schema equivalent, as it applies to
 * values with a `size` property rather than standard JSON Schema types.
 *
 * **Arbitrary**
 *
 * When generating test data with fast-check, this applies `minLength` and
 * `maxLength` constraints to ensure generated values have a size within the
 * specified range.
 *
 * @category Size checks
 * @since 4.0.0
 */
export function isSizeBetween(minimum, maximum, annotations) {
  minimum = Math.max(0, Math.floor(minimum));
  maximum = Math.max(0, Math.floor(maximum));
  return makeFilter(input => input.size >= minimum && input.size <= maximum, {
    expected: minimum === maximum ? `a value with a size of ${minimum}` : `a value with a size between ${minimum} and ${maximum}`,
    meta: {
      _tag: "isSizeBetween",
      minimum,
      maximum
    },
    [AST.STRUCTURAL_ANNOTATION_KEY]: true,
    toArbitraryConstraint: {
      array: {
        minLength: minimum,
        maxLength: maximum
      }
    },
    ...annotations
  });
}
/**
 * Validates that an object contains at least the specified number of
 * properties. This includes both string and symbol keys when counting
 * properties.
 *
 * **JSON Schema**
 *
 * This check corresponds to the `minProperties` constraint in JSON Schema.
 *
 * **Arbitrary**
 *
 * When generating test data with fast-check, this applies a `minLength`
 * constraint to the array of entries that is generated before being converted
 * to an object, ensuring the resulting object has at least the required number
 * of properties.
 *
 * @category Object checks
 * @since 4.0.0
 */
export function isMinProperties(minProperties, annotations) {
  minProperties = Math.max(0, Math.floor(minProperties));
  return makeFilter(input => Reflect.ownKeys(input).length >= minProperties, {
    expected: `a value with at least ${minProperties === 1 ? "1 entry" : `${minProperties} entries`}`,
    meta: {
      _tag: "isMinProperties",
      minProperties
    },
    [AST.STRUCTURAL_ANNOTATION_KEY]: true,
    toArbitraryConstraint: {
      array: {
        minLength: minProperties
      }
    },
    ...annotations
  });
}
/**
 * Validates that an object contains at most the specified number of properties.
 * This includes both string and symbol keys when counting properties.
 *
 * **JSON Schema**
 *
 * This check corresponds to the `maxProperties` constraint in JSON Schema.
 *
 * **Arbitrary**
 *
 * When generating test data with fast-check, this applies a `maxLength`
 * constraint to the array of entries that is generated before being converted
 * to an object, ensuring the resulting object has at most the required number
 * of properties.
 *
 * @category Object checks
 * @since 4.0.0
 */
export function isMaxProperties(maxProperties, annotations) {
  maxProperties = Math.max(0, Math.floor(maxProperties));
  return makeFilter(input => Reflect.ownKeys(input).length <= maxProperties, {
    expected: `a value with at most ${maxProperties === 1 ? "1 entry" : `${maxProperties} entries`}`,
    meta: {
      _tag: "isMaxProperties",
      maxProperties
    },
    [AST.STRUCTURAL_ANNOTATION_KEY]: true,
    toArbitraryConstraint: {
      array: {
        maxLength: maxProperties
      }
    },
    ...annotations
  });
}
/**
 * Validates that an object contains between `minimum` and `maximum` properties (inclusive).
 * This includes both string and symbol keys when counting properties.
 *
 * **JSON Schema**
 *
 * This check corresponds to `minProperties` and `maxProperties`
 * constraints in JSON Schema.
 *
 * **Arbitrary**
 *
 * When generating test data with fast-check, this applies `minLength` and
 * `maxLength` constraints to the array of entries that is generated before
 * being converted to an object.
 *
 * @category Object checks
 * @since 4.0.0
 */
export function isPropertiesLengthBetween(minimum, maximum, annotations) {
  minimum = Math.max(0, Math.floor(minimum));
  maximum = Math.max(0, Math.floor(maximum));
  return makeFilter(input => Reflect.ownKeys(input).length >= minimum && Reflect.ownKeys(input).length <= maximum, {
    expected: minimum === maximum ? `a value with exactly ${minimum === 1 ? "1 entry" : `${minimum} entries`}` : `a value with between ${minimum} and ${maximum} entries`,
    meta: {
      _tag: "isPropertiesLengthBetween",
      minimum,
      maximum
    },
    [AST.STRUCTURAL_ANNOTATION_KEY]: true,
    toArbitraryConstraint: {
      array: {
        minLength: minimum,
        maxLength: maximum
      }
    },
    ...annotations
  });
}
/**
 * Validates that all property names in an object satisfy the provided key
 * schema (encoded side of the schema).
 *
 * **JSON Schema**
 *
 * This check corresponds to the `propertyNames` constraint in JSON Schema.
 *
 * @category Object checks
 * @since 4.0.0
 */
export function isPropertyNames(keySchema, annotations) {
  const propertyNames = toEncoded(keySchema);
  const parser = Parser._issue(propertyNames.ast);
  return makeFilter((input, ast, options) => {
    const keys = Reflect.ownKeys(input);
    const issues = [];
    for (const key of keys) {
      const issue = parser(key, options);
      if (issue !== undefined) {
        issues.push(new Issue.Pointer([key], issue));
        if (options.errors === "first") break;
      }
    }
    if (Arr.isArrayNonEmpty(issues)) {
      return new Issue.Composite(ast, Option_.some(input), issues);
    }
    return true;
  }, {
    expected: "an object with property names matching the schema",
    meta: {
      _tag: "isPropertyNames",
      propertyNames: propertyNames.ast
    },
    [AST.STRUCTURAL_ANNOTATION_KEY]: true,
    ...annotations
  });
}
/**
 * Validates that all items in an array are unique according to the provided
 * equivalence function.
 *
 * **JSON Schema**
 *
 * This check corresponds to the `uniqueItems: true` constraint in JSON Schema.
 *
 * **Arbitrary**
 *
 * When generating test data with fast-check, this applies a `comparator`
 * constraint using the provided equivalence function to ensure generated arrays
 * contain only unique items.
 *
 * @since 4.0.0
 */
export function isUnique(annotations) {
  const equivalence = Equal.asEquivalence();
  return makeFilter(input => Arr.dedupeWith(input, equivalence).length === input.length, {
    expected: "an array with unique items",
    meta: {
      _tag: "isUnique"
    },
    toArbitraryConstraint: {
      array: {
        comparator: equivalence
      }
    },
    ...annotations
  });
}
// -----------------------------------------------------------------------------
// Built-in Schemas
// -----------------------------------------------------------------------------
/**
 * A schema for non-empty strings. Validates that a string has at least one
 * character.
 *
 * @since 4.0.0
 */
export const NonEmptyString = /*#__PURE__*/String.check(/*#__PURE__*/isNonEmpty());
/**
 * A schema representing a single character.
 *
 * @since 4.0.0
 */
export const Char = /*#__PURE__*/String.check(/*#__PURE__*/isLengthBetween(1, 1));
/**
 * @category Option
 * @since 4.0.0
 */
export function Option(value) {
  const schema = declareConstructor()([value], ([value]) => (input, ast, options) => {
    if (Option_.isOption(input)) {
      if (Option_.isNone(input)) {
        return Effect.succeedNone;
      }
      return Effect.mapBothEager(Parser.decodeUnknownEffect(value)(input.value, options), {
        onSuccess: Option_.some,
        onFailure: issue => new Issue.Composite(ast, Option_.some(input), [new Issue.Pointer(["value"], issue)])
      });
    }
    return Effect.fail(new Issue.InvalidType(ast, Option_.some(input)));
  }, {
    typeConstructor: {
      _tag: "effect/Option"
    },
    generation: {
      runtime: `Schema.Option(?)`,
      Type: `Option.Option<?>`,
      importDeclaration: `import * as Option from "effect/Option"`
    },
    expected: "Option",
    toCodec: ([value]) => link()(Union([Struct({
      _tag: Literal("Some"),
      value
    }), Struct({
      _tag: Literal("None")
    })]), Transformation.transform({
      decode: e => e._tag === "None" ? Option_.none() : Option_.some(e.value),
      encode: o => Option_.isSome(o) ? {
        _tag: "Some",
        value: o.value
      } : {
        _tag: "None"
      }
    })),
    toArbitrary: ([value]) => (fc, ctx) => {
      return fc.oneof(ctx?.isSuspend ? {
        maxDepth: 2,
        depthIdentifier: "Option"
      } : {}, fc.constant(Option_.none()), value.map(Option_.some));
    },
    toEquivalence: ([value]) => Option_.makeEquivalence(value),
    toFormatter: ([value]) => Option_.match({
      onNone: () => "none()",
      onSome: t => `some(${value(t)})`
    })
  });
  return make(schema.ast, {
    value
  });
}
/**
 * Decodes a nullable, required value `T` to a required `Option<T>` value.
 *
 * Decoding:
 * - `null` is decoded as `None`
 * - other values are decoded as `Some`
 *
 * Encoding:
 * - `None` is encoded as `null`
 * - `Some` is encoded as the value
 *
 * @category Option
 * @since 4.0.0
 */
export function OptionFromNullOr(schema) {
  return NullOr(schema).pipe(decodeTo(Option(toType(schema)), Transformation.optionFromNullOr()));
}
/**
 * Decodes an undefined-or value `T` to a required `Option<T>` value.
 *
 * Decoding:
 * - `undefined` is decoded as `None`
 * - other values are decoded as `Some`
 *
 * Encoding:
 * - `None` is encoded as `undefined`
 * - `Some` is encoded as the value
 *
 * @category Option
 * @since 4.0.0
 */
export function OptionFromUndefinedOr(schema) {
  return UndefinedOr(schema).pipe(decodeTo(Option(toType(schema)), Transformation.optionFromUndefinedOr()));
}
/**
 * Decodes a nullish value `T` to a required `Option<T>` value.
 *
 * Decoding:
 * - `null` and `undefined` are decoded as `None`
 * - other values are decoded as `Some`
 *
 * Encoding:
 * - `None` is encoded as `null` or `undefined` depending on the provided `options.onNoneEncoding` (defaults to `undefined`)
 * - `Some` is encoded as the value
 *
 * @category Option
 * @since 4.0.0
 */
export function OptionFromNullishOr(schema, options) {
  return NullishOr(schema).pipe(decodeTo(Option(toType(schema)), Transformation.optionFromNullishOr(options)));
}
/**
 * Decodes an optional value `A` to a required `Option<A>` value.
 *
 * Decoding:
 * - a missing key is decoded as `None`
 * - a present value is decoded as `Some`
 *
 * Encoding:
 * - `None` is encoded as missing key
 * - `Some` is encoded as the value
 *
 * @category Option
 * @since 4.0.0
 */
export function OptionFromOptionalKey(schema) {
  return optionalKey(schema).pipe(decodeTo(Option(toType(schema)), Transformation.optionFromOptionalKey()));
}
/**
 * Decodes an optional or `undefined` value `A` to an required `Option<A>`
 * value.
 *
 * Decoding:
 * - a missing key is decoded as `None`
 * - a present key with an `undefined` value is decoded as `None`
 * - all other values are decoded as `Some`
 *
 * Encoding:
 * - `None` is encoded as missing key
 * - `Some` is encoded as the value
 *
 * @category Option
 * @since 4.0.0
 */
export function OptionFromOptional(schema) {
  return optional(schema).pipe(decodeTo(Option(toType(schema)), Transformation.optionFromOptional()));
}
/**
 * @category Result
 * @since 4.0.0
 */
export function Result(success, failure) {
  const schema = declareConstructor()([success, failure], ([success, failure]) => (input, ast, options) => {
    if (!Result_.isResult(input)) {
      return Effect.fail(new Issue.InvalidType(ast, Option_.some(input)));
    }
    switch (input._tag) {
      case "Success":
        return Effect.mapBothEager(Parser.decodeEffect(success)(input.success, options), {
          onSuccess: Result_.succeed,
          onFailure: issue => new Issue.Composite(ast, Option_.some(input), [new Issue.Pointer(["success"], issue)])
        });
      case "Failure":
        return Effect.mapBothEager(Parser.decodeEffect(failure)(input.failure, options), {
          onSuccess: Result_.fail,
          onFailure: issue => new Issue.Composite(ast, Option_.some(input), [new Issue.Pointer(["failure"], issue)])
        });
    }
  }, {
    typeConstructor: {
      _tag: "effect/Result"
    },
    generation: {
      runtime: `Schema.Result(?, ?)`,
      Type: `Result.Result<?, ?>`,
      importDeclaration: `import * as Result from "effect/Result"`
    },
    expected: "Result",
    toCodec: ([success, failure]) => link()(Union([Struct({
      _tag: Literal("Success"),
      success
    }), Struct({
      _tag: Literal("Failure"),
      failure
    })]), Transformation.transform({
      decode: e => e._tag === "Success" ? Result_.succeed(e.success) : Result_.fail(e.failure),
      encode: r => Result_.isSuccess(r) ? {
        _tag: "Success",
        success: r.success
      } : {
        _tag: "Failure",
        failure: r.failure
      }
    })),
    toArbitrary: ([success, failure]) => (fc, ctx) => {
      return fc.oneof(ctx?.isSuspend ? {
        maxDepth: 2,
        depthIdentifier: "Result"
      } : {}, success.map(Result_.succeed), failure.map(Result_.fail));
    },
    toEquivalence: ([success, failure]) => Result_.makeEquivalence(success, failure),
    toFormatter: ([success, failure]) => Result_.match({
      onSuccess: t => `success(${success(t)})`,
      onFailure: t => `failure(${failure(t)})`
    })
  });
  return make(schema.ast, {
    success,
    failure
  });
}
/**
 * Creates a schema for the `Redacted` type, providing secure handling of
 * sensitive information.
 *
 * If the wrapped schema fails, the issue will be redacted to prevent both
 * the actual value and the schema details from being exposed.
 *
 * **Options**
 *
 * - `label`: When provided, the schema will behave as follows:
 *   - Values will be validated against the label in addition to the wrapped schema
 *   - The default JSON serializer will deserialize into a `Redacted` instance with the label
 *   - The arbitrary generator will produce a `Redacted` instance with the label
 *   - The formatter will return the label
 *
 * **Default JSON serializer**
 *
 * The default JSON serializer will fail when attempting to serialize a `Redacted` value,
 * but it will deserialize a value into a `Redacted` instance.
 *
 * @category Redacted
 * @since 4.0.0
 */
export function Redacted(value, options) {
  const decodeLabel = typeof options?.label === "string" ? Parser.decodeUnknownEffect(Literal(options.label)) : undefined;
  const schema = declareConstructor()([value], ([value]) => (input, ast, poptions) => {
    if (Redacted_.isRedacted(input)) {
      const label = decodeLabel !== undefined ? Effect.mapErrorEager(decodeLabel(input.label, poptions), issue => new Issue.Pointer(["label"], issue)) : Effect.void;
      return Effect.flatMapEager(label, () => Effect.mapBothEager(Parser.decodeUnknownEffect(value)(Redacted_.value(input), poptions), {
        onSuccess: () => input,
        onFailure: (/** ignore the actual issue because of security reasons */
        ) => {
          const oinput = Option_.some(input);
          return new Issue.Composite(ast, oinput, [new Issue.Pointer(["value"], new Issue.InvalidValue(oinput))]);
        }
      }));
    }
    return Effect.fail(new Issue.InvalidType(ast, Option_.some(input)));
  }, {
    typeConstructor: {
      _tag: "effect/Redacted"
    },
    generation: {
      runtime: `Schema.Redacted(?)`,
      Type: `Redacted.Redacted<?>`,
      importDeclaration: `import * as Redacted from "effect/Redacted"`
    },
    expected: "Redacted",
    toCodecJson: ([value]) => link()(redact(value), {
      decode: Getter.transform(e => Redacted_.make(e, {
        label: options?.label
      })),
      encode: Getter.forbidden(oe => "Cannot serialize Redacted" + (Option_.isSome(oe) && typeof oe.value.label === "string" ? ` with label: "${oe.value.label}"` : ""))
    }),
    toArbitrary: ([value]) => () => value.map(a => Redacted_.make(a, {
      label: options?.label
    })),
    toFormatter: () => globalThis.String,
    toEquivalence: ([value]) => Redacted_.makeEquivalence(value)
  });
  return make(schema.ast, {
    value
  });
}
/**
 * @category Redacted
 * @since 4.0.0
 */
export function redact(schema) {
  return schema.pipe(middlewareDecoding(Effect.mapErrorEager(Issue.redact)));
}
/**
 * @category Redacted
 * @since 4.0.0
 */
export function RedactedFromValue(value, options) {
  return redact(value).pipe(decodeTo(Redacted(toType(value), options), {
    decode: Getter.transform(t => Redacted_.make(t, {
      label: options?.label
    })),
    encode: Getter.forbidden(oe => "Cannot encode Redacted" + (Option_.isSome(oe) && typeof oe.value.label === "string" ? ` with label: "${oe.value.label}"` : ""))
  }));
}
/**
 * @category CauseReason
 * @since 4.0.0
 */
export function CauseReason(error, defect) {
  const schema = declareConstructor()([error, defect], ([error, defect]) => (input, ast, options) => {
    if (!Cause_.isReason(input)) {
      return Effect.fail(new Issue.InvalidType(ast, Option_.some(input)));
    }
    switch (input._tag) {
      case "Fail":
        return Effect.mapBothEager(Parser.decodeUnknownEffect(error)(input.error, options), {
          onSuccess: Cause_.makeFailReason,
          onFailure: issue => new Issue.Composite(ast, Option_.some(input), [new Issue.Pointer(["error"], issue)])
        });
      case "Die":
        return Effect.mapBothEager(Parser.decodeUnknownEffect(defect)(input.defect, options), {
          onSuccess: Cause_.makeDieReason,
          onFailure: issue => new Issue.Composite(ast, Option_.some(input), [new Issue.Pointer(["defect"], issue)])
        });
      case "Interrupt":
        return Effect.succeed(input);
    }
  }, {
    typeConstructor: {
      _tag: "effect/Cause/Failure"
    },
    generation: {
      runtime: `Schema.CauseReason(?, ?)`,
      Type: `Cause.Failure<?, ?>`,
      importDeclaration: `import * as Cause from "effect/Cause"`
    },
    expected: "Cause.Failure",
    toCodec: ([error, defect]) => link()(Union([Struct({
      _tag: Literal("Fail"),
      error
    }), Struct({
      _tag: Literal("Die"),
      defect
    }), Struct({
      _tag: Literal("Interrupt"),
      fiberId: UndefinedOr(Finite)
    })]), Transformation.transform({
      decode: e => {
        switch (e._tag) {
          case "Fail":
            return Cause_.makeFailReason(e.error);
          case "Die":
            return Cause_.makeDieReason(e.defect);
          case "Interrupt":
            return Cause_.makeInterruptReason(e.fiberId);
        }
      },
      encode: identity
    })),
    toArbitrary: ([error, defect]) => causeReasonToArbitrary(error, defect),
    toEquivalence: ([error, defect]) => causeReasonToEquivalence(error, defect),
    toFormatter: ([error, defect]) => causeReasonToFormatter(error, defect)
  });
  return make(schema.ast, {
    error,
    defect
  });
}
function causeReasonToArbitrary(error, defect) {
  return (fc, ctx) => {
    return fc.oneof(ctx?.isSuspend ? {
      maxDepth: 2,
      depthIdentifier: "Cause.Failure"
    } : {}, fc.constant(Cause_.makeInterruptReason()), fc.integer({
      min: 1
    }).map(Cause_.makeInterruptReason), error.map(e => Cause_.makeFailReason(e)), defect.map(d => Cause_.makeDieReason(d)));
  };
}
function causeReasonToEquivalence(error, defect) {
  return (a, b) => {
    if (a._tag !== b._tag) return false;
    switch (a._tag) {
      case "Fail":
        return error(a.error, b.error);
      case "Die":
        return defect(a.defect, b.defect);
      case "Interrupt":
        return a.fiberId === b.fiberId;
    }
  };
}
function causeReasonToFormatter(error, defect) {
  return t => {
    switch (t._tag) {
      case "Fail":
        return `Fail(${error(t.error)})`;
      case "Die":
        return `Die(${defect(t.defect)})`;
      case "Interrupt":
        return "Interrupt";
    }
  };
}
/**
 * @category Cause
 * @since 4.0.0
 */
export function Cause(error, defect) {
  const schema = declareConstructor()([error, defect], ([error, defect]) => {
    const failures = Array(CauseReason(error, defect));
    return (input, ast, options) => {
      if (!Cause_.isCause(input)) {
        return Effect.fail(new Issue.InvalidType(ast, Option_.some(input)));
      }
      return Effect.mapBothEager(Parser.decodeUnknownEffect(failures)(input.reasons, options), {
        onSuccess: Cause_.fromReasons,
        onFailure: issue => new Issue.Composite(ast, Option_.some(input), [new Issue.Pointer(["failures"], issue)])
      });
    };
  }, {
    typeConstructor: {
      _tag: "effect/Cause"
    },
    generation: {
      runtime: `Schema.Cause(?, ?)`,
      Type: `Cause.Cause<?, ?>`,
      importDeclaration: `import * as Cause from "effect/Cause"`
    },
    expected: "Cause",
    toCodec: ([error, defect]) => link()(Array(CauseReason(error, defect)), Transformation.transform({
      decode: Cause_.fromReasons,
      encode: ({
        reasons: failures
      }) => failures
    })),
    toArbitrary: ([error, defect]) => causeToArbitrary(error, defect),
    toEquivalence: ([error, defect]) => causeToEquivalence(error, defect),
    toFormatter: ([error, defect]) => causeToFormatter(error, defect)
  });
  return make(schema.ast, {
    error,
    defect
  });
}
function causeToArbitrary(error, defect) {
  return (fc, ctx) => {
    return fc.array(causeReasonToArbitrary(error, defect)(fc, ctx)).map(Cause_.fromReasons);
  };
}
function causeToEquivalence(error, defect) {
  const failures = Equivalence.Array(causeReasonToEquivalence(error, defect));
  return (a, b) => failures(a.reasons, b.reasons);
}
function causeToFormatter(error, defect) {
  const causeReason = causeReasonToFormatter(error, defect);
  return t => `Cause([${t.reasons.map(causeReason).join(", ")}])`;
}
const ErrorJsonEncoded = /*#__PURE__*/Struct({
  message: String,
  name: /*#__PURE__*/optionalKey(String),
  stack: /*#__PURE__*/optionalKey(String)
});
/**
 * A schema that represents `Error` objects.
 *
 * The default json serializer decodes to a struct with `name` and `message`
 * properties (stack is omitted for security).
 *
 * @category Schemas
 * @since 4.0.0
 */
export const Error = /*#__PURE__*/instanceOf(globalThis.Error, {
  typeConstructor: {
    _tag: "Error"
  },
  generation: {
    runtime: `Schema.Error`,
    Type: `globalThis.Error`
  },
  expected: "Error",
  toCodecJson: () => link()(ErrorJsonEncoded, Transformation.errorFromErrorJsonEncoded()),
  toArbitrary: () => fc => fc.string().map(message => new globalThis.Error(message))
});
/**
 * A schema that represents `Error` objects.
 *
 * The default json serializer decodes to a struct with `name`, `message` and
 * `stack` properties.
 *
 * @category Schemas
 * @since 4.0.0
 */
export const ErrorWithStack = /*#__PURE__*/instanceOf(globalThis.Error, {
  typeConstructor: {
    _tag: "ErrorWithStack"
  },
  generation: {
    runtime: `Schema.ErrorWithStack`,
    Type: `globalThis.Error`
  },
  expected: "Error",
  toCodecJson: () => link()(ErrorJsonEncoded, Transformation.errorFromErrorJsonEncoded({
    includeStack: true
  })),
  toArbitrary: () => fc => fc.string().map(message => new globalThis.Error(message))
});
const defectTransformation = /*#__PURE__*/new Transformation.Transformation(/*#__PURE__*/Getter.passthrough(), /*#__PURE__*/Getter.transform(u => {
  try {
    return JSON.parse(JSON.stringify(u));
  } catch {
    return format(u);
  }
}));
/**
 * A schema that represents defects.
 *
 * @category Constructors
 * @since 4.0.0
 */
export const Defect = /*#__PURE__*/Union([/*#__PURE__*/ErrorJsonEncoded.pipe(/*#__PURE__*/decodeTo(Error, /*#__PURE__*/Transformation.errorFromErrorJsonEncoded())), /*#__PURE__*/Any.pipe(/*#__PURE__*/decodeTo(/*#__PURE__*/Unknown.annotate({
  toCodecJson: () => link()(Any, defectTransformation),
  toArbitrary: () => fc => fc.json()
}), defectTransformation))]);
/**
 * A schema that represents defects, that also includes stack traces in the
 * encoded form.
 *
 * @category Constructors
 * @since 4.0.0
 */
export const DefectWithStack = /*#__PURE__*/Union([/*#__PURE__*/ErrorJsonEncoded.pipe(/*#__PURE__*/decodeTo(ErrorWithStack, /*#__PURE__*/Transformation.errorFromErrorJsonEncoded({
  includeStack: true
}))), /*#__PURE__*/Any.pipe(/*#__PURE__*/decodeTo(/*#__PURE__*/Unknown.annotate({
  toCodecJson: () => link()(Any, defectTransformation),
  toArbitrary: () => fc => fc.json()
}), defectTransformation))]);
/**
 * @category Exit
 * @since 4.0.0
 */
export function Exit(value, error, defect) {
  const schema = declareConstructor()([value, error, defect], ([value, error, defect]) => {
    const cause = Cause(error, defect);
    return (input, ast, options) => {
      if (!Exit_.isExit(input)) {
        return Effect.fail(new Issue.InvalidType(ast, Option_.some(input)));
      }
      switch (input._tag) {
        case "Success":
          return Effect.mapBothEager(Parser.decodeUnknownEffect(value)(input.value, options), {
            onSuccess: Exit_.succeed,
            onFailure: issue => new Issue.Composite(ast, Option_.some(input), [new Issue.Pointer(["value"], issue)])
          });
        case "Failure":
          return Effect.mapBothEager(Parser.decodeUnknownEffect(cause)(input.cause, options), {
            onSuccess: Exit_.failCause,
            onFailure: issue => new Issue.Composite(ast, Option_.some(input), [new Issue.Pointer(["cause"], issue)])
          });
      }
    };
  }, {
    typeConstructor: {
      _tag: "effect/Exit"
    },
    generation: {
      runtime: `Schema.Exit(?, ?, ?)`,
      Type: `Exit.Exit<?, ?, ?>`,
      importDeclaration: `import * as Exit from "effect/Exit"`
    },
    expected: "Exit",
    toCodec: ([value, error, defect]) => link()(Union([Struct({
      _tag: Literal("Success"),
      value
    }), Struct({
      _tag: Literal("Failure"),
      cause: Cause(error, defect)
    })]), Transformation.transform({
      decode: e => e._tag === "Success" ? Exit_.succeed(e.value) : Exit_.failCause(e.cause),
      encode: exit => Exit_.isSuccess(exit) ? {
        _tag: "Success",
        value: exit.value
      } : {
        _tag: "Failure",
        cause: exit.cause
      }
    })),
    toArbitrary: ([value, error, defect]) => (fc, ctx) => fc.oneof(ctx?.isSuspend ? {
      maxDepth: 2,
      depthIdentifier: "Exit"
    } : {}, value.map(v => Exit_.succeed(v)), causeToArbitrary(error, defect)(fc, ctx).map(cause => Exit_.failCause(cause))),
    toEquivalence: ([value, error, defect]) => {
      const cause = causeToEquivalence(error, defect);
      return (a, b) => {
        if (a._tag !== b._tag) return false;
        switch (a._tag) {
          case "Success":
            return value(a.value, b.value);
          case "Failure":
            return cause(a.cause, b.cause);
        }
      };
    },
    toFormatter: ([value, error, defect]) => {
      const cause = causeToFormatter(error, defect);
      return t => {
        switch (t._tag) {
          case "Success":
            return `Exit.Success(${value(t.value)})`;
          case "Failure":
            return `Exit.Failure(${cause(t.cause)})`;
        }
      };
    }
  });
  return make(schema.ast, {
    value,
    error,
    defect
  });
}
/**
 * Creates a schema that validates a `ReadonlyMap` where keys and values must
 * conform to the provided schemas.
 *
 * @category ReadonlyMap
 * @since 4.0.0
 */
export function ReadonlyMap(key, value) {
  const schema = declareConstructor()([key, value], ([key, value]) => {
    const array = Array(Tuple([key, value]));
    return (input, ast, options) => {
      if (input instanceof globalThis.Map) {
        return Effect.mapBothEager(Parser.decodeUnknownEffect(array)([...input], options), {
          onSuccess: array => new globalThis.Map(array),
          onFailure: issue => new Issue.Composite(ast, Option_.some(input), [new Issue.Pointer(["entries"], issue)])
        });
      }
      return Effect.fail(new Issue.InvalidType(ast, Option_.some(input)));
    };
  }, {
    typeConstructor: {
      _tag: "ReadonlyMap"
    },
    generation: {
      runtime: `Schema.ReadonlyMap(?, ?)`,
      Type: `globalThis.ReadonlyMap<?, ?>`
    },
    expected: "ReadonlyMap",
    toCodec: ([key, value]) => link()(Array(Tuple([key, value])), Transformation.transform({
      decode: e => new globalThis.Map(e),
      encode: map => [...map.entries()]
    })),
    toArbitrary: ([key, value]) => (fc, ctx) => {
      return fc.oneof(ctx?.isSuspend ? {
        maxDepth: 2,
        depthIdentifier: "ReadonlyMap"
      } : {}, fc.constant([]), fc.array(fc.tuple(key, value), ctx?.constraints?.array)).map(as => new globalThis.Map(as));
    },
    toEquivalence: ([key, value]) => Equal.makeCompareMap(key, value),
    toFormatter: ([key, value]) => t => {
      const size = t.size;
      if (size === 0) {
        return "ReadonlyMap(0) {}";
      }
      const entries = globalThis.Array.from(t.entries()).sort().map(([k, v]) => `${key(k)} => ${value(v)}`);
      return `ReadonlyMap(${size}) { ${entries.join(", ")} }`;
    }
  });
  return make(schema.ast, {
    key,
    value
  });
}
/**
 * Creates a schema that validates a `HashMap` where keys and values must
 * conform to the provided schemas.
 *
 * @category HashMap
 * @since 4.0.0
 */
export function HashMap(key, value) {
  const schema = declareConstructor()([key, value], ([key, value]) => {
    const entries = Array(Tuple([key, value]));
    return (input, ast, options) => {
      if (HashMap_.isHashMap(input)) {
        return Effect.mapBothEager(Parser.decodeUnknownEffect(entries)(HashMap_.toEntries(input), options), {
          onSuccess: HashMap_.fromIterable,
          onFailure: issue => new Issue.Composite(ast, Option_.some(input), [new Issue.Pointer(["entries"], issue)])
        });
      }
      return Effect.fail(new Issue.InvalidType(ast, Option_.some(input)));
    };
  }, {
    typeConstructor: {
      _tag: "effect/HashMap"
    },
    generation: {
      runtime: `Schema.HashMap(?, ?)`,
      Type: `HashMap.HashMap<?, ?>`,
      importDeclaration: `import * as HashMap from "effect/HashMap"`
    },
    expected: "HashMap",
    toCodec: ([key, value]) => link()(Array(Tuple([key, value])), Transformation.transform({
      decode: HashMap_.fromIterable,
      encode: HashMap_.toEntries
    })),
    toArbitrary: ([key, value]) => (fc, ctx) => {
      return fc.oneof(ctx?.isSuspend ? {
        maxDepth: 2,
        depthIdentifier: "HashMap"
      } : {}, fc.constant([]), fc.array(fc.tuple(key, value), ctx?.constraints?.array)).map(HashMap_.fromIterable);
    },
    toEquivalence: ([key, value]) => Equal.makeCompareMap(key, value),
    toFormatter: ([key, value]) => t => {
      const size = HashMap_.size(t);
      if (size === 0) {
        return "HashMap(0) {}";
      }
      const entries = HashMap_.toEntries(t).sort().map(([k, v]) => `${key(k)} => ${value(v)}`);
      return `HashMap(${size}) { ${entries.join(", ")} }`;
    }
  });
  return make(schema.ast, {
    key,
    value
  });
}
/**
 * @category ReadonlySet
 * @since 4.0.0
 */
export function ReadonlySet(value) {
  const schema = declareConstructor()([value], ([value]) => {
    const array = Array(value);
    return (input, ast, options) => {
      if (input instanceof globalThis.Set) {
        return Effect.mapBothEager(Parser.decodeUnknownEffect(array)([...input], options), {
          onSuccess: array => new globalThis.Set(array),
          onFailure: issue => new Issue.Composite(ast, Option_.some(input), [new Issue.Pointer(["values"], issue)])
        });
      }
      return Effect.fail(new Issue.InvalidType(ast, Option_.some(input)));
    };
  }, {
    typeConstructor: {
      _tag: "ReadonlySet"
    },
    generation: {
      runtime: `Schema.ReadonlySet(?)`,
      Type: `globalThis.ReadonlySet<?>`
    },
    expected: "ReadonlySet",
    toCodec: ([value]) => link()(Array(value), Transformation.transform({
      decode: e => new globalThis.Set(e),
      encode: set => [...set.values()]
    })),
    toArbitrary: ([value]) => (fc, ctx) => {
      return fc.oneof(ctx?.isSuspend ? {
        maxDepth: 2,
        depthIdentifier: "ReadonlySet"
      } : {}, fc.constant([]), fc.array(value, ctx?.constraints?.array)).map(as => new globalThis.Set(as));
    },
    toEquivalence: ([value]) => Equal.makeCompareSet(value),
    toFormatter: ([value]) => t => {
      const size = t.size;
      if (size === 0) {
        return "ReadonlySet(0) {}";
      }
      const values = globalThis.Array.from(t.values()).sort().map(v => `${value(v)}`);
      return `ReadonlySet(${size}) { ${values.join(", ")} }`;
    }
  });
  return make(schema.ast, {
    value
  });
}
/**
 * Creates a schema that validates a `HashSet` where values must conform to the
 * provided schema.
 *
 * @category HashSet
 * @since 4.0.0
 */
export function HashSet(value) {
  const schema = declareConstructor()([value], ([value]) => {
    const values = Array(value);
    return (input, ast, options) => {
      if (HashSet_.isHashSet(input)) {
        return Effect.mapBothEager(Parser.decodeUnknownEffect(values)(Arr.fromIterable(input), options), {
          onSuccess: HashSet_.fromIterable,
          onFailure: issue => new Issue.Composite(ast, Option_.some(input), [new Issue.Pointer(["values"], issue)])
        });
      }
      return Effect.fail(new Issue.InvalidType(ast, Option_.some(input)));
    };
  }, {
    typeConstructor: {
      _tag: "effect/HashSet"
    },
    generation: {
      runtime: `Schema.HashSet(?)`,
      Type: `HashSet.HashSet<?>`
    },
    expected: "HashSet",
    toCodec: ([value]) => link()(Array(value), Transformation.transform({
      decode: HashSet_.fromIterable,
      encode: Arr.fromIterable
    })),
    toArbitrary: ([value]) => (fc, ctx) => {
      return fc.oneof(ctx?.isSuspend ? {
        maxDepth: 2,
        depthIdentifier: "HashSet"
      } : {}, fc.constant([]), fc.array(value, ctx?.constraints?.array)).map(HashSet_.fromIterable);
    },
    toEquivalence: ([value]) => Equal.makeCompareSet(value),
    toFormatter: ([value]) => t => {
      const size = HashSet_.size(t);
      if (size === 0) {
        return "HashSet(0) {}";
      }
      const values = globalThis.Array.from(t).sort().map(v => `${value(v)}`);
      return `HashSet(${size}) { ${values.join(", ")} }`;
    }
  });
  return make(schema.ast, {
    value
  });
}
/**
 * Creates a schema that validates a `Chunk` where values must conform to the
 * provided schema.
 *
 * @category Chunk
 * @since 4.0.0
 */
export function Chunk(value) {
  const schema = declareConstructor()([value], ([value]) => {
    const values = Array(value);
    return (input, ast, options) => {
      if (Chunk_.isChunk(input)) {
        return Effect.mapBothEager(Parser.decodeUnknownEffect(values)(Arr.fromIterable(input), options), {
          onSuccess: Chunk_.fromIterable,
          onFailure: issue => new Issue.Composite(ast, Option_.some(input), [new Issue.Pointer(["values"], issue)])
        });
      }
      return Effect.fail(new Issue.InvalidType(ast, Option_.some(input)));
    };
  }, {
    typeConstructor: {
      _tag: "effect/Chunk"
    },
    generation: {
      runtime: `Schema.Chunk(?)`,
      Type: `Chunk.Chunk<?>`
    },
    expected: "Chunk",
    toCodec: ([value]) => link()(Array(value), Transformation.transform({
      decode: Chunk_.fromIterable,
      encode: Arr.fromIterable
    })),
    toArbitrary: ([value]) => (fc, ctx) => {
      return fc.oneof(ctx?.isSuspend ? {
        maxDepth: 2,
        depthIdentifier: "Chunk"
      } : {}, fc.constant([]), fc.array(value, ctx?.constraints?.array)).map(Chunk_.fromIterable);
    },
    toEquivalence: ([value]) => Chunk_.makeEquivalence(value),
    toFormatter: ([value]) => t => {
      const size = Chunk_.size(t);
      if (size === 0) {
        return "Chunk(0) {}";
      }
      const values = globalThis.Array.from(t).sort().map(v => `${value(v)}`);
      return `Chunk(${size}) { ${values.join(", ")} }`;
    }
  });
  return make(schema.ast, {
    value
  });
}
/**
 * @since 4.0.0
 */
export const RegExp = /*#__PURE__*/instanceOf(globalThis.RegExp, {
  typeConstructor: {
    _tag: "RegExp"
  },
  generation: {
    runtime: `Schema.RegExp`,
    Type: `globalThis.RegExp`
  },
  expected: "RegExp",
  toCodecJson: () => link()(Struct({
    source: String,
    flags: String
  }), Transformation.transformOrFail({
    decode: e => Effect.try({
      try: () => new globalThis.RegExp(e.source, e.flags),
      catch: e => new Issue.InvalidValue(Option_.some(e), {
        message: globalThis.String(e)
      })
    }),
    encode: regExp => Effect.succeed({
      source: regExp.source,
      flags: regExp.flags
    })
  })),
  toArbitrary: () => fc => fc.tuple(fc.constantFrom(".", ".*", "\\d+", "\\w+", "[a-z]+", "[A-Z]+", "[0-9]+", "^[a-zA-Z0-9]+$", "^\\d{4}-\\d{2}-\\d{2}$" // date pattern
  ), fc.uniqueArray(fc.constantFrom("g", "i", "m", "s", "u", "y"), {
    minLength: 0,
    maxLength: 6
  }).map(flags => flags.join(""))).map(([source, flags]) => new globalThis.RegExp(source, flags)),
  toEquivalence: () => (a, b) => a.source === b.source && a.flags === b.flags
});
/**
 * A schema for JavaScript `URL` objects.
 *
 * **Default JSON serializer**
 *
 * - encodes `URL` as a `string`
 *
 * @since 4.0.0
 * @category URL
 */
export const URL = /*#__PURE__*/instanceOf(globalThis.URL, {
  typeConstructor: {
    _tag: "URL"
  },
  generation: {
    runtime: `Schema.URL`,
    Type: `globalThis.URL`
  },
  expected: "URL",
  toCodecJson: () => link()(String.annotate({
    expected: "a string that will be decoded as a URL"
  }), Transformation.urlFromString),
  toArbitrary: () => fc => fc.webUrl().map(s => new globalThis.URL(s)),
  toEquivalence: () => (a, b) => a.toString() === b.toString()
});
/**
 * A transformation schema that decodes a `string` into a `URL`.
 *
 * Decoding:
 * - A **valid** URL `string` is decoded as a `URL`
 *
 * Encoding:
 * - A `URL` is encoded as a `string`
 *
 * @category URL
 * @since 4.0.0
 */
export const URLFromString = /*#__PURE__*/String.annotate({
  expected: "a string that will be decoded as a URL"
}).pipe(/*#__PURE__*/decodeTo(URL, Transformation.urlFromString)); // TODO: remove duplication with URL schema
/**
 * A schema for JavaScript `Date` objects.
 *
 * This schema accepts any `Date` instance, including invalid dates (e.g., `new
 * Date("invalid")`). For validating only valid dates, use `ValidDate` instead.
 *
 * @since 4.0.0
 */
export const Date = /*#__PURE__*/instanceOf(globalThis.Date, {
  typeConstructor: {
    _tag: "Date"
  },
  generation: {
    runtime: `Schema.Date`,
    Type: `globalThis.Date`
  },
  expected: "Date",
  toCodecJson: () => link()(String.annotate({
    expected: "a string in ISO 8601 format that will be decoded as a Date"
  }), Transformation.transform({
    decode: s => new globalThis.Date(s),
    encode: formatDate
  })),
  toArbitrary: () => (fc, ctx) => fc.date(ctx?.constraints?.date)
});
/**
 * A schema for **valid** JavaScript `Date` objects.
 *
 * This schema accepts `Date` instances but rejects invalid dates (such as `new
 * Date("invalid")`).
 *
 * @since 4.0.0
 */
export const DateValid = /*#__PURE__*/Date.check(/*#__PURE__*/isDateValid());
/**
 * A schema for `Duration` values.
 *
 * **Default JSON serializer**
 *
 * - encodes `Duration` as a `string`
 *
 * @since 4.0.0
 */
export const Duration = /*#__PURE__*/declare(Duration_.isDuration, {
  typeConstructor: {
    _tag: "effect/Duration"
  },
  generation: {
    runtime: `Schema.Duration`,
    Type: `Duration.Duration`,
    importDeclaration: `import * as Duration from "effect/Duration"`
  },
  expected: "Duration",
  toCodecJson: () => link()(Union([Struct({
    _tag: Literal("Infinity")
  }), Struct({
    _tag: Literal("NegativeInfinity")
  }), Struct({
    _tag: Literal("Nanos"),
    value: BigInt
  }), Struct({
    _tag: Literal("Millis"),
    value: Int
  })]), Transformation.transform({
    decode: e => {
      switch (e._tag) {
        case "Infinity":
          return Duration_.infinity;
        case "NegativeInfinity":
          return Duration_.negativeInfinity;
        case "Nanos":
          return Duration_.nanos(e.value);
        case "Millis":
          return Duration_.millis(e.value);
      }
    },
    encode: duration => {
      switch (duration.value._tag) {
        case "Infinity":
          return {
            _tag: "Infinity"
          };
        case "NegativeInfinity":
          return {
            _tag: "NegativeInfinity"
          };
        case "Nanos":
          return {
            _tag: "Nanos",
            value: duration.value.nanos
          };
        case "Millis":
          return {
            _tag: "Millis",
            value: duration.value.millis
          };
      }
    }
  })),
  toArbitrary: () => fc => fc.oneof(fc.constant(Duration_.infinity), fc.constant(Duration_.negativeInfinity), fc.bigInt().map(Duration_.nanos), fc.maxSafeInteger().map(Duration_.millis)),
  toFormatter: () => globalThis.String,
  toEquivalence: () => Duration_.Equivalence
});
/**
 * A transformation schema that decodes a non-negative `bigint` into a
 * `Duration`, treating the `bigint` value as the duration in nanoseconds.
 *
 * Decoding:
 * - A non-negative `bigint` representing nanoseconds is decoded as a `Duration`
 *
 * Encoding:
 * - A `Duration` is encoded to a non-negative `bigint` representing nanoseconds
 *
 * @category Duration
 * @since 4.0.0
 */
export const DurationFromNanos = /*#__PURE__*/BigInt.check(isGreaterThanOrEqualToBigInt(0n)).pipe(/*#__PURE__*/decodeTo(Duration, Transformation.durationFromNanos));
/**
 * A transformation schema that decodes a non-negative (possibly infinite)
 * integer into a `Duration`, treating the integer value as the duration in
 * milliseconds.
 *
 * Decoding:
 * - A non-negative (possibly infinite) integer representing milliseconds is
 *   decoded as a `Duration`
 *
 * Encoding:
 * - A `Duration` is encoded to a non-negative (possibly infinite) integer
 *   representing milliseconds
 *
 * @category Duration
 * @since 4.0.0
 */
export const DurationFromMillis = /*#__PURE__*/Number.check(isGreaterThanOrEqualTo(0)).pipe(/*#__PURE__*/decodeTo(Duration, Transformation.durationFromMillis));
/**
 * A schema for `BigDecimal` values.
 *
 * **Default JSON serializer**
 *
 * - encodes `BigDecimal` as a `string`
 *
 * @since 4.0.0
 */
export const BigDecimal = /*#__PURE__*/declare(BigDecimal_.isBigDecimal, {
  typeConstructor: {
    _tag: "effect/BigDecimal"
  },
  generation: {
    runtime: `Schema.BigDecimal`,
    Type: `BigDecimal.BigDecimal`,
    importDeclaration: `import * as BigDecimal from "effect/BigDecimal"`
  },
  expected: "BigDecimal",
  toCodecJson: () => link()(String.annotate({
    expected: "a string that will be decoded as a BigDecimal"
  }), Transformation.bigDecimalFromString),
  toArbitrary: () => fc => fc.tuple(fc.bigInt(), fc.integer({
    min: 0,
    max: 20
  })).map(([value, scale]) => BigDecimal_.make(value, scale)),
  toFormatter: () => bd => BigDecimal_.format(bd),
  toEquivalence: () => BigDecimal_.Equivalence
});
/**
 * A transformation schema that decodes a JSON-encoded string into an `unknown` value.
 *
 * Decoding:
 * - A `string` is decoded as an `unknown` value.
 * - If the string is not valid JSON, decoding fails.
 *
 * Encoding:
 * - Any value is encoded as a JSON string using `JSON.stringify`.
 * - If the value is not a valid JSON value, encoding fails.
 *
 * **Example**
 *
 * ```ts
 * import { Schema } from "effect"
 *
 * Schema.decodeUnknownSync(Schema.UnknownFromJsonString)(`{"a":1,"b":2}`)
 * // => { a: 1, b: 2 }
 * ```
 *
 * @since 4.0.0
 */
export const UnknownFromJsonString = /*#__PURE__*/fromJsonString(Unknown);
/**
 * Returns a schema that decodes a JSON string and then decodes the parsed value
 * using the given schema.
 *
 * This is useful when working with JSON-encoded strings where the actual
 * structure of the value is known and described by an existing schema.
 *
 * The resulting schema first parses the input string as JSON, and then runs the
 * provided schema on the parsed result.
 *
 * **Example**
 *
 * ```ts
 * import { Schema } from "effect"
 *
 * const schema = Schema.Struct({ a: Schema.Number })
 * const schemaFromJsonString = Schema.fromJsonString(schema)
 *
 * Schema.decodeUnknownSync(schemaFromJsonString)(`{"a":1,"b":2}`)
 * // => { a: 1 }
 * ```
 *
 * **Json Schema Generation**
 *
 * When using `fromJsonString` with `draft-2020-12` or `openApi3.1`, the
 * resulting schema will be a JSON Schema with a `contentSchema` property that
 * contains the JSON Schema for the given schema.
 *
 * **Example**
 *
 * ```ts
 * import { Schema } from "effect"
 *
 * const original = Schema.Struct({ a: Schema.String })
 * const schema = Schema.fromJsonString(original)
 *
 * const document = Schema.toJsonSchemaDocument(schema)
 *
 * console.log(JSON.stringify(document, null, 2))
 * // {
 * //   "source": "draft-2020-12",
 * //   "schema": {
 * //     "type": "string",
 * //     "contentMediaType": "application/json",
 * //     "contentSchema": {
 * //       "type": "object",
 * //       "properties": {
 * //         "a": {
 * //           "type": "string"
 * //         }
 * //       },
 * //       "required": [
 * //         "a"
 * //       ],
 * //       "additionalProperties": false
 * //     }
 * //   },
 * //   "definitions": {}
 * // }
 * ```
 *
 * @since 4.0.0
 */
export function fromJsonString(schema) {
  return String.annotate({
    expected: "a string that will be decoded as JSON",
    contentMediaType: "application/json",
    contentSchema: AST.toEncoded(schema.ast)
  }).pipe(decodeTo(schema, Transformation.fromJsonString));
}
/**
 * @since 4.0.0
 */
export const File = /*#__PURE__*/instanceOf(globalThis.File, {
  typeConstructor: {
    _tag: "File"
  },
  generation: {
    runtime: `Schema.File`,
    Type: `globalThis.File`
  },
  expected: "File",
  toCodecJson: () => link()(Struct({
    data: String.check(isBase64()),
    type: String,
    name: String,
    lastModified: Number
  }), Transformation.transformOrFail({
    decode: e => Result_.match(Encoding.decodeBase64(e.data), {
      onFailure: error => Effect.fail(new Issue.InvalidValue(Option_.some(e.data), {
        message: error.message
      })),
      onSuccess: bytes => {
        const buffer = new globalThis.Uint8Array(bytes);
        return Effect.succeed(new globalThis.File([buffer], e.name, {
          type: e.type,
          lastModified: e.lastModified
        }));
      }
    }),
    encode: file => Effect.tryPromise({
      try: async () => {
        const bytes = new globalThis.Uint8Array(await file.arrayBuffer());
        return {
          data: Encoding.encodeBase64(bytes),
          type: file.type,
          name: file.name,
          lastModified: file.lastModified
        };
      },
      catch: e => new Issue.InvalidValue(Option_.some(file), {
        message: globalThis.String(e)
      })
    })
  }))
});
/**
 * @since 4.0.0
 */
export const FormData = /*#__PURE__*/instanceOf(globalThis.FormData, {
  typeConstructor: {
    _tag: "FormData"
  },
  generation: {
    runtime: `Schema.FormData`,
    Type: `globalThis.FormData`
  },
  expected: "FormData",
  toCodecJson: () => link()(Array(Tuple([String, Union([Struct({
    _tag: tag("String"),
    value: String
  }), Struct({
    _tag: tag("File"),
    value: File
  })])])), Transformation.transformOrFail({
    decode: e => {
      const out = new globalThis.FormData();
      for (const [key, entry] of e) {
        out.append(key, entry.value);
      }
      return Effect.succeed(out);
    },
    encode: formData => {
      return Effect.succeed(globalThis.Array.from(formData.entries()).map(([key, value]) => {
        if (typeof value === "string") {
          return [key, {
            _tag: "String",
            value
          }];
        } else {
          return [key, {
            _tag: "File",
            value
          }];
        }
      }));
    }
  }))
});
/**
 * `Schema.fromFormData` returns a schema that reads a `FormData` instance,
 * converts it into a tree record using bracket notation, and then decodes the
 * resulting structure using the provided schema.
 *
 * The decoding process has two steps:
 *
 * 1. Parse `FormData` into a nested tree record.
 * 2. Decode the parsed value with the given schema.
 *
 * **Example** (Decoding a flat structure)
 *
 * ```ts
 * import { Schema } from "effect"
 *
 * const schema = Schema.fromFormData(
 *   Schema.Struct({
 *     a: Schema.String
 *   })
 * )
 *
 * const formData = new FormData()
 * formData.append("a", "1")
 * formData.append("b", "2")
 *
 * console.log(String(Schema.decodeUnknownExit(schema)(formData)))
 * // Success({"a":"1"})
 * ```
 *
 * You can express nested values using bracket notation.
 *
 * **Example** (Nested fields)
 *
 * ```ts
 * import { Schema } from "effect"
 *
 * const schema = Schema.fromFormData(
 *   Schema.Struct({
 *     a: Schema.String,
 *     b: Schema.Struct({
 *       c: Schema.String,
 *       d: Schema.String
 *     })
 *   })
 * )
 *
 * const formData = new FormData()
 * formData.append("a", "1")
 * formData.append("b[c]", "2")
 * formData.append("b[d]", "3")
 *
 * console.log(String(Schema.decodeUnknownExit(schema)(formData)))
 * // Success({"a":"1","b":{"c":"2","d":"3"}})
 * ```
 *
 * If you want to decode values that are not strings, use
 * `Schema.toCodecStringTree` with the `keepDeclarations: true` option.
 * This serializer preserves values such as numbers and `Blob` objects when
 * compatible with the schema.
 *
 * **Example** (Parsing non-string values)
 *
 * ```ts
 * import { Schema } from "effect"
 *
 * const schema = Schema.fromFormData(
 *   Schema.toCodecStringTree(
 *     Schema.Struct({
 *       a: Schema.Int
 *     }),
 *     { keepDeclarations: true }
 *   )
 * )
 *
 * const formData = new FormData()
 * formData.append("a", "1")
 *
 * console.log(String(Schema.decodeUnknownExit(schema)(formData)))
 * // Success({"a":1}) // Note: the value is a number
 * ```
 *
 * @since 4.0.0
 */
export function fromFormData(schema) {
  return FormData.pipe(decodeTo(schema, Transformation.fromFormData));
}
/**
 * @since 4.0.0
 */
export const URLSearchParams = /*#__PURE__*/instanceOf(globalThis.URLSearchParams, {
  typeConstructor: {
    _tag: "URLSearchParams"
  },
  generation: {
    runtime: `Schema.URLSearchParams`,
    Type: `globalThis.URLSearchParams`
  },
  expected: "URLSearchParams",
  toCodecJson: () => link()(String.annotate({
    expected: "a query string that will be decoded as URLSearchParams"
  }), Transformation.transform({
    decode: e => new globalThis.URLSearchParams(e),
    encode: params => params.toString()
  }))
});
/**
 * `Schema.fromURLSearchParams` returns a schema that reads a `URLSearchParams`
 * instance, converts it into a tree record using bracket notation, and then
 * decodes the resulting structure using the provided schema.
 *
 * The decoding process has two steps:
 *
 * 1. Parse `URLSearchParams` into a nested tree record.
 * 2. Decode the parsed value with the given schema.
 *
 * **Example** (Decoding a flat structure)
 *
 * ```ts
 * import { Schema } from "effect"
 *
 * const schema = Schema.fromURLSearchParams(
 *   Schema.Struct({
 *     a: Schema.String
 *   })
 * )
 *
 * const urlSearchParams = new URLSearchParams("a=1&b=2")
 *
 * console.log(String(Schema.decodeUnknownExit(schema)(urlSearchParams)))
 * // Success({"a":"1"})
 * ```
 * You can express nested values using bracket notation.
 *
 * **Example** (Nested fields)
 *
 * ```ts
 * import { Schema } from "effect"
 *
 * const schema = Schema.fromURLSearchParams(
 *   Schema.Struct({
 *     a: Schema.String,
 *     b: Schema.Struct({
 *       c: Schema.String,
 *       d: Schema.String
 *     })
 *   })
 * )
 *
 * const urlSearchParams = new URLSearchParams("a=1&b[c]=2&b[d]=3")
 *
 * console.log(String(Schema.decodeUnknownExit(schema)(urlSearchParams)))
 * // Success({"a":"1","b":{"c":"2","d":"3"}})
 * ```
 *
 * If you want to decode values that are not strings, use
 * `Schema.toCodecStringTree`. This serializer preserves values such as
 * numbers when compatible with the schema.
 *
 * **Example** (Parsing non-string values)
 *
 * ```ts
 * import { Schema } from "effect"
 *
 * const schema = Schema.fromURLSearchParams(
 *   Schema.toCodecStringTree(
 *     Schema.Struct({
 *       a: Schema.Int
 *     })
 *   )
 * )
 *
 * const urlSearchParams = new URLSearchParams("a=1&b=2")
 *
 * console.log(String(Schema.decodeUnknownExit(schema)(urlSearchParams)))
 * // Success({"a":1}) // Note: the value is a number
 * ```
 *
 * @since 4.0.0
 */
export function fromURLSearchParams(schema) {
  return URLSearchParams.pipe(decodeTo(schema, Transformation.fromURLSearchParams));
}
/**
 * A schema for finite numbers, rejecting `NaN`, `Infinity`, and `-Infinity`.
 *
 * @since 4.0.0
 */
export const Finite = /*#__PURE__*/Number.check(/*#__PURE__*/isFinite());
/**
 * A schema for integers, rejecting `NaN`, `Infinity`, and `-Infinity`.
 *
 * @since 4.0.0
 */
export const Int = /*#__PURE__*/Number.check(/*#__PURE__*/isInt());
/**
 * A transformation schema that parses a string into a number.
 *
 * Decoding:
 * - A `string` is decoded as a finite number.
 *
 * Encoding:
 * - A number is encoded as a `string`.
 *
 * @since 4.0.0
 */
export const NumberFromString = /*#__PURE__*/String.annotate({
  expected: "a string that will be decoded as a number"
}).pipe(/*#__PURE__*/decodeTo(Number, Transformation.numberFromString));
/**
 * A transformation schema that parses a string into a finite number.
 *
 * Decoding:
 * - A `string` is decoded as a finite number, rejecting `NaN`, `Infinity`, and
 *   `-Infinity` values.
 *
 * Encoding:
 * - A finite number is encoded as a `string`.
 *
 * @since 4.0.0
 */
export const FiniteFromString = /*#__PURE__*/String.annotate({
  expected: "a string that will be decoded as a finite number"
}).pipe(/*#__PURE__*/decodeTo(Finite, Transformation.numberFromString));
/**
 * A schema for strings that contains no leading or trailing whitespaces.
 *
 * @since 4.0.0
 */
export const Trimmed = /*#__PURE__*/String.check(/*#__PURE__*/isTrimmed());
/**
 * A transformation schema that trims whitespace from a string.
 *
 * Decoding:
 * - A `string` is decoded as a string with no leading or trailing whitespaces.
 *
 * Encoding:
 * - The trimmed string is encoded as is.
 *
 * @since 4.0.0
 */
export const Trim = /*#__PURE__*/String.annotate({
  expected: "a string that will be decoded as a trimmed string"
}).pipe(/*#__PURE__*/decodeTo(Trimmed, /*#__PURE__*/Transformation.trim()));
/**
 * @since 4.0.0
 */
export const PropertyKey = /*#__PURE__*/Union([Finite, Symbol, String]);
/**
 * @since 4.0.0
 */
export const StandardSchemaV1FailureResult = /*#__PURE__*/Struct({
  issues: /*#__PURE__*/Array(/*#__PURE__*/Struct({
    message: String,
    path: /*#__PURE__*/optional(/*#__PURE__*/Array(/*#__PURE__*/Union([PropertyKey, /*#__PURE__*/Struct({
      key: PropertyKey
    })])))
  }))
});
/**
 * A boolean parsed from 0 or 1.
 *
 * @category Boolean
 * @since 4.0.0
 */
export const BooleanFromBit = /*#__PURE__*/Literals([0, 1]).pipe(/*#__PURE__*/decodeTo(Boolean, /*#__PURE__*/Transformation.transform({
  decode: bit => bit === 1,
  encode: bool => bool ? 1 : 0
})));
const Base64String = /*#__PURE__*/String.annotate({
  expected: "a base64 encoded string that will be decoded as Uint8Array",
  format: "byte",
  contentEncoding: "base64"
});
/**
 * A schema for JavaScript `Uint8Array` objects.
 *
 * **Default JSON serializer**
 *
 * The default JSON serializer encodes Uint8Array as a Base64 encoded string.
 *
 * @category Uint8Array
 * @since 4.0.0
 */
export const Uint8Array = /*#__PURE__*/instanceOf(globalThis.Uint8Array, {
  typeConstructor: {
    _tag: "Uint8Array"
  },
  generation: {
    runtime: `Schema.Uint8Array`,
    Type: `globalThis.Uint8Array`
  },
  expected: "Uint8Array",
  toCodecJson: () => link()(Base64String, Transformation.uint8ArrayFromBase64String),
  toArbitrary: () => fc => fc.uint8Array()
});
/**
 * A transformation schema that decodes a base64 encoded string into a
 * `Uint8Array`.
 *
 * Decoding:
 * - A **valid** base64 encoded string is decoded as a `Uint8Array`.
 *
 * Encoding:
 * - A `Uint8Array` is encoded as a base64-encoded string.
 *
 * @category Uint8Array
 * @since 4.0.0
 */
export const Uint8ArrayFromBase64 = /*#__PURE__*/Base64String.pipe(/*#__PURE__*/decodeTo(Uint8Array, Transformation.uint8ArrayFromBase64String));
/**
 * A transformation schema that decodes a base64 (URL) encoded string into a
 * `Uint8Array`.
 *
 * Decoding:
 * - A **valid** base64 (URL) encoded string is decoded as a `Uint8Array`.
 *
 * Encoding:
 * - A `Uint8Array` is encoded as a base64 (URL) encoded string.
 *
 * @category Uint8Array
 * @since 4.0.0
 */
export const Uint8ArrayFromBase64Url = /*#__PURE__*/String.annotate({
  expected: "a base64 (URL) encoded string that will be decoded as a Uint8Array"
}).pipe(/*#__PURE__*/decodeTo(Uint8Array, {
  decode: /*#__PURE__*/Getter.decodeBase64Url(),
  encode: /*#__PURE__*/Getter.encodeBase64Url()
}));
/**
 * A transformation schema that decodes a hex encoded string into a
 * `Uint8Array`.
 *
 * Decoding:
 * - A **valid** hex encoded string is decoded as a `Uint8Array`.
 *
 * Encoding:
 * - A `Uint8Array` is encoded as a hex encoded string.
 *
 * @category Uint8Array
 * @since 4.0.0
 */
export const Uint8ArrayFromHex = /*#__PURE__*/String.annotate({
  expected: "a hex encoded string that will be decoded as a Uint8Array"
}).pipe(/*#__PURE__*/decodeTo(Uint8Array, {
  decode: /*#__PURE__*/Getter.decodeHex(),
  encode: /*#__PURE__*/Getter.encodeHex()
}));
/**
 * A schema for `DateTime.Utc` values.
 *
 * **Default JSON serializer**
 *
 * - encodes `DateTime.Utc` as a UTC ISO string
 *
 * @category DateTime
 * @since 4.0.0
 */
export const DateTimeUtc = /*#__PURE__*/declare(u => DateTime.isDateTime(u) && DateTime.isUtc(u), {
  typeConstructor: {
    _tag: "effect/DateTime.Utc"
  },
  generation: {
    runtime: `Schema.DateTimeUtc`,
    Type: `DateTime.Utc`,
    importDeclaration: `import * as DateTime from "effect/DateTime"`
  },
  expected: "DateTime.Utc",
  toCodecJson: () => link()(String, Transformation.dateTimeUtcFromString),
  toArbitrary: () => (fc, ctx) => fc.date({
    noInvalidDate: true,
    ...ctx?.constraints?.date
  }).map(date => DateTime.fromDateUnsafe(date)),
  toFormatter: () => utc => utc.toString(),
  toEquivalence: () => DateTime.Equivalence
});
/**
 * A transformation schema that decodes a `Date` into a `DateTime.Utc`.
 *
 * Decoding:
 * - A **valid** `Date` is decoded as a `DateTime.Utc`
 *
 * Encoding:
 * - A `DateTime.Utc` is encoded as a `Date`
 *
 * @category DateTime
 * @since 4.0.0
 */
export const DateTimeUtcFromDate = /*#__PURE__*/DateValid.pipe(/*#__PURE__*/decodeTo(DateTimeUtc, {
  decode: /*#__PURE__*/Getter.dateTimeUtcFromInput(),
  encode: /*#__PURE__*/Getter.transform(DateTime.toDateUtc)
}));
/**
 * A transformation schema that decodes a string into a `DateTime.Utc`.
 *
 * Decoding:
 * - A `string` that can be parsed by `Date.parse` is decoded as a
 *   `DateTime.Utc`
 *
 * Encoding:
 * - A `DateTime.Utc` is encoded as a `string` in ISO 8601 format, ignoring any
 *   time zone.
 *
 * @category DateTime
 * @since 4.0.0
 */
export const DateTimeUtcFromString = /*#__PURE__*/String.annotate({
  expected: "a string that will be decoded as a DateTime.Utc"
}).pipe(/*#__PURE__*/decodeTo(DateTimeUtc, Transformation.dateTimeUtcFromString));
/**
 * A transformation schema that decodes a number into a `DateTime.Utc`.
 *
 * Decoding:
 * - A number of milliseconds since the Unix epoch is decoded as a `DateTime.Utc`
 *
 * Encoding:
 * - A `DateTime.Utc` is encoded as a number of milliseconds since the Unix epoch.
 *
 * @category DateTime
 * @since 4.0.0
 */
export const DateTimeUtcFromMillis = /*#__PURE__*/Number.pipe(/*#__PURE__*/decodeTo(DateTimeUtc, {
  decode: /*#__PURE__*/Getter.dateTimeUtcFromInput(),
  encode: /*#__PURE__*/Getter.transform(DateTime.toEpochMillis)
}));
/**
 * A schema for `DateTime.TimeZone.Offset` values.
 *
 * **Default JSON serializer**
 *
 * - encodes `DateTime.TimeZone.Offset` as a number (offset in milliseconds)
 *
 * @category DateTime
 * @since 4.0.0
 */
export const TimeZoneOffset = /*#__PURE__*/declare(DateTime.isTimeZoneOffset, {
  typeConstructor: {
    _tag: "effect/DateTime.TimeZone.Offset"
  },
  generation: {
    runtime: `Schema.TimeZoneOffset`,
    Type: `DateTime.TimeZone.Offset`,
    importDeclaration: `import * as DateTime from "effect/DateTime"`
  },
  expected: "DateTime.TimeZone.Offset",
  toCodecJson: () => link()(Number, Transformation.timeZoneOffsetFromNumber),
  toArbitrary: () => fc => fc.integer({
    min: -12 * 60 * 60 * 1000,
    max: 14 * 60 * 60 * 1000
  }).map(n => DateTime.zoneMakeOffset(n)),
  toFormatter: () => tz => DateTime.zoneToString(tz),
  toEquivalence: () => (a, b) => a.offset === b.offset
});
/**
 * A schema for `DateTime.TimeZone.Named` values.
 *
 * **Default JSON serializer**
 *
 * - encodes `DateTime.TimeZone.Named` as a string (IANA time zone identifier)
 *
 * @category DateTime
 * @since 4.0.0
 */
export const TimeZoneNamed = /*#__PURE__*/declare(DateTime.isTimeZoneNamed, {
  typeConstructor: {
    _tag: "effect/DateTime.TimeZone.Named"
  },
  generation: {
    runtime: `Schema.TimeZoneNamed`,
    Type: `DateTime.TimeZone.Named`,
    importDeclaration: `import * as DateTime from "effect/DateTime"`
  },
  expected: "DateTime.TimeZone.Named",
  toCodecJson: () => link()(String.annotate({
    expected: "an IANA time zone identifier"
  }), Transformation.timeZoneNamedFromString),
  toArbitrary: () => fc => fc.constantFrom(...["UTC", "Europe/London", "America/New_York", "Asia/Tokyo", "Australia/Sydney"].map(DateTime.zoneMakeNamedUnsafe)),
  toFormatter: () => tz => DateTime.zoneToString(tz),
  toEquivalence: () => (a, b) => a.id === b.id
});
/**
 * A schema for `DateTime.TimeZone` values.
 *
 * **Default JSON serializer**
 *
 * - encodes `DateTime.TimeZone` as a string (IANA identifier or offset like
 *   `+03:00`)
 *
 * @category DateTime
 * @since 4.0.0
 */
export const TimeZone = /*#__PURE__*/declare(DateTime.isTimeZone, {
  typeConstructor: {
    _tag: "effect/DateTime.TimeZone"
  },
  generation: {
    runtime: `Schema.TimeZone`,
    Type: `DateTime.TimeZone`,
    importDeclaration: `import * as DateTime from "effect/DateTime"`
  },
  expected: "DateTime.TimeZone",
  toCodecJson: () => link()(String.annotate({
    expected: "a time zone string (IANA identifier or offset like +03:00)"
  }), Transformation.timeZoneFromString),
  toArbitrary: () => fc => fc.oneof(fc.integer({
    min: -12 * 60 * 60 * 1000,
    max: 14 * 60 * 60 * 1000
  }).map(n => DateTime.zoneMakeOffset(n)), fc.constantFrom(...["UTC", "Europe/London", "America/New_York", "Asia/Tokyo", "Australia/Sydney"].map(DateTime.zoneMakeNamedUnsafe))),
  toFormatter: () => tz => DateTime.zoneToString(tz),
  toEquivalence: () => (a, b) => DateTime.zoneToString(a) === DateTime.zoneToString(b)
});
/**
 * A schema for `DateTime.Zoned` values.
 *
 * **Default JSON serializer**
 *
 * - encodes `DateTime.Zoned` as a string in the format
 *   `YYYY-MM-DDTHH:mm:ss.sss+HH:MM[Time/Zone]`
 *
 * @category DateTime
 * @since 4.0.0
 */
export const DateTimeZoned = /*#__PURE__*/declare(u => DateTime.isDateTime(u) && DateTime.isZoned(u), {
  typeConstructor: {
    _tag: "effect/DateTime.Zoned"
  },
  generation: {
    runtime: `Schema.DateTimeZoned`,
    Type: `DateTime.Zoned`,
    importDeclaration: `import * as DateTime from "effect/DateTime"`
  },
  expected: "DateTime.Zoned",
  toCodecJson: () => link()(String.annotate({
    expected: "a zoned DateTime string (e.g. 2024-01-01T00:00:00.000+00:00[Europe/London])"
  }), Transformation.dateTimeZonedFromString),
  toArbitrary: () => (fc, ctx) => fc.tuple(fc.date({
    noInvalidDate: true,
    min: new globalThis.Date(-8640000000000000 + 14 * 60 * 60 * 1000),
    max: new globalThis.Date(8640000000000000 - 14 * 60 * 60 * 1000),
    ...ctx?.constraints?.date
  }), fc.constantFrom("UTC", "Europe/London", "America/New_York", "Asia/Tokyo", "Australia/Sydney")).map(([date, zone]) => DateTime.makeZonedUnsafe(date, {
    timeZone: zone
  })),
  toFormatter: () => zoned => DateTime.formatIsoZoned(zoned),
  toEquivalence: () => DateTime.Equivalence
});
const immerable = /*#__PURE__*/globalThis.Symbol.for("immer-draftable");
function makeClass(Inherited, identifier, struct, annotations) {
  const getClassSchema = getClassSchemaFactory(struct, identifier, annotations);
  const ClassTypeId = getClassTypeId(identifier); // HMR support
  return class extends Inherited {
    constructor(...[input, options]) {
      const props = input ?? {};
      if (options?.disableValidation) {
        super(props, options);
      } else {
        const validated = struct.makeUnsafe(props, options);
        super({
          ...props,
          ...validated
        }, {
          ...options,
          disableValidation: true
        });
      }
    }
    toString() {
      return `${identifier}(${format({
        ...this
      })})`;
    }
    static [TypeId] = TypeId;
    get [ClassTypeId]() {
      return ClassTypeId;
    }
    static [immerable] = true;
    static identifier = identifier;
    static fields = struct.fields;
    static get ast() {
      return getClassSchema(this).ast;
    }
    static pipe() {
      return Pipeable.pipeArguments(this, arguments);
    }
    static rebuild(ast) {
      return getClassSchema(this).rebuild(ast);
    }
    static makeUnsafe(input, options) {
      return new this(input, options);
    }
    static makeOption(input, options) {
      return Parser.makeOption(getClassSchema(this))(input, options);
    }
    static annotate(annotations) {
      return this.rebuild(AST.annotate(this.ast, annotations));
    }
    static annotateKey(annotations) {
      return this.rebuild(AST.annotateKey(this.ast, annotations));
    }
    static check(...checks) {
      return this.rebuild(AST.appendChecks(this.ast, checks));
    }
    static extend(identifier) {
      return (newFields, annotations) => {
        const fields = {
          ...struct.fields,
          ...newFields
        };
        return makeClass(this, identifier, makeStruct(AST.struct(fields, struct.ast.checks, {
          identifier
        }), fields), annotations);
      };
    }
    static mapFields(f, options) {
      return struct.mapFields(f, options);
    }
  };
}
function getClassTransformation(self) {
  return new Transformation.Transformation(Getter.transform(input => new self(input)), Getter.passthrough());
}
function getClassTypeId(identifier) {
  return `~effect/Schema/Class/${identifier}`;
}
function getClassSchemaFactory(from, identifier, annotations) {
  let memo;
  return self => {
    if (memo === undefined) {
      const transformation = getClassTransformation(self);
      const to = make(new AST.Declaration([from.ast], () => (input, ast) => {
        return input instanceof self || Predicate.hasProperty(input, getClassTypeId(identifier)) ? Effect.succeed(input) : Effect.fail(new Issue.InvalidType(ast, Option_.some(input)));
      }, {
        identifier,
        [AST.ClassTypeId]: ([from]) => new AST.Link(from, transformation),
        toCodec: ([from]) => new AST.Link(from.ast, transformation),
        toArbitrary: ([from]) => () => from.map(args => new self(args)),
        toFormatter: ([from]) => t => `${self.identifier}(${from(t)})`,
        "~sentinels": AST.collectSentinels(from.ast),
        ...annotations
      }));
      memo = from.pipe(decodeTo(to, transformation));
    }
    return memo;
  };
}
function isStruct(schema) {
  return isSchema(schema);
}
/**
 * @category Constructors
 * @since 4.0.0
 */
export const Class = identifier => (schema, annotations) => {
  const struct = isStruct(schema) ? schema : Struct(schema);
  return makeClass(Data.Class, identifier, struct, annotations);
};
/**
 * @category Constructors
 * @since 4.0.0
 */
export const TaggedClass = identifier => {
  return (tagValue, schema, annotations) => {
    return Class(identifier ?? tagValue)(isStruct(schema) ? schema.mapFields(fields => ({
      _tag: tag(tagValue),
      ...fields
    }), {
      unsafePreserveChecks: true
    }) : TaggedStruct(tagValue, schema), annotations);
  };
};
/**
 * @category Constructors
 * @since 4.0.0
 */
export const ErrorClass = identifier => (schema, annotations) => {
  const struct = isStruct(schema) ? schema : Struct(schema);
  return makeClass(core.Error, identifier, struct, annotations);
};
/**
 * @category Constructors
 * @since 4.0.0
 */
export const TaggedErrorClass = identifier => {
  return (tagValue, schema, annotations) => {
    return ErrorClass(identifier ?? tagValue)(isStruct(schema) ? schema.mapFields(fields => ({
      _tag: tag(tagValue),
      ...fields
    }), {
      unsafePreserveChecks: true
    }) : TaggedStruct(tagValue, schema), annotations);
  };
};
/**
 * @category Arbitrary
 * @since 4.0.0
 */
export function toArbitraryLazy(schema) {
  const lawc = InternalArbitrary.memoized(schema.ast);
  return fc => lawc(fc, {});
}
/**
 * @category Arbitrary
 * @since 4.0.0
 */
export function toArbitrary(schema) {
  return toArbitraryLazy(schema)(FastCheck);
}
// -----------------------------------------------------------------------------
// Formatter
// -----------------------------------------------------------------------------
/**
 * **Technical Note**
 *
 * This annotation cannot be added to `Annotations.Bottom` because it would make
 * the schema invariant.
 *
 * @category Formatter
 * @since 4.0.0
 */
export function overrideToFormatter(toFormatter) {
  return self => {
    return self.annotate({
      toFormatter
    });
  };
}
/**
 * @category Formatter
 * @since 4.0.0
 */
export function toFormatter(schema, options) {
  return recur(schema.ast);
  function recur(ast) {
    // ---------------------------------------------
    // handle annotation
    // ---------------------------------------------
    const annotation = InternalAnnotations.resolve(ast)?.["toFormatter"];
    if (typeof annotation === "function") {
      return annotation(AST.isDeclaration(ast) ? ast.typeParameters.map(recur) : []);
    }
    // ---------------------------------------------
    // handle onBefore
    // ---------------------------------------------
    if (options?.onBefore) {
      const onBefore = options.onBefore(ast, recur);
      if (onBefore !== undefined) {
        return onBefore;
      }
    }
    // ---------------------------------------------
    // handle base case
    // ---------------------------------------------
    return on(ast);
  }
  function on(ast) {
    switch (ast._tag) {
      default:
        return format;
      case "Never":
        return () => "never";
      case "Void":
        return () => "void";
      case "Arrays":
        {
          const elements = ast.elements.map(recur);
          const rest = ast.rest.map(recur);
          return t => {
            const out = [];
            let i = 0;
            // ---------------------------------------------
            // handle elements
            // ---------------------------------------------
            for (; i < elements.length; i++) {
              if (t.length < i + 1) {
                if (AST.isOptional(ast.elements[i])) {
                  continue;
                }
              } else {
                out.push(elements[i](t[i]));
              }
            }
            // ---------------------------------------------
            // handle rest element
            // ---------------------------------------------
            if (rest.length > 0) {
              const [head, ...tail] = rest;
              for (; i < t.length - tail.length; i++) {
                out.push(head(t[i]));
              }
              // ---------------------------------------------
              // handle post rest elements
              // ---------------------------------------------
              for (let j = 0; j < tail.length; j++) {
                i += j;
                out.push(tail[j](t[i]));
              }
            }
            return "[" + out.join(", ") + "]";
          };
        }
      case "Objects":
        {
          const propertySignatures = ast.propertySignatures.map(ps => recur(ps.type));
          const indexSignatures = ast.indexSignatures.map(is => recur(is.type));
          if (ast.propertySignatures.length === 0 && ast.indexSignatures.length === 0) {
            return format;
          }
          return t => {
            const out = [];
            const visited = new Set();
            // ---------------------------------------------
            // handle property signatures
            // ---------------------------------------------
            for (let i = 0; i < propertySignatures.length; i++) {
              const ps = ast.propertySignatures[i];
              const name = ps.name;
              visited.add(name);
              if (AST.isOptional(ps.type) && !Object.hasOwn(t, name)) {
                continue;
              }
              out.push(`${formatPropertyKey(name)}: ${propertySignatures[i](t[name])}`);
            }
            // ---------------------------------------------
            // handle index signatures
            // ---------------------------------------------
            for (let i = 0; i < indexSignatures.length; i++) {
              const keys = AST.getIndexSignatureKeys(t, ast.indexSignatures[i].parameter);
              for (const key of keys) {
                if (visited.has(key)) {
                  continue;
                }
                visited.add(key);
                out.push(`${formatPropertyKey(key)}: ${indexSignatures[i](t[key])}`);
              }
            }
            return out.length > 0 ? "{ " + out.join(", ") + " }" : "{}";
          };
        }
      case "Union":
        {
          const getCandidates = t => AST.getCandidates(t, ast.types);
          return t => {
            const candidates = getCandidates(t);
            const refinements = candidates.map(Parser._is);
            for (let i = 0; i < candidates.length; i++) {
              const is = refinements[i];
              if (is(t)) {
                return recur(candidates[i])(t);
              }
            }
            return format(t);
          };
        }
      case "Suspend":
        {
          const get = AST.memoizeThunk(() => recur(ast.thunk()));
          return t => get()(t);
        }
    }
  }
}
// -----------------------------------------------------------------------------
// Equivalence
// -----------------------------------------------------------------------------
/**
 * **Technical Note**
 *
 * This annotation cannot be added to `Annotations.Bottom` because it would make
 * the schema invariant.
 *
 * @category Equivalence
 * @since 4.0.0
 */
export function overrideToEquivalence(toEquivalence) {
  return self => self.annotate({
    toEquivalence
  });
}
/**
 * @category Equivalence
 * @since 4.0.0
 */
export function toEquivalence(schema) {
  return InternalEquivalence.toEquivalence(schema.ast);
}
// -----------------------------------------------------------------------------
// Representation
// -----------------------------------------------------------------------------
/**
 * @category Representation
 * @since 4.0.0
 */
export function toRepresentation(schema) {
  return InternalStandard.fromAST(schema.ast);
}
/**
 * Returns a JSON Schema Document (draft-2020-12).
 *
 * You can use the `options` parameter to return a different target JSON Schema.
 *
 * @category JsonSchema
 * @since 4.0.0
 */
export function toJsonSchemaDocument(schema, options) {
  const sd = toRepresentation(schema);
  const jd = InternalStandard.toJsonSchemaDocument(sd, options);
  return {
    dialect: "draft-2020-12",
    schema: jd.schema,
    definitions: jd.definitions
  };
}
// -----------------------------------------------------------------------------
// Canonical Codecs
// -----------------------------------------------------------------------------
/**
 * @category Canonical Codecs
 * @since 4.0.0
 */
export function toCodecJson(schema) {
  return make(InternalToCodec.toCodecJson(schema.ast));
}
/**
 * @category Canonical Codecs
 * @since 4.0.0
 */
export function toCodecIso(schema) {
  return make(InternalToCodec.toCodecIso(AST.toType(schema.ast)));
}
export function toCodecStringTree(schema, options) {
  if (options?.keepDeclarations === true) {
    return make(toCodecEnsureArray(serializerStringTreeKeepDeclarations(schema.ast)));
  } else {
    return make(toCodecEnsureArray(serializerStringTree(schema.ast)));
  }
}
/**
 * @category Canonical Codecs
 * @since 4.0.0
 */
export function toEncoderXml(codec, options) {
  const rootName = InternalAnnotations.resolveIdentifier(codec.ast) ?? InternalAnnotations.resolveTitle(codec.ast);
  const serialize = encodeEffect(toCodecStringTree(codec));
  return t => serialize(t).pipe(Effect.map(stringTree => stringTreeToXml(stringTree, {
    rootName,
    ...options
  })));
}
function stringTreeToXml(value, options) {
  const rootName = options.rootName ?? "root";
  const arrayItemName = options.arrayItemName ?? "item";
  const pretty = options.pretty ?? true;
  const indent = options.indent ?? "  ";
  const sortKeys = options.sortKeys ?? true;
  const seen = new Set();
  const lines = [];
  recur(rootName, value, 0);
  return lines.join(pretty ? "\n" : "");
  function push(depth, text) {
    lines.push(pretty ? indent.repeat(depth) + text : text);
  }
  function recur(tagName, node, depth, originalNameForMeta) {
    const {
      attrs,
      safe
    } = xml.tagInfo(tagName, originalNameForMeta);
    if (node === undefined) {
      push(depth, `<${safe}${attrs}/>`);
    } else if (typeof node === "string") {
      push(depth, `<${safe}${attrs}>${xml.escapeText(node)}</${safe}>`);
    } else if (typeof node !== "object" || node === null) {
      push(depth, `<${safe}${attrs}>${xml.escapeText(format(node))}</${safe}>`);
    } else {
      if (seen.has(node)) throw new globalThis.Error("Cycle detected while serializing to XML.", {
        cause: node
      });
      seen.add(node);
      try {
        if (globalThis.globalThis.Array.isArray(node)) {
          if (node.length === 0) {
            push(depth, `<${safe}${attrs}/>`);
            return;
          }
          push(depth, `<${safe}${attrs}>`);
          for (const item of node) recur(arrayItemName, item, depth + 1);
          push(depth, `</${safe}>`);
          return;
        }
        const obj = node;
        const keys = Object.keys(obj);
        if (sortKeys) keys.sort();
        if (keys.length === 0) {
          push(depth, `<${safe}${attrs}/>`);
          return;
        }
        push(depth, `<${safe}${attrs}>`);
        for (const k of keys) {
          recur(xml.parseTagName(k).safe, obj[k], depth + 1, k);
        }
        push(depth, `</${safe}>`);
      } finally {
        seen.delete(node);
      }
    }
  }
}
const xml = {
  escapeText(s) {
    return s.replace(/&/g, "&amp;").replace(/</g, "&lt;").replace(/>/g, "&gt;");
  },
  escapeAttribute(s) {
    return s.replace(/&/g, "&amp;").replace(/"/g, "&quot;").replace(/</g, "&lt;").replace(/>/g, "&gt;");
  },
  parseTagName(name) {
    const original = name;
    let safe = name;
    if (!/^[A-Za-z_]/.test(safe)) safe = "_" + safe;
    safe = safe.replace(/[^A-Za-z0-9._-]/g, "_");
    if (/^xml/i.test(safe)) safe = "_" + safe;
    return {
      safe,
      changed: safe !== original
    };
  },
  tagInfo(name, original) {
    const {
      changed,
      safe
    } = xml.parseTagName(name);
    const needsMeta = changed || original && original !== name;
    const attrs = needsMeta ? ` data-name="${xml.escapeAttribute(original ?? name)}"` : "";
    return {
      safe,
      attrs
    };
  }
};
function getStringTreePriority(ast) {
  switch (ast._tag) {
    case "Null":
    case "Boolean":
    case "Number":
    case "BigInt":
    case "Symbol":
    case "UniqueSymbol":
      return 0;
    default:
      return 1;
  }
}
const treeReorder = /*#__PURE__*/InternalToCodec.makeReorder(getStringTreePriority);
function serializerTree(ast, recur, onMissingAnnotation) {
  switch (ast._tag) {
    case "Declaration":
      {
        const getLink = ast.annotations?.toCodecJson ?? ast.annotations?.toCodec;
        if (Predicate.isFunction(getLink)) {
          const tps = AST.isDeclaration(ast) ? ast.typeParameters.map(tp => make(recur(AST.toEncoded(tp)))) : [];
          const link = getLink(tps);
          const to = recur(link.to);
          return AST.replaceEncoding(ast, to === link.to ? [link] : [new AST.Link(to, link.transformation)]);
        }
        return onMissingAnnotation(ast);
      }
    case "Null":
      return AST.replaceEncoding(ast, [nullToString]);
    case "Boolean":
      return AST.replaceEncoding(ast, [booleanToString]);
    case "Unknown":
    case "ObjectKeyword":
      return AST.replaceEncoding(ast, [AST.unknownToStringTree]);
    case "Enum":
    case "Number":
    case "Literal":
    case "UniqueSymbol":
    case "Symbol":
    case "BigInt":
      return ast.toCodecStringTree();
    case "Objects":
      {
        if (ast.propertySignatures.some(ps => typeof ps.name !== "string")) {
          throw new globalThis.Error("Objects property names must be strings", {
            cause: ast
          });
        }
        return ast.recur(recur);
      }
    case "Union":
      {
        const sortedTypes = treeReorder(ast.types);
        if (sortedTypes !== ast.types) {
          return new AST.Union(sortedTypes, ast.mode, ast.annotations, ast.checks, ast.encoding, ast.context).recur(recur);
        }
        return ast.recur(recur);
      }
    case "Arrays":
    case "Suspend":
      return ast.recur(recur);
  }
  // `Schema.Any` is used as an escape hatch
  return ast;
}
const nullToString = /*#__PURE__*/new AST.Link(/*#__PURE__*/new AST.Literal("null"), /*#__PURE__*/new Transformation.Transformation(/*#__PURE__*/Getter.transform(() => null), /*#__PURE__*/Getter.transform(() => "null")));
const booleanToString = /*#__PURE__*/new AST.Link(/*#__PURE__*/new AST.Union([/*#__PURE__*/new AST.Literal("true"), /*#__PURE__*/new AST.Literal("false")], "anyOf"), /*#__PURE__*/new Transformation.Transformation(/*#__PURE__*/Getter.transform(s => s === "true"), /*#__PURE__*/Getter.String()));
const serializerStringTree = /*#__PURE__*/AST.toCodec(ast => {
  const out = serializerTree(ast, serializerStringTree, ast => AST.replaceEncoding(ast, [unknownToUndefined]));
  if (out !== ast && AST.isOptional(ast)) {
    return AST.optionalKeyLastLink(out);
  }
  return out;
});
const unknownToUndefined = /*#__PURE__*/new AST.Link(AST.undefined, /*#__PURE__*/new Transformation.Transformation(/*#__PURE__*/Getter.passthrough(), /*#__PURE__*/Getter.transform(() => undefined)));
const serializerStringTreeKeepDeclarations = /*#__PURE__*/AST.toCodec(ast => {
  const out = serializerTree(ast, serializerStringTreeKeepDeclarations, identity);
  if (out !== ast && AST.isOptional(ast)) {
    return AST.optionalKeyLastLink(out);
  }
  return out;
});
const SERIALIZER_ENSURE_ARRAY = "~effect/Schema/SERIALIZER_ENSURE_ARRAY";
const toCodecEnsureArray = /*#__PURE__*/AST.toCodec(ast => {
  if (AST.isUnion(ast) && ast.annotations?.[SERIALIZER_ENSURE_ARRAY]) {
    return ast;
  }
  const out = onSerializerEnsureArray(ast);
  if (AST.isArrays(out)) {
    const ensure = new AST.Union([out, AST.decodeTo(AST.string, out, new Transformation.Transformation(Getter.split(), Getter.passthrough()))], "anyOf", {
      [SERIALIZER_ENSURE_ARRAY]: true
    });
    return AST.isOptional(ast) ? AST.optionalKey(ensure) : ensure;
  }
  return out;
});
function onSerializerEnsureArray(ast) {
  switch (ast._tag) {
    default:
      return ast;
    case "Declaration":
    case "Arrays":
    case "Objects":
    case "Union":
    case "Suspend":
      return ast.recur(toCodecEnsureArray);
  }
}
// -----------------------------------------------------------------------------
// Optic APIs
// -----------------------------------------------------------------------------
/**
 * @category Optic
 * @since 4.0.0
 */
export function toIso(schema) {
  const serializer = toCodecIso(schema);
  return Optic_.makeIso(Parser.encodeSync(serializer), Parser.decodeSync(serializer));
}
/**
 * @category Optic
 * @since 4.0.0
 */
export function toIsoSource(_) {
  return Optic_.id();
}
/**
 * @category Optic
 * @since 4.0.0
 */
export function toIsoFocus(_) {
  return Optic_.id();
}
/**
 * **Technical Note**
 *
 * This annotation cannot be added to `Annotations.Bottom` because it changes
 * the schema type.
 *
 * @category Optic
 * @since 4.0.0
 */
export function overrideToCodecIso(to, transformation) {
  return schema => {
    return make(AST.annotate(schema.ast, {
      toCodecIso: () => new AST.Link(to.ast, Transformation.make(transformation))
    }), {
      schema
    });
  };
}
// -----------------------------------------------------------------------------
// Differ APIs
// -----------------------------------------------------------------------------
/**
 * @category JsonPatch
 * @since 4.0.0
 */
export function toDifferJsonPatch(schema) {
  const serializer = toCodecJson(schema); // TODO: remove this cast
  const get = Parser.encodeSync(serializer);
  const set = Parser.decodeSync(serializer);
  return {
    empty: [],
    diff: (oldValue, newValue) => JsonPatch.get(get(oldValue), get(newValue)),
    combine: (first, second) => [...first, ...second],
    patch: (oldValue, patch) => {
      const value = get(oldValue);
      const patched = JsonPatch.apply(patch, value);
      return Object.is(patched, value) ? oldValue : set(patched);
    }
  };
}
/**
 * @category Tree
 * @since 4.0.0
 */
export function Tree(node) {
  const Tree$ref = suspend(() => Tree);
  const Tree = Union([node, Array(Tree$ref), Record(String, Tree$ref)]);
  return Tree;
}
/**
 * @category JSON
 * @since 4.0.0
 */
export const Json = /*#__PURE__*/make(AST.Json);
/**
 * @category JSON
 * @since 4.0.0
 */
export const MutableJson = /*#__PURE__*/make(AST.MutableJson);
// -----------------------------------------------------------------------------
// Annotations
// -----------------------------------------------------------------------------
/**
 * Return all the typed annotations from the schema.
 *
 * @category Schema Resolvers
 * @since 4.0.0
 */
export function resolveInto(schema) {
  return InternalAnnotations.resolve(schema.ast);
}
