// effect/dist/Semaphore.js
/**
 * @since 2.0.0
 */
import * as Effect from "./Effect.js";
import * as internal from "./internal/effect.js";
import * as MutableHashMap from "./MutableHashMap.js";
import * as Option from "./Option.js";
/**
 * Unsafely creates a new Semaphore.
 *
 * **Previously Known As**
 *
 * This API replaces the following from Effect 3.x:
 *
 * - `Effect.makeSemaphoreUnsafe`
 *
 * @example
 * ```ts
 * import { Effect, Semaphore } from "effect"
 *
 * const semaphore = Semaphore.makeUnsafe(3)
 *
 * const task = (id: number) =>
 *   semaphore.withPermits(1)(
 *     Effect.gen(function*() {
 *       yield* Effect.log(`Task ${id} started`)
 *       yield* Effect.sleep("1 second")
 *       yield* Effect.log(`Task ${id} completed`)
 *     })
 *   )
 *
 * // Only 3 tasks can run concurrently
 * const program = Effect.all([
 *   task(1),
 *   task(2),
 *   task(3),
 *   task(4),
 *   task(5)
 * ], { concurrency: "unbounded" })
 * ```
 *
 * @since 2.0.0
 * @category constructors
 */
export const makeUnsafe = internal.makeSemaphoreUnsafe;
/**
 * Creates a new Semaphore.
 *
 * **Previously Known As**
 *
 * This API replaces the following from Effect 3.x:
 *
 * - `Effect.makeSemaphore`
 *
 * @example
 * ```ts
 * import { Effect, Semaphore } from "effect"
 *
 * const program = Effect.gen(function*() {
 *   const semaphore = yield* Semaphore.make(2)
 *
 *   const task = (id: number) =>
 *     semaphore.withPermits(1)(
 *       Effect.gen(function*() {
 *         yield* Effect.log(`Task ${id} acquired permit`)
 *         yield* Effect.sleep("1 second")
 *         yield* Effect.log(`Task ${id} releasing permit`)
 *       })
 *     )
 *
 *   // Run 4 tasks, but only 2 can run concurrently
 *   yield* Effect.all([task(1), task(2), task(3), task(4)])
 * })
 * ```
 *
 * @since 2.0.0
 * @category constructors
 */
export const make = internal.makeSemaphore;
// -----------------------------------------------------------------------------
// Partitioned
// -----------------------------------------------------------------------------
/**
 * @since 3.19.4
 * @category models
 */
export const PartitionedTypeId = "~effect/PartitionedSemaphore";
/**
 * Creates a `Partitioned` semaphore unsafely.
 *
 * **Previously Known As**
 *
 * This API replaces the following from Effect 3.x:
 *
 * - `PartitionedSemaphore.makeUnsafe`
 *
 * @since 3.19.4
 * @category constructors
 */
export const makePartitionedUnsafe = options => {
  const maxPermits = Math.max(0, options.permits);
  if (!Number.isFinite(maxPermits)) {
    return {
      [PartitionedTypeId]: PartitionedTypeId,
      withPermits: () => effect => effect
    };
  }
  let totalPermits = maxPermits;
  let waitingPermits = 0;
  const partitions = MutableHashMap.empty();
  const take = (key, permits) => Effect.callback(resume => {
    if (maxPermits < permits) {
      resume(Effect.never);
      return;
    }
    if (totalPermits >= permits) {
      totalPermits -= permits;
      resume(Effect.void);
      return;
    }
    const needed = permits - totalPermits;
    const taken = permits - needed;
    if (totalPermits > 0) {
      totalPermits = 0;
    }
    waitingPermits += needed;
    const waiters = Option.getOrElse(MutableHashMap.get(partitions, key), () => {
      const set = new Set();
      MutableHashMap.set(partitions, key, set);
      return set;
    });
    const entry = {
      permits: needed,
      resume: () => {
        cleanup();
        resume(Effect.void);
      }
    };
    const cleanup = () => {
      waiters.delete(entry);
      if (waiters.size === 0) {
        MutableHashMap.remove(partitions, key);
      }
    };
    waiters.add(entry);
    return Effect.sync(() => {
      cleanup();
      waitingPermits -= entry.permits;
      if (taken > 0) {
        releaseUnsafe(taken);
      }
    });
  });
  let iterator = partitions[Symbol.iterator]();
  const releaseUnsafe = permits => {
    while (permits > 0) {
      if (waitingPermits === 0) {
        totalPermits += permits;
        return;
      }
      let state = iterator.next();
      if (state.done) {
        iterator = partitions[Symbol.iterator]();
        state = iterator.next();
        if (state.done) {
          return;
        }
      }
      const waiter = state.value[1].values().next().value;
      if (waiter === undefined) {
        continue;
      }
      waiter.permits -= 1;
      waitingPermits -= 1;
      if (waiter.permits === 0) {
        waiter.resume();
      }
      permits -= 1;
    }
  };
  return {
    [PartitionedTypeId]: PartitionedTypeId,
    withPermits: (key, permits) => {
      const takePermits = take(key, permits);
      return effect => Effect.uninterruptibleMask(restore => Effect.flatMap(restore(takePermits), () => Effect.ensuring(restore(effect), Effect.sync(() => releaseUnsafe(permits)))));
    }
  };
};
/**
 * Creates a `Partitioned` semaphore.
 *
 * **Previously Known As**
 *
 * This API replaces the following from Effect 3.x:
 *
 * - `PartitionedSemaphore.make`
 *
 * @since 3.19.4
 * @category constructors
 */
export const makePartitioned = options => Effect.sync(() => makePartitionedUnsafe(options));
