// effect/dist/ServiceMap.js
import * as Equal from "./Equal.js";
import { constant, dual } from "./Function.js";
import * as Hash from "./Hash.js";
import { exitSucceed, PipeInspectableProto, withFiber, YieldableProto } from "./internal/core.js";
import * as Option from "./Option.js";
import { hasProperty } from "./Predicate.js";
const ServiceTypeId = "~effect/ServiceMap/Service";
/**
 * @example
 * ```ts
 * import { ServiceMap } from "effect"
 *
 * // Create a simple service
 * const Database = ServiceMap.Service<{
 *   query: (sql: string) => string
 * }>("Database")
 *
 * // Create a service class
 * class Config extends ServiceMap.Service<Config, {
 *   port: number
 * }>()("Config") {}
 *
 * // Use the services to create service maps
 * const db = ServiceMap.make(Database, {
 *   query: (sql) => `Result: ${sql}`
 * })
 * const config = ServiceMap.make(Config, { port: 8080 })
 * ```
 *
 * @since 4.0.0
 * @category Constructors
 */
export const Service = function () {
  const prevLimit = Error.stackTraceLimit;
  Error.stackTraceLimit = 2;
  const err = new Error();
  Error.stackTraceLimit = prevLimit;
  function KeyClass() {}
  const self = KeyClass;
  Object.setPrototypeOf(self, ServiceProto);
  Object.defineProperty(self, "stack", {
    get() {
      return err.stack;
    }
  });
  if (arguments.length > 0) {
    self.key = arguments[0];
    if (arguments[1]?.defaultValue) {
      self[ReferenceTypeId] = ReferenceTypeId;
      self.defaultValue = arguments[1].defaultValue;
    }
    return self;
  }
  return function (key, options) {
    self.key = key;
    if (options?.make) {
      ;
      self.make = options.make;
    }
    return self;
  };
};
const ServiceProto = {
  [ServiceTypeId]: {
    _Service: _ => _,
    _Identifier: _ => _
  },
  ...PipeInspectableProto,
  ...YieldableProto,
  toJSON() {
    return {
      _id: "Service",
      key: this.key,
      stack: this.stack
    };
  },
  asEffect() {
    const fn = this.asEffect = constant(withFiber(fiber => exitSucceed(get(fiber.services, this))));
    return fn();
  },
  of(self) {
    return self;
  },
  serviceMap(self) {
    return make(this, self);
  },
  use(f) {
    return withFiber(fiber => f(get(fiber.services, this)));
  },
  useSync(f) {
    return withFiber(fiber => exitSucceed(f(get(fiber.services, this))));
  }
};
const ReferenceTypeId = "~effect/ServiceMap/Reference";
const TypeId = "~effect/ServiceMap";
/**
 * @example
 * ```ts
 * import { ServiceMap } from "effect"
 *
 * // Create a service map from a Map (unsafe)
 * const map = new Map([
 *   ["Logger", { log: (msg: string) => console.log(msg) }]
 * ])
 *
 * const services = ServiceMap.makeUnsafe(map)
 * ```
 *
 * @since 4.0.0
 * @category Constructors
 */
export const makeUnsafe = mapUnsafe => {
  const self = Object.create(Proto);
  self.mapUnsafe = mapUnsafe;
  return self;
};
const Proto = {
  ...PipeInspectableProto,
  [TypeId]: {
    _Services: _ => _
  },
  toJSON() {
    return {
      _id: "ServiceMap",
      services: Array.from(this.mapUnsafe).map(([key, value]) => ({
        key,
        value
      }))
    };
  },
  [Equal.symbol](that) {
    if (!isServiceMap(that) || this.mapUnsafe.size !== that.mapUnsafe.size) return false;
    for (const k of this.mapUnsafe.keys()) {
      if (!that.mapUnsafe.has(k) || !Equal.equals(this.mapUnsafe.get(k), that.mapUnsafe.get(k))) {
        return false;
      }
    }
    return true;
  },
  [Hash.symbol]() {
    return Hash.number(this.mapUnsafe.size);
  }
};
/**
 * Checks if the provided argument is a `ServiceMap`.
 *
 * @example
 * ```ts
 * import { ServiceMap } from "effect"
 * import * as assert from "node:assert"
 *
 * assert.strictEqual(ServiceMap.isServiceMap(ServiceMap.empty()), true)
 * ```
 *
 * @since 4.0.0
 * @category Guards
 */
export const isServiceMap = u => hasProperty(u, TypeId);
/**
 * Checks if the provided argument is a `Key`.
 *
 * @example
 * ```ts
 * import { ServiceMap } from "effect"
 * import * as assert from "node:assert"
 *
 * assert.strictEqual(ServiceMap.isKey(ServiceMap.Service("Service")), true)
 * ```
 *
 * @since 4.0.0
 * @category Guards
 */
export const isKey = u => hasProperty(u, ServiceTypeId);
/**
 * Checks if the provided argument is a `Reference`.
 *
 * @example
 * ```ts
 * import { ServiceMap } from "effect"
 * import * as assert from "node:assert"
 *
 * const LoggerRef = ServiceMap.Reference("Logger", {
 *   defaultValue: () => ({ log: (msg: string) => console.log(msg) })
 * })
 *
 * assert.strictEqual(ServiceMap.isReference(LoggerRef), true)
 * assert.strictEqual(ServiceMap.isReference(ServiceMap.Service("Key")), false)
 * ```
 *
 * @since 4.0.0
 * @category Guards
 */
export const isReference = u => hasProperty(u, ReferenceTypeId);
/**
 * Returns an empty `ServiceMap`.
 *
 * @example
 * ```ts
 * import { ServiceMap } from "effect"
 * import * as assert from "node:assert"
 *
 * assert.strictEqual(ServiceMap.isServiceMap(ServiceMap.empty()), true)
 * ```
 *
 * @since 4.0.0
 * @category Constructors
 */
export const empty = () => emptyServiceMap;
const emptyServiceMap = /*#__PURE__*/makeUnsafe(/*#__PURE__*/new Map());
/**
 * Creates a new `ServiceMap` with a single service associated to the key.
 *
 * @example
 * ```ts
 * import { ServiceMap } from "effect"
 * import * as assert from "node:assert"
 *
 * const Port = ServiceMap.Service<{ PORT: number }>("Port")
 *
 * const Services = ServiceMap.make(Port, { PORT: 8080 })
 *
 * assert.deepStrictEqual(ServiceMap.get(Services, Port), { PORT: 8080 })
 * ```
 *
 * @since 4.0.0
 * @category Constructors
 */
export const make = (key, service) => makeUnsafe(new Map([[key.key, service]]));
/**
 * Adds a service to a given `ServiceMap`.
 *
 * @example
 * ```ts
 * import { pipe, ServiceMap } from "effect"
 * import * as assert from "node:assert"
 *
 * const Port = ServiceMap.Service<{ PORT: number }>("Port")
 * const Timeout = ServiceMap.Service<{ TIMEOUT: number }>("Timeout")
 *
 * const someServiceMap = ServiceMap.make(Port, { PORT: 8080 })
 *
 * const Services = pipe(
 *   someServiceMap,
 *   ServiceMap.add(Timeout, { TIMEOUT: 5000 })
 * )
 *
 * assert.deepStrictEqual(ServiceMap.get(Services, Port), { PORT: 8080 })
 * assert.deepStrictEqual(ServiceMap.get(Services, Timeout), { TIMEOUT: 5000 })
 * ```
 *
 * @since 4.0.0
 * @category Adders
 */
export const add = /*#__PURE__*/dual(3, (self, key, service) => {
  const map = new Map(self.mapUnsafe);
  map.set(key.key, service);
  return makeUnsafe(map);
});
/**
 * @since 4.0.0
 * @category Adders
 */
export const addOrOmit = /*#__PURE__*/dual(3, (self, key, service) => {
  const map = new Map(self.mapUnsafe);
  if (service._tag === "None") {
    map.delete(key.key);
  } else {
    map.set(key.key, service.value);
  }
  return makeUnsafe(map);
});
/**
 * Get a service from the context that corresponds to the given key, or
 * use the fallback value.
 *
 * @example
 * ```ts
 * import { ServiceMap } from "effect"
 * import * as assert from "node:assert"
 *
 * const Logger = ServiceMap.Service<{ log: (msg: string) => void }>("Logger")
 * const Database = ServiceMap.Service<{ query: (sql: string) => string }>(
 *   "Database"
 * )
 *
 * const services = ServiceMap.make(Logger, {
 *   log: (msg: string) => console.log(msg)
 * })
 *
 * const logger = ServiceMap.getOrElse(services, Logger, () => ({ log: () => {} }))
 * const database = ServiceMap.getOrElse(
 *   services,
 *   Database,
 *   () => ({ query: () => "fallback" })
 * )
 *
 * assert.deepStrictEqual(logger, { log: (msg: string) => console.log(msg) })
 * assert.deepStrictEqual(database, { query: () => "fallback" })
 * ```
 *
 * @since 4.0.0
 * @category Getters
 */
export const getOrElse = /*#__PURE__*/dual(3, (self, key, orElse) => {
  if (self.mapUnsafe.has(key.key)) {
    return self.mapUnsafe.get(key.key);
  }
  return isReference(key) ? getDefaultValue(key) : orElse();
});
/**
 * @since 4.0.0
 * @category Getters
 */
export const getOrUndefined = /*#__PURE__*/dual(2, (self, key) => self.mapUnsafe.get(key.key));
/**
 * Get a service from the context that corresponds to the given key.
 *
 * This function is unsafe because if the key is not present in the context, a
 * runtime error will be thrown.
 *
 * For a safer version see {@link getOption}.
 *
 * @param self - The `ServiceMap` to search for the service.
 * @param service - The `Service` of the service to retrieve.
 *
 * @example
 * ```ts
 * import { ServiceMap } from "effect"
 * import * as assert from "node:assert"
 *
 * const Port = ServiceMap.Service<{ PORT: number }>("Port")
 * const Timeout = ServiceMap.Service<{ TIMEOUT: number }>("Timeout")
 *
 * const Services = ServiceMap.make(Port, { PORT: 8080 })
 *
 * assert.deepStrictEqual(ServiceMap.getUnsafe(Services, Port), { PORT: 8080 })
 * assert.throws(() => ServiceMap.getUnsafe(Services, Timeout))
 * ```
 *
 * @since 4.0.0
 * @category unsafe
 */
export const getUnsafe = /*#__PURE__*/dual(2, (self, service) => {
  if (!self.mapUnsafe.has(service.key)) {
    if (ReferenceTypeId in service) return getDefaultValue(service);
    throw serviceNotFoundError(service);
  }
  return self.mapUnsafe.get(service.key);
});
/**
 * Get a service from the context that corresponds to the given key.
 *
 * @param self - The `ServiceMap` to search for the service.
 * @param service - The `Service` of the service to retrieve.
 *
 * @example
 * ```ts
 * import { pipe, ServiceMap } from "effect"
 * import * as assert from "node:assert"
 *
 * const Port = ServiceMap.Service<{ PORT: number }>("Port")
 * const Timeout = ServiceMap.Service<{ TIMEOUT: number }>("Timeout")
 *
 * const Services = pipe(
 *   ServiceMap.make(Port, { PORT: 8080 }),
 *   ServiceMap.add(Timeout, { TIMEOUT: 5000 })
 * )
 *
 * assert.deepStrictEqual(ServiceMap.get(Services, Timeout), { TIMEOUT: 5000 })
 * ```
 *
 * @since 4.0.0
 * @category Getters
 */
export const get = getUnsafe;
/**
 * @example
 * ```ts
 * import { ServiceMap } from "effect"
 * import * as assert from "node:assert"
 *
 * const LoggerRef = ServiceMap.Reference("Logger", {
 *   defaultValue: () => ({ log: (msg: string) => console.log(msg) })
 * })
 *
 * const services = ServiceMap.empty()
 * const logger = ServiceMap.getReferenceUnsafe(services, LoggerRef)
 *
 * assert.deepStrictEqual(logger, { log: (msg: string) => console.log(msg) })
 * ```
 *
 * @since 4.0.0
 * @category unsafe
 */
export const getReferenceUnsafe = (self, service) => {
  if (!self.mapUnsafe.has(service.key)) {
    return getDefaultValue(service);
  }
  return self.mapUnsafe.get(service.key);
};
const defaultValueCacheKey = "~effect/ServiceMap/defaultValue";
const getDefaultValue = ref => {
  if (defaultValueCacheKey in ref) {
    return ref[defaultValueCacheKey];
  }
  return ref[defaultValueCacheKey] = ref.defaultValue();
};
const serviceNotFoundError = service => {
  const error = new Error(`Service not found${service.key ? `: ${String(service.key)}` : ""}`);
  if (service.stack) {
    const lines = service.stack.split("\n");
    if (lines.length > 2) {
      const afterAt = lines[2].match(/at (.*)/);
      if (afterAt) {
        error.message = error.message + ` (defined at ${afterAt[1]})`;
      }
    }
  }
  if (error.stack) {
    const lines = error.stack.split("\n");
    lines.splice(1, 3);
    error.stack = lines.join("\n");
  }
  return error;
};
/**
 * Get the value associated with the specified key from the context wrapped in
 * an `Option` object. If the key is not found, the `Option` object will be
 * `None`.
 *
 * @param self - The `ServiceMap` to search for the service.
 * @param service - The `Service` of the service to retrieve.
 *
 * @example
 * ```ts
 * import { Option, ServiceMap } from "effect"
 * import * as assert from "node:assert"
 *
 * const Port = ServiceMap.Service<{ PORT: number }>("Port")
 * const Timeout = ServiceMap.Service<{ TIMEOUT: number }>("Timeout")
 *
 * const Services = ServiceMap.make(Port, { PORT: 8080 })
 *
 * assert.deepStrictEqual(
 *   ServiceMap.getOption(Services, Port),
 *   Option.some({ PORT: 8080 })
 * )
 * assert.deepStrictEqual(ServiceMap.getOption(Services, Timeout), Option.none())
 * ```
 *
 * @since 4.0.0
 * @category Getters
 */
export const getOption = /*#__PURE__*/dual(2, (self, service) => {
  if (self.mapUnsafe.has(service.key)) {
    return Option.some(self.mapUnsafe.get(service.key));
  }
  return isReference(service) ? Option.some(getDefaultValue(service)) : Option.none();
});
/**
 * Merges two `ServiceMap`s, returning a new `ServiceMap` containing the services of both.
 *
 * @param self - The first `ServiceMap` to merge.
 * @param that - The second `ServiceMap` to merge.
 *
 * @example
 * ```ts
 * import { ServiceMap } from "effect"
 * import * as assert from "node:assert"
 *
 * const Port = ServiceMap.Service<{ PORT: number }>("Port")
 * const Timeout = ServiceMap.Service<{ TIMEOUT: number }>("Timeout")
 *
 * const firstServiceMap = ServiceMap.make(Port, { PORT: 8080 })
 * const secondServiceMap = ServiceMap.make(Timeout, { TIMEOUT: 5000 })
 *
 * const Services = ServiceMap.merge(firstServiceMap, secondServiceMap)
 *
 * assert.deepStrictEqual(ServiceMap.get(Services, Port), { PORT: 8080 })
 * assert.deepStrictEqual(ServiceMap.get(Services, Timeout), { TIMEOUT: 5000 })
 * ```
 *
 * @since 4.0.0
 * @category Utils
 */
export const merge = /*#__PURE__*/dual(2, (self, that) => {
  if (self.mapUnsafe.size === 0) return that;
  if (that.mapUnsafe.size === 0) return self;
  const map = new Map(self.mapUnsafe);
  that.mapUnsafe.forEach((value, key) => map.set(key, value));
  return makeUnsafe(map);
});
/**
 * Merges any number of `ServiceMap`s, returning a new `ServiceMap` containing the services of all.
 *
 * @example
 * ```ts
 * import { ServiceMap } from "effect"
 * import * as assert from "node:assert"
 *
 * const Port = ServiceMap.Service<{ PORT: number }>("Port")
 * const Timeout = ServiceMap.Service<{ TIMEOUT: number }>("Timeout")
 * const Host = ServiceMap.Service<{ HOST: string }>("Host")
 *
 * const firstServiceMap = ServiceMap.make(Port, { PORT: 8080 })
 * const secondServiceMap = ServiceMap.make(Timeout, { TIMEOUT: 5000 })
 * const thirdServiceMap = ServiceMap.make(Host, { HOST: "localhost" })
 *
 * const Services = ServiceMap.mergeAll(
 *   firstServiceMap,
 *   secondServiceMap,
 *   thirdServiceMap
 * )
 *
 * assert.deepStrictEqual(ServiceMap.get(Services, Port), { PORT: 8080 })
 * assert.deepStrictEqual(ServiceMap.get(Services, Timeout), { TIMEOUT: 5000 })
 * assert.deepStrictEqual(ServiceMap.get(Services, Host), { HOST: "localhost" })
 * ```
 *
 * @since 3.12.0
 */
export const mergeAll = (...ctxs) => {
  const map = new Map();
  for (let i = 0; i < ctxs.length; i++) {
    ctxs[i].mapUnsafe.forEach((value, key) => {
      map.set(key, value);
    });
  }
  return makeUnsafe(map);
};
/**
 * Returns a new `ServiceMap` that contains only the specified services.
 *
 * @param self - The `ServiceMap` to prune services from.
 * @param services - The list of `Service`s to be included in the new `ServiceMap`.
 *
 * @example
 * ```ts
 * import { Option, pipe, ServiceMap } from "effect"
 * import * as assert from "node:assert"
 *
 * const Port = ServiceMap.Service<{ PORT: number }>("Port")
 * const Timeout = ServiceMap.Service<{ TIMEOUT: number }>("Timeout")
 *
 * const someServiceMap = pipe(
 *   ServiceMap.make(Port, { PORT: 8080 }),
 *   ServiceMap.add(Timeout, { TIMEOUT: 5000 })
 * )
 *
 * const Services = pipe(someServiceMap, ServiceMap.pick(Port))
 *
 * assert.deepStrictEqual(
 *   ServiceMap.getOption(Services, Port),
 *   Option.some({ PORT: 8080 })
 * )
 * assert.deepStrictEqual(ServiceMap.getOption(Services, Timeout), Option.none())
 * ```
 *
 * @since 4.0.0
 * @category Utils
 */
export const pick = (...services) => self => {
  const map = new Map();
  const keySet = new Set(services.map(key => key.key));
  self.mapUnsafe.forEach((value, key) => {
    if (keySet.has(key)) {
      map.set(key, value);
    }
  });
  return makeUnsafe(map);
};
/**
 * @example
 * ```ts
 * import { Option, pipe, ServiceMap } from "effect"
 * import * as assert from "node:assert"
 *
 * const Port = ServiceMap.Service<{ PORT: number }>("Port")
 * const Timeout = ServiceMap.Service<{ TIMEOUT: number }>("Timeout")
 *
 * const someServiceMap = pipe(
 *   ServiceMap.make(Port, { PORT: 8080 }),
 *   ServiceMap.add(Timeout, { TIMEOUT: 5000 })
 * )
 *
 * const Services = pipe(someServiceMap, ServiceMap.omit(Timeout))
 *
 * assert.deepStrictEqual(
 *   ServiceMap.getOption(Services, Port),
 *   Option.some({ PORT: 8080 })
 * )
 * assert.deepStrictEqual(ServiceMap.getOption(Services, Timeout), Option.none())
 * ```
 *
 * @since 4.0.0
 * @category Utils
 */
export const omit = (...keys) => self => {
  const map = new Map(self.mapUnsafe);
  for (let i = 0; i < keys.length; i++) {
    map.delete(keys[i].key);
  }
  return makeUnsafe(map);
};
/**
 * Creates a service map key with a default value.
 *
 * **Details**
 *
 * `ServiceMap.Reference` allows you to create a key that can hold a value. You
 * can provide a default value for the service, which will automatically be used
 * when the context is accessed, or override it with a custom implementation
 * when needed.
 *
 * @example
 * ```ts
 * import { ServiceMap } from "effect"
 *
 * // Create a reference with a default value
 * const LoggerRef = ServiceMap.Reference("Logger", {
 *   defaultValue: () => ({ log: (msg: string) => console.log(msg) })
 * })
 *
 * // The reference provides the default value when accessed from an empty context
 * const services = ServiceMap.empty()
 * const logger = ServiceMap.get(services, LoggerRef)
 *
 * // You can also override the default value
 * const customServices = ServiceMap.make(LoggerRef, {
 *   log: (msg: string) => `Custom: ${msg}`
 * })
 * const customLogger = ServiceMap.get(customServices, LoggerRef)
 * ```
 *
 * @since 4.0.0
 * @category References
 */
export const Reference = Service;
