// effect/dist/Utils.js
/**
 * @since 2.0.0
 */
import { identity } from "./Function.js";
import { isObject } from "./Predicate.js";
/**
 * @since 2.0.0
 */
const GenKindTypeId = "~effect/Utils/GenKind";
/**
 * @category predicates
 * @since 3.0.6
 */
export const isGenKind = u => isObject(u) && GenKindTypeId in u;
class GenKindImpl {
  value;
  constructor(value) {
    this.value = value;
  }
  get _F() {
    return identity;
  }
  get _R() {
    return _ => _;
  }
  get _O() {
    return _ => _;
  }
  get _E() {
    return _ => _;
  }
  [GenKindTypeId] = GenKindTypeId;
  [Symbol.iterator]() {
    return new SingleShotGen(this);
  }
}
/**
 * @category constructors
 * @since 2.0.0
 */
export class SingleShotGen {
  called = false;
  self;
  constructor(self) {
    this.self = self;
  }
  /**
   * @since 2.0.0
   */
  next(a) {
    return this.called ? {
      value: a,
      done: true
    } : (this.called = true, {
      value: this.self,
      done: false
    });
  }
  /**
   * @since 2.0.0
   */
  [Symbol.iterator]() {
    return new SingleShotGen(this.self);
  }
}
/**
 * @category constructors
 * @since 2.0.0
 */
export const makeGenKind = kind => new GenKindImpl(kind);
const InternalTypeId = "~effect/Effect/internal";
const standard = {
  [InternalTypeId]: body => {
    return body();
  }
};
const forced = {
  [InternalTypeId]: body => {
    try {
      return body();
    } finally {
      //
    }
  }
};
const isNotOptimizedAway = /*#__PURE__*/standard[InternalTypeId](() => new Error().stack)?.includes(InternalTypeId) === true;
/**
 * @since 3.2.2
 * @status experimental
 * @category tracing
 */
export const internalCall = isNotOptimizedAway ? standard[InternalTypeId] : forced[InternalTypeId];
const genConstructor = function* () {}.constructor;
/**
 * @example
 * ```ts
 * import { Utils } from "effect"
 *
 * function* generatorFn() {
 *   yield 1
 *   yield 2
 * }
 *
 * console.log(Utils.isGeneratorFunction(generatorFn)) // true
 * console.log(Utils.isGeneratorFunction(() => {})) // false
 * ```
 *
 * @category predicates
 * @since 3.11.0
 */
export const isGeneratorFunction = u => isObject(u) && u.constructor === genConstructor;
