// @platformatic/kafka/dist/clients/producer/producer.js
import { randomUUID } from 'crypto';
import { createPromisifiedCallback, kCallbackPromise, runConcurrentCallbacks } from "../../apis/callbacks.js";
import { FindCoordinatorKeyTypes, ProduceAcks } from "../../apis/enumerations.js";
import { createDiagnosticContext, producerInitIdempotentChannel, producerSendsChannel, producerTransactionsChannel } from "../../diagnostic.js";
import { UserError } from "../../errors.js";
import { murmur2 } from "../../protocol/murmur2.js";
import { runAsyncSeries } from "../../registries/abstract.js";
import { kInstance, kTransaction, kTransactionAddOffsets, kTransactionAddPartitions, kTransactionCancel, kTransactionCommitOffset, kTransactionEnd, kTransactionFindCoordinator, kTransactionPrepare } from "../../symbols.js";
import { emitExperimentalApiWarning, NumericMap } from "../../utils.js";
import { Base, kAfterCreate, kCheckNotClosed, kClosed, kGetApi, kGetBootstrapConnection, kGetConnection, kMetadata, kOptions, kPerformDeduplicated, kPerformWithRetry, kPrometheus, kValidateOptions } from "../base/base.js";
import { ensureMetric } from "../metrics.js";
import { produceOptionsValidator, producerOptionsValidator, producerStreamOptionsValidator, sendOptionsValidator } from "./options.js";
import { ProducerStream } from "./producer-stream.js";
import { Transaction } from "./transaction.js";
// Don't move this function as being in the same file will enable V8 to remove.
// For futher info, ask Matteo.
export function noopSerializer(data) {
    return data;
}
export class Producer extends Base {
    #partitionsRoundRobin;
    // These two values should be serializable and loadable in the constructor in order to restore
    // the idempotent producer status.
    #producerInfo;
    #sequences;
    #keySerializer;
    #valueSerializer;
    #headerKeySerializer;
    #headerValueSerializer;
    #metricsProducedMessages;
    #coordinatorId;
    #transaction;
    #streams;
    #sendOperation;
    constructor(options) {
        if (options.idempotent) {
            options.maxInflights = 1;
            options.acks = ProduceAcks.ALL;
            options.retries = Number.MAX_SAFE_INTEGER;
        }
        else {
            options.idempotent = false;
        }
        options.repeatOnStaleMetadata ??= true;
        super(options);
        this[kValidateOptions](options, producerOptionsValidator, '/options');
        if (options.beforeSerialization) {
            emitExperimentalApiWarning('beforeSerialization');
        }
        if (options.registry) {
            emitExperimentalApiWarning('registry (Confluent Schema Registry integration)');
        }
        let serializers = options.serializers;
        if (options.registry) {
            if (options.beforeSerialization) {
                throw new UserError('/options/beforeSerialization cannot be provided when /options/registry is provided.');
            }
            else if (options.serializers) {
                throw new UserError('/options/serializers cannot be provided when /options/registry is provided.');
            }
            serializers = options.registry.getSerializers();
        }
        this.#partitionsRoundRobin = new NumericMap();
        this.#sequences = new NumericMap();
        this.#keySerializer = serializers?.key ?? noopSerializer;
        this.#valueSerializer = serializers?.value ?? noopSerializer;
        this.#headerKeySerializer = serializers?.headerKey ?? noopSerializer;
        this.#headerValueSerializer = serializers?.headerValue ?? noopSerializer;
        this.#streams = new Set();
        this[kOptions].transactionalId ??= randomUUID();
        if (this[kPrometheus]) {
            ensureMetric(this[kPrometheus], 'Gauge', 'kafka_producers', 'Number of active Kafka producers').inc();
            this.#metricsProducedMessages = ensureMetric(this[kPrometheus], 'Counter', 'kafka_produced_messages', 'Number of produced Kafka messages');
        }
        if (options.registry) {
            this.#sendOperation = this.#beforeSerialization.bind(this, options.registry.getBeforeSerializationHook());
        }
        else if (options.beforeSerialization) {
            this.#sendOperation = this.#beforeSerialization.bind(this, options.beforeSerialization);
        }
        else {
            this.#sendOperation = this.#send.bind(this);
        }
        this[kAfterCreate]('producer');
    }
    get producerId() {
        return this.#producerInfo?.producerId;
    }
    get producerEpoch() {
        return this.#producerInfo?.producerEpoch;
    }
    get transaction() {
        return this.#transaction;
    }
    get coordinatorId() {
        return this.#coordinatorId;
    }
    get streamsCount() {
        return this.#streams.size;
    }
    close(force, callback) {
        if (typeof force === 'function') {
            callback = force;
            force = false;
        }
        if (!callback) {
            callback = createPromisifiedCallback();
        }
        if (this[kClosed]) {
            callback(null);
            return callback[kCallbackPromise];
        }
        this[kClosed] = true;
        this.#performClose(force, callback);
        return callback[kCallbackPromise];
    }
    #performClose(force, callback) {
        for (const stream of this.#streams) {
            /* c8 ignore next 3 - Hard to test */
            if (stream.closed || stream.destroyed) {
                this.#streams.delete(stream);
            }
        }
        if (this.#streams.size) {
            if (!force) {
                this[kClosed] = false;
                callback(new UserError('Cannot close producer while producing messages.'));
                return;
            }
            runConcurrentCallbacks('Closing streams failed.', this.#streams, (stream, concurrentCallback) => {
                stream.close(concurrentCallback);
            }, error => {
                /* c8 ignore next 5 - Hard to test */
                if (error) {
                    this[kClosed] = false;
                    callback(error);
                    return;
                }
                this.#streams.clear();
                this.#performClose(false, callback);
            });
            return;
        }
        super.close(error => {
            if (error) {
                this[kClosed] = false;
                callback(error);
                return;
            }
            if (this[kPrometheus]) {
                ensureMetric(this[kPrometheus], 'Gauge', 'kafka_producers', 'Number of active Kafka producers').dec();
            }
            callback(null);
        });
    }
    initIdempotentProducer(options, callback) {
        if (!callback) {
            callback = createPromisifiedCallback();
        }
        if (this[kCheckNotClosed](callback)) {
            return callback[kCallbackPromise];
        }
        const validationError = this[kValidateOptions](options, produceOptionsValidator, '/options', false);
        if (validationError) {
            callback(validationError);
            return callback[kCallbackPromise];
        }
        producerInitIdempotentChannel.traceCallback(this.#initIdempotentProducer, 1, createDiagnosticContext({ client: this, operation: 'initIdempotentProducer', options }), this, options, callback);
        return callback[kCallbackPromise];
    }
    send(options, callback) {
        if (!callback) {
            callback = createPromisifiedCallback();
        }
        if (this[kCheckNotClosed](callback)) {
            return callback[kCallbackPromise];
        }
        if (this.#transaction && options[kTransaction] !== this.#transaction[kInstance]) {
            callback(new UserError(options[kTransaction]
                ? 'The producer is in use by another transaction.'
                : 'The producer is in use by a transaction.'));
            return callback[kCallbackPromise];
        }
        const validationError = this[kValidateOptions](options, sendOptionsValidator, '/options', false);
        if (validationError) {
            callback(validationError);
            return callback[kCallbackPromise];
        }
        const idempotent = options.idempotent ?? this[kOptions].idempotent;
        if (idempotent) {
            if (typeof options.producerId !== 'undefined' || typeof options.producerEpoch !== 'undefined') {
                callback(new UserError('Cannot specify producerId or producerEpoch when using idempotent producer.'));
                return callback[kCallbackPromise];
            }
            if (typeof options.acks !== 'undefined' && options.acks !== ProduceAcks.ALL) {
                callback(new UserError('Idempotent producer requires acks to be ALL (-1).'));
                return callback[kCallbackPromise];
            }
        }
        options.acks ??= idempotent ? ProduceAcks.ALL : ProduceAcks.LEADER;
        producerSendsChannel.traceCallback(this.#sendOperation, 1, createDiagnosticContext({ client: this, operation: 'send', options }), this, options, callback);
        return callback[kCallbackPromise];
    }
    asStream(options) {
        options ??= {};
        const validationError = this[kValidateOptions](options, producerStreamOptionsValidator, '/options', false);
        if (validationError) {
            throw validationError;
        }
        const stream = new ProducerStream(this, options);
        this.#streams.add(stream);
        stream.once('close', () => {
            this.#streams.delete(stream);
        });
        return stream;
    }
    beginTransaction(options, callback) {
        if (!callback) {
            callback = createPromisifiedCallback();
        }
        if (this[kCheckNotClosed](callback)) {
            return callback[kCallbackPromise];
        }
        if (!options) {
            options = {};
        }
        const validationError = this[kValidateOptions](options, produceOptionsValidator, '/options', false);
        if (validationError) {
            callback(validationError);
            return callback[kCallbackPromise];
        }
        if (!this[kOptions].idempotent) {
            callback(new UserError('Cannot begin a transaction on a non-idempotent producer.'));
            return callback[kCallbackPromise];
        }
        if (this.#transaction) {
            callback(new UserError('There is already an active transaction.'));
            return callback[kCallbackPromise];
        }
        this.#transaction = new Transaction(this);
        producerTransactionsChannel.traceCallback(this.#transaction[kTransactionPrepare], 2, createDiagnosticContext({ client: this, operation: 'begin' }), this.#transaction, this.#producerInfo?.transactionalId, options, callback);
        return callback[kCallbackPromise];
    }
    [kTransactionFindCoordinator](callback) {
        if (this.#coordinatorId !== undefined) {
            callback(null);
            return;
        }
        const transactionalId = this.#transaction.id;
        this[kPerformDeduplicated]('findCoordinator', deduplicateCallback => {
            this[kPerformWithRetry]('findCoordinator', retryCallback => {
                this[kGetBootstrapConnection]((error, connection) => {
                    if (error) {
                        retryCallback(error);
                        return;
                    }
                    this[kGetApi]('FindCoordinator', (error, api) => {
                        if (error) {
                            retryCallback(error);
                            return;
                        }
                        api(connection, FindCoordinatorKeyTypes.TRANSACTION, [transactionalId], retryCallback);
                    });
                });
            }, (error, response) => {
                if (error) {
                    deduplicateCallback(error);
                    return;
                }
                const groupInfo = response.coordinators.find(coordinator => coordinator.key === transactionalId);
                this.#coordinatorId = groupInfo.nodeId;
                deduplicateCallback(null);
            }, 0);
        }, callback);
    }
    [kTransactionAddPartitions](transactionId, topicsPartitions, callback) {
        if (transactionId !== this.#transaction?.[kInstance]) {
            callback(new UserError('The producer is in use by another transaction.'));
            return;
        }
        const transactions = [];
        for (const [topic, partitionsSet] of topicsPartitions) {
            transactions.push({
                name: topic,
                partitions: Array.from(partitionsSet)
            });
        }
        this[kPerformDeduplicated]('addPartitionsToTransaction', deduplicateCallback => {
            this[kPerformWithRetry]('addPartitionsToTransaction', retryCallback => {
                this.#getCoordinatorConnection((error, connection) => {
                    if (error) {
                        retryCallback(error);
                        return;
                    }
                    this[kGetApi]('AddPartitionsToTxn', (error, api) => {
                        if (error) {
                            retryCallback(error);
                            return;
                        }
                        api(connection, [
                            {
                                transactionalId: this.#transaction.id,
                                producerId: this.#producerInfo.producerId,
                                producerEpoch: this.#producerInfo.producerEpoch,
                                topics: transactions,
                                verifyOnly: false
                            }
                        ], error => {
                            if (error) {
                                retryCallback(error);
                                return;
                            }
                            retryCallback(null);
                        });
                    });
                });
            }, deduplicateCallback);
        }, callback);
    }
    [kTransactionAddOffsets](transactionId, groupId, callback) {
        if (transactionId !== this.#transaction?.[kInstance]) {
            callback(new UserError('The producer is in use by another transaction.'));
            return;
        }
        this[kPerformDeduplicated]('addOffsetsToTransaction', deduplicateCallback => {
            this[kPerformWithRetry]('addOffsetsToTransaction', retryCallback => {
                this.#getCoordinatorConnection((error, connection) => {
                    if (error) {
                        retryCallback(error);
                        return;
                    }
                    this[kGetApi]('AddOffsetsToTxn', (error, api) => {
                        if (error) {
                            retryCallback(error);
                            return;
                        }
                        api(connection, this.#transaction.id, this.#producerInfo.producerId, this.#producerInfo.producerEpoch, groupId, error => {
                            if (error) {
                                retryCallback(error);
                                return;
                            }
                            retryCallback(null);
                        });
                    });
                });
            }, deduplicateCallback);
        }, callback);
    }
    [kTransactionCommitOffset](transactionId, message, callback) {
        if (transactionId !== this.#transaction?.[kInstance]) {
            callback(new UserError('The producer is in use by another transaction.'));
            return;
        }
        this[kPerformDeduplicated]('commit', deduplicateCallback => {
            this[kPerformWithRetry]('commit', retryCallback => {
                this[kMetadata]({ topics: [message.topic] }, (error, metadata) => {
                    if (error) {
                        retryCallback(error);
                        return;
                    }
                    const { groupId, generationId, memberId, coordinatorId } = message.metadata
                        .consumer;
                    this[kGetConnection](metadata.brokers.get(coordinatorId), (error, connection) => {
                        if (error) {
                            retryCallback(error);
                            return;
                        }
                        this[kGetApi]('TxnOffsetCommit', (error, api) => {
                            if (error) {
                                retryCallback(error);
                                return;
                            }
                            const { topic, partition } = message;
                            api(connection, this.#transaction.id, groupId, this.#producerInfo.producerId, this.#producerInfo.producerEpoch, generationId, memberId, null, [
                                {
                                    name: topic,
                                    partitions: [
                                        {
                                            partitionIndex: message.partition,
                                            committedOffset: message.offset + 1n,
                                            committedLeaderEpoch: metadata.topics.get(topic).partitions[partition].leaderEpoch
                                        }
                                    ]
                                }
                            ], error => {
                                if (error) {
                                    retryCallback(error);
                                    return;
                                }
                                retryCallback(null);
                            });
                        });
                    });
                });
            }, deduplicateCallback);
        }, callback);
    }
    [kTransactionEnd](transactionId, commit, callback) {
        if (transactionId !== this.#transaction?.[kInstance]) {
            callback(new UserError('The producer is in use by another transaction.'));
            return;
        }
        this[kPerformDeduplicated]('endTransaction', deduplicateCallback => {
            this[kPerformWithRetry]('endTransaction', retryCallback => {
                this.#getCoordinatorConnection((error, connection) => {
                    if (error) {
                        retryCallback(error);
                        return;
                    }
                    this[kGetApi]('EndTxn', (error, api) => {
                        if (error) {
                            retryCallback(error);
                            return;
                        }
                        api(connection, this.#transaction.id, this.#producerInfo.producerId, this.#producerInfo.producerEpoch, commit, error => {
                            if (error) {
                                this.#handleFencingError(error);
                                retryCallback(error);
                                return;
                            }
                            this.#transaction = undefined;
                            retryCallback(null);
                        });
                    });
                });
            }, deduplicateCallback);
        }, callback);
    }
    [kTransactionCancel](transactionId, callback) {
        if (transactionId !== this.#transaction?.[kInstance]) {
            callback(new UserError('The producer is in use by another transaction.'));
            return;
        }
        this.#transaction = undefined;
        callback(null);
    }
    #initIdempotentProducer(options, callback) {
        const transactionalId = this.#transaction?.id;
        this[kPerformDeduplicated]('initProducerId', deduplicateCallback => {
            this[kPerformWithRetry]('initProducerId', retryCallback => {
                const connector = transactionalId
                    ? this.#getCoordinatorConnection.bind(this)
                    : this[kGetBootstrapConnection].bind(this);
                connector((error, connection) => {
                    if (error) {
                        retryCallback(error);
                        return;
                    }
                    this[kGetApi]('InitProducerId', (error, api) => {
                        if (error) {
                            retryCallback(error);
                            return;
                        }
                        api(connection, transactionalId, this[kOptions].timeout, options.producerId ?? this[kOptions].producerId ?? 0n, options.producerEpoch ?? this[kOptions].producerEpoch ?? 0, retryCallback);
                    });
                });
            }, (error, response) => {
                if (error) {
                    this.#handleFencingError(error);
                    deduplicateCallback(error);
                    return;
                }
                this.#producerInfo = {
                    producerId: response.producerId,
                    producerEpoch: response.producerEpoch,
                    transactionalId
                };
                this.#sequences.clear();
                deduplicateCallback(null, this.#producerInfo);
            }, 0);
        }, callback);
    }
    #send(options, callback) {
        options.idempotent ??= this[kOptions].idempotent;
        options.repeatOnStaleMetadata ??= this[kOptions].repeatOnStaleMetadata;
        options.partitioner ??= this[kOptions].partitioner;
        const { idempotent, partitioner } = options;
        let producerId = options.producerId;
        let producerEpoch = options.producerEpoch;
        // If the producer was transactional and is not anymore, reset
        if (!this.#transaction && this.#producerInfo?.transactionalId) {
            this.#producerInfo = undefined;
        }
        // Initialize the idempotent producer if required
        if (options.idempotent) {
            if (!this.#producerInfo) {
                const { messages, ...initOptions } = options;
                this.initIdempotentProducer(initOptions, error => {
                    if (error) {
                        callback(error);
                        return;
                    }
                    this.#send(options, callback);
                });
                return;
            }
            producerId = this.#producerInfo.producerId;
            producerEpoch = this.#producerInfo.producerEpoch;
        }
        const produceOptions = {
            compression: options.compression ?? this[kOptions].compression,
            producerId,
            producerEpoch,
            sequences: idempotent ? this.#sequences : undefined,
            transactionalId: this.#transaction?.id
        };
        // Build messages records out of messages
        const topics = new Set();
        const messages = [];
        for (const message of options.messages) {
            const topic = message.topic;
            let key;
            let value;
            let headers = new Map();
            const serializedHeaders = new Map();
            const metadata = message.metadata;
            try {
                if (message.headers) {
                    headers =
                        message.headers instanceof Map
                            ? message.headers
                            : new Map(Object.entries(message.headers));
                    for (const [key, value] of headers) {
                        serializedHeaders.set(this.#headerKeySerializer(key, metadata), this.#headerValueSerializer(value, metadata));
                    }
                }
                key = this.#keySerializer(message.key, headers, message);
                value = this.#valueSerializer(message.value, headers, message);
            }
            catch (error) {
                callback(new UserError('Failed to serialize a message.', { cause: error }));
                return;
            }
            let partition = 0;
            if (typeof message.partition !== 'number') {
                if (partitioner) {
                    partition = partitioner(message);
                }
                else if (key) {
                    partition = murmur2(key) >>> 0;
                }
                else {
                    // Use the roundrobin
                    partition = this.#partitionsRoundRobin.postIncrement(topic, 1, 0);
                }
            }
            else {
                partition = message.partition;
            }
            topics.add(topic);
            messages.push({
                key,
                value,
                headers: serializedHeaders,
                topic,
                partition,
                timestamp: message.timestamp
            });
        }
        const topicsArray = Array.from(topics);
        if (this.#transaction) {
            // Get the metadata for topics, we need to normalize the partitions
            this[kMetadata]({ topics: topicsArray, autocreateTopics: options.autocreateTopics }, (error, metadata) => {
                if (error) {
                    callback(error);
                    return;
                }
                this.#transaction?.addPartitions(metadata, messages, error => {
                    if (error) {
                        callback(error);
                        return;
                    }
                    this.#performSend(topicsArray, messages, options, produceOptions, callback);
                });
            });
        }
        else {
            this.#performSend(topicsArray, messages, options, produceOptions, callback);
        }
    }
    #performSend(topics, messages, sendOptions, produceOptions, callback) {
        // Get the metadata with the topic/partitions informations
        this[kMetadata]({ topics, autocreateTopics: sendOptions.autocreateTopics }, (error, metadata) => {
            if (error) {
                callback(error);
                return;
            }
            const messagesByDestination = new Map();
            // Track the number of messages per partition so we can update the sequence number
            const messagesPerPartition = new NumericMap();
            // Normalize the partition of all messages, then enqueue them to their destination
            for (const message of messages) {
                message.partition %= metadata.topics.get(message.topic).partitionsCount;
                const { topic, partition } = message;
                const leader = metadata.topics.get(topic).partitions[partition].leader;
                let destination = messagesByDestination.get(leader);
                if (!destination) {
                    destination = [];
                    messagesByDestination.set(leader, destination);
                }
                const messagePartitionKey = `${message.topic}:${partition}`;
                messagesPerPartition.postIncrement(messagePartitionKey, 1, 0);
                destination.push(message);
            }
            // Track nodes so that we can get their ID for delayed write reporting
            const nodes = [];
            runConcurrentCallbacks('Producing messages failed.', messagesByDestination, ([destination, destinationMessages], concurrentCallback) => {
                nodes.push(destination);
                this.#performSingleDestinationSend(topics, destinationMessages, this[kOptions].timeout, sendOptions.acks, sendOptions.autocreateTopics, sendOptions.repeatOnStaleMetadata, produceOptions, concurrentCallback);
            }, (error, apiResults) => {
                if (error) {
                    callback(error);
                    return;
                }
                this.#metricsProducedMessages?.inc(messages.length);
                const results = {};
                if (sendOptions.acks === ProduceAcks.NO_RESPONSE) {
                    const unwritableNodes = [];
                    for (let i = 0; i < apiResults.length; i++) {
                        if (apiResults[i] === false) {
                            unwritableNodes.push(nodes[i]);
                        }
                    }
                    results.unwritableNodes = unwritableNodes;
                }
                else {
                    const topics = [];
                    for (const result of apiResults) {
                        for (const { name, partitionResponses } of result.responses) {
                            for (const partitionResponse of partitionResponses) {
                                topics.push({
                                    topic: name,
                                    partition: partitionResponse.index,
                                    offset: partitionResponse.baseOffset
                                });
                                const partitionKey = `${name}:${partitionResponse.index}`;
                                this.#sequences.postIncrement(partitionKey, messagesPerPartition.get(partitionKey), 0);
                            }
                        }
                        results.offsets = topics;
                    }
                }
                callback(null, results);
            });
        });
    }
    #performSingleDestinationSend(topics, messages, timeout, acks, autocreateTopics, repeatOnStaleMetadata, produceOptions, callback) {
        // Get the metadata with the topic/partitions informations
        this[kMetadata]({ topics, autocreateTopics }, (error, metadata) => {
            if (error) {
                callback(error);
                return;
            }
            const { topic, partition } = messages[0];
            const leader = metadata.topics.get(topic).partitions[partition].leader;
            this[kPerformWithRetry]('produce', retryCallback => {
                this[kGetConnection](metadata.brokers.get(leader), (error, connection) => {
                    if (error) {
                        retryCallback(error);
                        return;
                    }
                    this[kGetApi]('Produce', (error, api) => {
                        if (error) {
                            retryCallback(error);
                            return;
                        }
                        api(connection, acks, timeout, messages, produceOptions, retryCallback);
                    });
                });
            }, (error, results) => {
                if (error) {
                    // If the last error was due to stale metadata, we retry the operation with this set of messages
                    // since the partition is already set, it should attempt on the new destination
                    const hasStaleMetadata = error.findBy('hasStaleMetadata', true);
                    if (hasStaleMetadata && repeatOnStaleMetadata) {
                        this.clearMetadata();
                        this.#performSingleDestinationSend(topics, messages, timeout, acks, autocreateTopics, false, produceOptions, callback);
                        return;
                    }
                    callback(error);
                    return;
                }
                callback(error, results);
            }, 0, [], error => {
                return repeatOnStaleMetadata && !!error.findBy('hasStaleMetadata', true);
            });
        });
    }
    #getCoordinatorConnection(callback) {
        // Get a connection to the coordinator
        this[kMetadata]({}, (error, metadata) => {
            if (error) {
                callback(error);
                return;
            }
            this[kPerformWithRetry]('getCoordinatorConnection', retryCallback => {
                this[kGetConnection](metadata.brokers.get(this.#coordinatorId), (error, connection) => {
                    if (error) {
                        retryCallback(error);
                        return;
                    }
                    retryCallback(null, connection);
                });
            }, callback);
        });
    }
    #handleFencingError(error) {
        if (error.findBy?.('producerFenced', true)) {
            this.#producerInfo = undefined;
            this.#transaction = undefined;
        }
    }
    #beforeSerialization(hook, options, callback) {
        // Create the pre-serialization requests
        const requests = [];
        for (const message of options.messages) {
            requests.push([message.key, 'key', message]);
            requests.push([message.value, 'value', message]);
            if (typeof message.headers !== 'undefined') {
                const headers = message.headers instanceof Map
                    ? message.headers
                    : new Map(Object.entries(message.headers));
                for (const [headerKey, headerValue] of headers) {
                    requests.push([headerKey, 'headerKey', message]);
                    requests.push([headerValue, 'headerValue', message]);
                }
            }
        }
        runAsyncSeries((request, cb) => {
            const [data, type, message] = request;
            const result = hook(data, type, message, cb);
            if (typeof result?.then === 'function') {
                result.then(() => cb(null), cb);
            }
        }, requests, 0, error => {
            if (error) {
                callback(error);
                return;
            }
            this.#send(options, callback);
        });
    }
}
