// @platformatic/kafka/dist/clients/producer/producer-stream.js
import { Writable } from 'node:stream';
import { createPromisifiedCallback, kCallbackPromise } from "../../apis/callbacks.js";
import { notifyCreation } from "../../diagnostic.js";
import { defaultProducerStreamOptions } from "./options.js";
import { ProducerStreamReportModes } from "./types.js";
let currentInstance = 0;
export class ProducerStream extends Writable {
    #producer;
    #batchSize;
    #batchTime;
    #reportMode;
    #produceOptions;
    #buffer;
    #timer;
    #flushing;
    #batchId;
    #pendingWriteCallback;
    instance;
    constructor(producer, options) {
        super({
            objectMode: true,
            highWaterMark: options.highWaterMark
        });
        const { batchSize, batchTime, reportMode, highWaterMark: _highWaterMark, ...sendOptions } = options;
        this.#producer = producer;
        this.#batchSize = batchSize ?? defaultProducerStreamOptions.batchSize;
        this.#batchTime = batchTime ?? defaultProducerStreamOptions.batchTime;
        this.#reportMode = reportMode ?? defaultProducerStreamOptions.reportMode;
        this.#produceOptions = sendOptions;
        this.#buffer = [];
        this.#timer = null;
        this.#flushing = false;
        this.#batchId = 0;
        this.#pendingWriteCallback = null;
        this.instance = currentInstance++;
        notifyCreation('producer-stream', this);
    }
    get producer() {
        return this.#producer;
    }
    close(callback) {
        if (!callback) {
            callback = createPromisifiedCallback();
        }
        this.end((error) => callback(error ?? null));
        return callback[kCallbackPromise];
    }
    _write(message, _, callback) {
        this.#buffer.push(message);
        const finalizeAfterFlush = this.#buffer.length >= this.writableHighWaterMark;
        if (this.#buffer.length >= this.#batchSize) {
            this.#clearTimer();
            this.#flush();
        }
        else {
            this.#scheduleFlush();
        }
        if (finalizeAfterFlush) {
            this.#pendingWriteCallback = callback;
            return;
        }
        callback(null);
    }
    _writev(chunks, callback) {
        for (const { chunk } of chunks) {
            this.#buffer.push(chunk);
        }
        const finalizeAfterFlush = this.#buffer.length >= this.writableHighWaterMark;
        if (this.#buffer.length >= this.#batchSize) {
            this.#clearTimer();
            this.#flush();
        }
        else {
            this.#scheduleFlush();
        }
        if (finalizeAfterFlush) {
            this.#pendingWriteCallback = callback;
            return;
        }
        callback(null);
    }
    _final(callback) {
        this.#clearTimer();
        if (!this.#buffer.length && !this.#flushing) {
            callback(null);
            return;
        }
        const self = this;
        function onError(error) {
            self.removeListener('error', onError);
            self.removeListener('flush', onFlush);
            callback(error);
        }
        function onFlush() {
            if (self.#buffer.length || self.#flushing) {
                return;
            }
            self.removeListener('error', onError);
            self.removeListener('flush', onFlush);
            callback(null);
        }
        this.on('error', onError);
        this.on('flush', onFlush);
        this.#flush();
    }
    _destroy(error, callback) {
        this.#clearTimer();
        this.#buffer = [];
        this.#flushWriteCallbacks(error);
        callback(error);
    }
    #scheduleFlush() {
        if (this.#timer || this.#batchTime < 0) {
            return;
        }
        this.#timer = setTimeout(() => {
            this.#timer = null;
            this.#flush();
        }, this.#batchTime);
    }
    #clearTimer() {
        if (this.#timer) {
            clearTimeout(this.#timer);
            this.#timer = null;
        }
    }
    #flush() {
        if (this.#flushing || this.destroyed) {
            return;
        }
        /* c8 ignore next 3 - Hard to test */
        if (this.#buffer.length === 0) {
            return;
        }
        this.#flushing = true;
        const pending = this.#buffer.splice(0, this.#batchSize);
        const batchId = this.#batchId++;
        const startedAt = Date.now();
        this.#producer.send({ messages: pending, ...this.#produceOptions }, (error, result) => {
            this.#flushing = false;
            if (error) {
                this.#flushWriteCallbacks(error);
                this.destroy(error);
                return;
            }
            this.#emitReports(this.#reportMode, batchId, pending, result);
            this.emit('flush', { batchId, count: pending.length, duration: Date.now() - startedAt, result: result });
            if (this.#buffer.length >= this.#batchSize) {
                this.#flush();
            }
            else if (this.#buffer.length > 0) {
                this.#scheduleFlush();
            }
            this.#flushWriteCallbacks();
        });
    }
    #flushWriteCallbacks(error) {
        if (!this.#pendingWriteCallback || this.#buffer.length >= this.writableHighWaterMark) {
            return;
        }
        const callback = this.#pendingWriteCallback;
        this.#pendingWriteCallback = null;
        callback(error ?? null);
    }
    #emitReports(mode, batchId, pending, result) {
        if (mode === ProducerStreamReportModes.NONE) {
            return;
        }
        if (mode === ProducerStreamReportModes.BATCH) {
            const report = {
                batchId,
                count: pending.length,
                result
            };
            this.emit('delivery-report', report);
            return;
        }
        for (let i = 0; i < pending.length; i++) {
            const report = {
                batchId,
                index: i,
                message: pending[i]
            };
            this.emit('delivery-report', report);
        }
    }
}
