// @platformatic/kafka/dist/clients/producer/transaction.js
import { createPromisifiedCallback, kCallbackPromise } from "../../apis/callbacks.js";
import { createDiagnosticContext, producerTransactionsChannel } from "../../diagnostic.js";
import { UserError } from "../../errors.js";
import { kInstance, kTransaction, kTransactionAddOffsets, kTransactionAddPartitions, kTransactionCancel, kTransactionCommitOffset, kTransactionEnd, kTransactionFindCoordinator, kTransactionPrepare } from "../../symbols.js";
import { kOptions } from "../base/base.js";
let currentInstance = 0;
export class Transaction {
    id;
    #completed;
    #producer;
    #partitions;
    #consumerGroups;
    [kInstance];
    constructor(producer) {
        this[kInstance] = currentInstance++;
        this.id = producer[kOptions].transactionalId;
        this.#producer = producer;
        this.#partitions = new Set();
        this.#consumerGroups = new Set();
        this.#completed = false;
    }
    get completed() {
        return this.#completed;
    }
    [kTransactionPrepare](transactionalId, options, callback) {
        this.#producer[kTransactionFindCoordinator](error => {
            if (error) {
                callback(error);
                return;
            }
            // If the idempotent producer was already initialized for this transactional id, nothing else is required
            if (transactionalId === this.id) {
                callback(null, this);
                return;
            }
            this.#producer.initIdempotentProducer(options, error => {
                if (error) {
                    callback(error);
                    return;
                }
                callback(null, this);
            });
        });
    }
    send(options, callback) {
        if (!callback) {
            callback = createPromisifiedCallback();
        }
        if (this.#completed) {
            callback(new UserError('Cannot produce to an already completed transaction.'));
            return callback[kCallbackPromise];
        }
        this.#producer.send({ ...options, [kTransaction]: this[kInstance] }, callback);
        return callback[kCallbackPromise];
    }
    addPartitions(metadata, messages, callback) {
        if (!callback) {
            callback = createPromisifiedCallback();
        }
        // Scan all messages to find the partitions we need to add to the transaction
        const newAdditions = new Set();
        const topicPartitions = new Map();
        for (const message of messages) {
            message.partition %= metadata.topics.get(message.topic).partitionsCount;
            const { topic, partition } = message;
            const key = `${topic}:${partition}`;
            if (!this.#partitions.has(key)) {
                newAdditions.add(key);
                let partitionsSet = topicPartitions.get(topic);
                if (!partitionsSet) {
                    partitionsSet = new Set();
                    topicPartitions.set(topic, partitionsSet);
                }
                partitionsSet.add(partition);
            }
        }
        // Nothing to add
        if (topicPartitions.size === 0) {
            callback(null);
            return;
        }
        if (this.#completed) {
            callback(new UserError('Cannot add partitions to an already completed transaction.'));
            return callback[kCallbackPromise];
        }
        producerTransactionsChannel.traceCallback(this.#producer[kTransactionAddPartitions], 2, createDiagnosticContext({ client: this.#producer, transaction: this, operation: 'addPartitions' }), this.#producer, this[kInstance], topicPartitions, error => {
            if (!error) {
                for (const key of newAdditions) {
                    this.#partitions.add(key);
                }
            }
            callback(error);
        });
        return callback[kCallbackPromise];
    }
    addConsumer(consumer, callback) {
        if (!callback) {
            callback = createPromisifiedCallback();
        }
        const groupId = consumer.groupId;
        if (this.#consumerGroups.has(groupId)) {
            callback(null);
            return;
        }
        if (this.#completed) {
            callback(new UserError('Cannot add a consumer to an already completed transaction.'));
            return callback[kCallbackPromise];
        }
        producerTransactionsChannel.traceCallback(this.#producer[kTransactionAddOffsets], 2, createDiagnosticContext({
            client: this.#producer,
            transaction: this,
            operation: 'addOffsetsToTransaction',
            groupId
        }), this.#producer, this[kInstance], groupId, error => {
            if (!error) {
                this.#consumerGroups.add(groupId);
            }
            callback(error);
        });
        return callback[kCallbackPromise];
    }
    addOffset(message, callback) {
        if (!callback) {
            callback = createPromisifiedCallback();
        }
        if (this.#completed) {
            callback(new UserError('Cannot add an offset to an already completed transaction.'));
            return callback[kCallbackPromise];
        }
        if (!this.#consumerGroups.has(message.metadata?.consumer?.groupId)) {
            callback(new UserError('Cannot add an offset to a transaction when the consumer group of the message has not been added to the transaction.'));
            return callback[kCallbackPromise];
        }
        producerTransactionsChannel.traceCallback(this.#producer[kTransactionCommitOffset], 2, createDiagnosticContext({ client: this.#producer, transaction: this, operation: 'commitOffset', message }), this.#producer, this[kInstance], message, callback);
        return callback[kCallbackPromise];
    }
    commit(callback) {
        if (!callback) {
            callback = createPromisifiedCallback();
        }
        if (this.#completed) {
            callback(new UserError('Cannot commit an already completed transaction.'));
            return callback[kCallbackPromise];
        }
        producerTransactionsChannel.traceCallback(this.#producer[kTransactionEnd], 2, createDiagnosticContext({ client: this.#producer, transaction: this, operation: 'commit' }), this.#producer, this[kInstance], true, this.#updateCompletionCallback.bind(this, callback));
        return callback[kCallbackPromise];
    }
    abort(callback) {
        if (!callback) {
            callback = createPromisifiedCallback();
        }
        if (this.#completed) {
            callback(new UserError('Cannot abort an already completed transaction.'));
            return callback[kCallbackPromise];
        }
        producerTransactionsChannel.traceCallback(this.#producer[kTransactionEnd], 2, createDiagnosticContext({ client: this.#producer, transaction: this, operation: 'abort' }), this.#producer, this[kInstance], false, this.#updateCompletionCallback.bind(this, callback));
        return callback[kCallbackPromise];
    }
    cancel(callback) {
        if (!callback) {
            callback = createPromisifiedCallback();
        }
        this.#producer[kTransactionCancel](this[kInstance], this.#updateCompletionCallback.bind(this, callback));
        return callback[kCallbackPromise];
    }
    #updateCompletionCallback(callback, error) {
        if (!error) {
            this.#completed = true;
        }
        callback(error);
    }
}
