// @effect/platform-node/dist/NodeHttpIncomingMessage.js
/**
 * @since 1.0.0
 */
import * as Effect from "effect/Effect";
import * as Inspectable from "effect/Inspectable";
import * as Headers from "effect/unstable/http/Headers";
import * as IncomingMessage from "effect/unstable/http/HttpIncomingMessage";
import * as UrlParams from "effect/unstable/http/UrlParams";
import * as NodeStream from "./NodeStream.js";
/**
 * @since 1.0.0
 * @category Constructors
 */
export class NodeHttpIncomingMessage extends Inspectable.Class {
  /**
   * @since 1.0.0
   */
  [IncomingMessage.TypeId];
  source;
  onError;
  remoteAddressOverride;
  constructor(source, onError, remoteAddressOverride) {
    super();
    this[IncomingMessage.TypeId] = IncomingMessage.TypeId;
    this.source = source;
    this.onError = onError;
    this.remoteAddressOverride = remoteAddressOverride;
  }
  get headers() {
    return Headers.fromInput(this.source.headers);
  }
  get remoteAddress() {
    return this.remoteAddressOverride ?? this.source.socket.remoteAddress;
  }
  textEffect;
  get text() {
    if (this.textEffect) {
      return this.textEffect;
    }
    this.textEffect = Effect.runSync(Effect.cached(Effect.flatMap(IncomingMessage.MaxBodySize.asEffect(), maxBodySize => NodeStream.toString(() => this.source, {
      onError: this.onError,
      maxBytes: maxBodySize
    }))));
    return this.textEffect;
  }
  get textUnsafe() {
    return Effect.runSync(this.text);
  }
  get json() {
    return Effect.flatMap(this.text, text => Effect.try({
      try: () => text === "" ? null : JSON.parse(text),
      catch: this.onError
    }));
  }
  get jsonUnsafe() {
    return Effect.runSync(this.json);
  }
  get urlParamsBody() {
    return Effect.flatMap(this.text, _ => Effect.try({
      try: () => UrlParams.fromInput(new URLSearchParams(_)),
      catch: this.onError
    }));
  }
  get stream() {
    return NodeStream.fromReadable({
      evaluate: () => this.source,
      onError: this.onError
    });
  }
  get arrayBuffer() {
    return Effect.withFiber(fiber => NodeStream.toArrayBuffer(() => this.source, {
      onError: this.onError,
      maxBytes: fiber.getRef(IncomingMessage.MaxBodySize)
    }));
  }
}
