// effect/dist/unstable/eventlog/EventLog.js
/**
 * @since 4.0.0
 */
import * as Effect from "../../Effect.js";
import * as FiberMap from "../../FiberMap.js";
import { identity } from "../../Function.js";
import * as Layer from "../../Layer.js";
import { pipeArguments } from "../../Pipeable.js";
import * as Predicate from "../../Predicate.js";
import * as PubSub from "../../PubSub.js";
import * as Queue from "../../Queue.js";
import * as Redacted from "../../Redacted.js";
import * as Schema from "../../Schema.js";
import * as Semaphore from "../../Semaphore.js";
import * as ServiceMap from "../../ServiceMap.js";
import { Reactivity } from "../reactivity/Reactivity.js";
import * as ReactivityLayer from "../reactivity/Reactivity.js";
import * as Event from "./Event.js";
import { Entry, EventJournal, makeEntryIdUnsafe } from "./EventJournal.js";
/**
 * @since 4.0.0
 * @category schema
 */
export const SchemaTypeId = "~effect/eventlog/EventLog/Schema";
/**
 * @since 4.0.0
 * @category schema
 */
export const isEventLogSchema = u => Predicate.hasProperty(u, SchemaTypeId);
/**
 * @since 4.0.0
 * @category schema
 */
export const schema = (...groups) => {
  const EventLog = Object.assign(function EventLog() {}, {
    [SchemaTypeId]: SchemaTypeId,
    groups
  });
  return EventLog;
};
/**
 * @since 4.0.0
 * @category handlers
 */
export const HandlersTypeId = "~effect/eventlog/EventLog/Handlers";
/**
 * @since 4.0.0
 * @category models
 */
export class Identity extends /*#__PURE__*/ServiceMap.Service()("effect/eventlog/EventLog/Identity") {}
/**
 * @since 4.0.0
 * @category schema
 */
export const IdentitySchema = /*#__PURE__*/Schema.Struct({
  publicKey: Schema.String,
  privateKey: /*#__PURE__*/Schema.Redacted(Schema.Uint8ArrayFromBase64)
});
const IdentityEncodedSchema = /*#__PURE__*/Schema.Struct({
  publicKey: Schema.String,
  privateKey: Schema.Uint8ArrayFromBase64
});
const IdentityStringSchema = /*#__PURE__*/Schema.fromJsonString(IdentityEncodedSchema);
/**
 * @since 4.0.0
 * @category constructors
 */
export const decodeIdentityString = value => {
  const decoded = Schema.decodeUnknownSync(IdentityStringSchema)(value);
  return {
    publicKey: decoded.publicKey,
    privateKey: Redacted.make(decoded.privateKey)
  };
};
/**
 * @since 4.0.0
 * @category constructors
 */
export const encodeIdentityString = identity => Schema.encodeSync(IdentityStringSchema)({
  publicKey: identity.publicKey,
  privateKey: Redacted.value(identity.privateKey)
});
/**
 * @since 4.0.0
 * @category constructors
 */
export const makeIdentityUnsafe = () => ({
  publicKey: globalThis.crypto.randomUUID(),
  privateKey: Redacted.make(globalThis.crypto.getRandomValues(new Uint8Array(32)))
});
const handlersProto = {
  [HandlersTypeId]: {
    _Events: identity
  },
  handle(tag, handler) {
    return makeHandlers({
      group: this.group,
      services: this.services,
      handlers: {
        ...this.handlers,
        [tag]: {
          event: this.group.events[tag],
          services: this.services,
          handler
        }
      }
    });
  },
  pipe() {
    return pipeArguments(this, arguments);
  }
};
const makeHandlers = options => Object.assign(Object.create(handlersProto), options);
/**
 * @since 4.0.0
 * @category handlers
 */
export const group = (group, f) => Layer.effectServices(Effect.gen(function* () {
  const services = yield* Effect.services();
  const result = f(makeHandlers({
    group: group,
    handlers: {},
    services
  }));
  const handlers = Effect.isEffect(result) ? yield* result : result;
  const serviceMap = new Map();
  for (const tag in handlers.handlers) {
    const handler = handlers.handlers[tag];
    serviceMap.set(handler.event.key, handlers.handlers[tag]);
  }
  return ServiceMap.makeUnsafe(serviceMap);
}));
/**
 * @since 4.0.0
 * @category compaction
 */
export const groupCompaction = (group, effect) => Layer.effectDiscard(Effect.gen(function* () {
  const log = yield* EventLog;
  const services = yield* Effect.services();
  yield* log.registerCompaction({
    events: Object.keys(group.events),
    effect: Effect.fnUntraced(function* ({
      entries,
      write
    }) {
      const isEventTag = tag => tag in group.events;
      const decodePayload = (tag, payload) => Schema.decodeUnknownEffect(group.events[tag].payloadMsgPack)(payload).pipe(Effect.updateServices(input => ServiceMap.merge(services, input)), Effect.orDie);
      const writePayload = Effect.fnUntraced(function* (timestamp, tag, payload) {
        const event = group.events[tag];
        const entry = new Entry({
          id: makeEntryIdUnsafe({
            msecs: timestamp
          }),
          event: tag,
          payload: yield* Schema.encodeUnknownEffect(event.payloadMsgPack)(payload).pipe(Effect.orDie),
          primaryKey: event.primaryKey(payload)
        }, {
          disableValidation: true
        });
        yield* write(entry);
      });
      const byPrimaryKey = new Map();
      for (const entry of entries) {
        if (!isEventTag(entry.event)) {
          continue;
        }
        const payload = yield* decodePayload(entry.event, entry.payload);
        const record = byPrimaryKey.get(entry.primaryKey);
        const taggedPayload = {
          _tag: entry.event,
          payload
        };
        if (record) {
          record.entries.push(entry);
          record.taggedPayloads.push(taggedPayload);
        } else {
          byPrimaryKey.set(entry.primaryKey, {
            entries: [entry],
            taggedPayloads: [taggedPayload]
          });
        }
      }
      for (const [primaryKey, {
        entries,
        taggedPayloads
      }] of byPrimaryKey) {
        yield* Effect.orDie(effect({
          primaryKey,
          entries,
          events: taggedPayloads,
          write(tag, payload) {
            return Effect.orDie(writePayload(entries[0].createdAtMillis, tag, payload));
          }
        }).pipe(Effect.updateServices(input => ServiceMap.merge(services, input))));
      }
    })
  });
})).pipe(Layer.provide(layerEventLog));
/**
 * @since 4.0.0
 * @category reactivity
 */
export const groupReactivity = (group, keys) => Effect.gen(function* () {
  const log = yield* EventLog;
  if (!Array.isArray(keys)) {
    yield* log.registerReactivity(keys);
    return;
  }
  const obj = {};
  for (const tag in group.events) {
    obj[tag] = keys;
  }
  yield* log.registerReactivity(obj);
}).pipe(Layer.effectDiscard, Layer.provide(layerEventLog));
/**
 * @since 4.0.0
 * @category tags
 */
export class EventLog extends /*#__PURE__*/ServiceMap.Service()("effect/eventlog/EventLog") {}
const make = /*#__PURE__*/Effect.gen(function* () {
  const identity = yield* Identity;
  const journal = yield* EventJournal;
  const services = yield* Effect.services();
  const remotes = yield* FiberMap.make();
  const compactors = new Map();
  const journalSemaphore = yield* Semaphore.make(1);
  const reactivity = yield* Reactivity;
  const reactivityKeys = {};
  const runRemote = Effect.fnUntraced(function* (remote) {
    const startSequence = yield* journal.nextRemoteSequence(remote.id);
    const changes = yield* remote.changes(identity, startSequence);
    yield* Queue.takeAll(changes).pipe(Effect.flatMap(entries => journal.writeFromRemote({
      remoteId: remote.id,
      entries: entries.flat(),
      compact: compactors.size > 0 ? Effect.fnUntraced(function* (remoteEntries) {
        const brackets = [];
        let uncompacted = [];
        let uncompactedRemote = [];
        let index = 0;
        while (index < remoteEntries.length) {
          const remoteEntry = remoteEntries[index];
          const compactor = compactors.get(remoteEntry.entry.event);
          if (!compactor) {
            uncompacted.push(remoteEntry.entry);
            uncompactedRemote.push(remoteEntry);
            index++;
            continue;
          }
          if (uncompacted.length > 0) {
            brackets.push([uncompacted, uncompactedRemote]);
            uncompacted = [];
            uncompactedRemote = [];
          }
          const entries = [remoteEntry.entry];
          const remoteGroup = [remoteEntry];
          const compacted = [];
          index++;
          while (index < remoteEntries.length) {
            const nextRemoteEntry = remoteEntries[index];
            if (!compactor.events.has(nextRemoteEntry.entry.event)) {
              break;
            }
            entries.push(nextRemoteEntry.entry);
            remoteGroup.push(nextRemoteEntry);
            index++;
          }
          yield* compactor.effect({
            entries,
            write(entry) {
              return Effect.sync(() => {
                compacted.push(entry);
              });
            }
          }).pipe(Effect.orDie);
          brackets.push([compacted, remoteGroup]);
        }
        if (uncompacted.length > 0) {
          brackets.push([uncompacted, uncompactedRemote]);
        }
        return brackets;
      }) : undefined,
      effect: Effect.fnUntraced(function* ({
        conflicts,
        entry
      }) {
        const handler = services.mapUnsafe.get(Event.serviceKey(entry.event));
        if (!handler) {
          return yield* Effect.logDebug(`Event handler not found for: "${entry.event}"`);
        }
        const decodePayload = Schema.decodeUnknownEffect(handler.event.payloadMsgPack);
        const decodedConflicts = new Array(conflicts.length);
        for (let i = 0; i < conflicts.length; i++) {
          decodedConflicts[i] = {
            entry: conflicts[i],
            payload: yield* decodePayload(conflicts[i].payload).pipe(Effect.updateServices(input => ServiceMap.merge(handler.services, input)))
          };
        }
        yield* decodePayload(entry.payload).pipe(Effect.flatMap(payload => handler.handler({
          payload,
          entry,
          conflicts: decodedConflicts
        })), Effect.updateServices(input => ServiceMap.merge(handler.services, input)), Effect.asVoid);
        if (reactivityKeys[entry.event]) {
          for (const key of reactivityKeys[entry.event]) {
            yield* Effect.sync(() => reactivity.invalidateUnsafe({
              [key]: [entry.primaryKey]
            }));
          }
        }
      }, Effect.catchCause(Effect.logError), (effect, {
        entry
      }) => Effect.annotateLogs(effect, {
        service: "EventLog",
        effect: "writeFromRemote",
        entryId: entry.idString
      }))
    }).pipe(journalSemaphore.withPermits(1))), Effect.catchCause(Effect.logError), Effect.forever, Effect.annotateLogs({
      service: "EventLog",
      effect: "runRemote consume"
    }), Effect.forkScoped);
    const write = journal.withRemoteUncommited(remote.id, entries => remote.write(identity, entries));
    yield* Effect.addFinalizer(() => Effect.ignore(write));
    yield* write;
    const changesSub = yield* journal.changes;
    return yield* PubSub.takeAll(changesSub).pipe(Effect.andThen(Effect.sleep("500 millis")), Effect.andThen(write), Effect.catchCause(Effect.logError), Effect.forever);
  }, Effect.scoped, Effect.provideService(Identity, identity), Effect.interruptible);
  const writeHandler = Effect.fnUntraced(function* (handler, options) {
    const payload = yield* Schema.encodeUnknownEffect(handler.event.payloadMsgPack)(options.payload).pipe(Effect.orDie);
    return yield* journalSemaphore.withPermits(1)(journal.write({
      event: options.event,
      primaryKey: handler.event.primaryKey(options.payload),
      payload,
      effect: entry => handler.handler({
        payload: options.payload,
        entry,
        conflicts: []
      }).pipe(Effect.updateServices(input => ServiceMap.merge(handler.services, input)), Effect.tap(() => Effect.sync(() => {
        if (reactivityKeys[entry.event]) {
          for (const key of reactivityKeys[entry.event]) {
            reactivity.invalidateUnsafe({
              [key]: [entry.primaryKey]
            });
          }
        }
      })))
    }));
  });
  const eventLogWrite = options => {
    const handler = services.mapUnsafe.get(Event.serviceKey(options.event));
    if (handler === undefined) {
      return Effect.die(`Event handler not found for: "${options.event}"`);
    }
    return writeHandler(handler, options);
  };
  return EventLog.of({
    write: eventLogWrite,
    entries: journal.entries,
    registerRemote: remote => Effect.acquireRelease(FiberMap.run(remotes, remote.id, runRemote(remote)), () => FiberMap.remove(remotes, remote.id)).pipe(Effect.asVoid),
    registerCompaction: options => Effect.acquireRelease(Effect.sync(() => {
      const events = new Set(options.events);
      const compactor = {
        events,
        effect: options.effect
      };
      for (const event of options.events) {
        compactors.set(event, compactor);
      }
    }), () => Effect.sync(() => {
      for (const event of options.events) {
        compactors.delete(event);
      }
    })),
    registerReactivity: keys => Effect.sync(() => {
      Object.assign(reactivityKeys, keys);
    }),
    destroy: journal.destroy
  });
});
/**
 * @since 4.0.0
 * @category layers
 */
export const layerEventLog = /*#__PURE__*/Layer.effect(EventLog, make).pipe(/*#__PURE__*/Layer.provide(ReactivityLayer.layer));
/**
 * @since 4.0.0
 * @category layers
 */
export const layer = _schema => layerEventLog;
/**
 * @since 4.0.0
 * @category client
 */
export const makeClient = schema => Effect.gen(function* () {
  const log = yield* EventLog;
  return (event, payload) => log.write({
    schema,
    event,
    payload
  });
});
