// effect/dist/unstable/eventlog/EventLogEncryption.js
/**
 * @since 4.0.0
 */
import * as Effect from "../../Effect.js";
import * as Layer from "../../Layer.js";
import * as Redacted from "../../Redacted.js";
import * as Schema from "../../Schema.js";
import * as ServiceMap from "../../ServiceMap.js";
import { Entry, EntryId, RemoteEntry } from "./EventJournal.js";
/**
 * @since 4.0.0
 * @category models
 */
export const EncryptedEntry = /*#__PURE__*/Schema.Struct({
  entryId: EntryId,
  encryptedEntry: Schema.Uint8Array
});
/**
 * @since 4.0.0
 * @category models
 */
export const EncryptedRemoteEntry = /*#__PURE__*/Schema.Struct({
  sequence: Schema.Number,
  iv: Schema.Uint8Array,
  entryId: EntryId,
  encryptedEntry: Schema.Uint8Array
});
const toArrayBuffer = data => {
  const buffer = new ArrayBuffer(data.byteLength);
  new Uint8Array(buffer).set(data);
  return buffer;
};
const toBufferSource = data => new Uint8Array(toArrayBuffer(data));
/**
 * @since 4.0.0
 * @category services
 */
export class EventLogEncryption extends /*#__PURE__*/ServiceMap.Service()("effect/eventlog/EventLogEncryption") {}
/**
 * @since 4.0.0
 * @category encryption
 */
export const makeEncryptionSubtle = crypto => Effect.sync(() => {
  const keyCache = new WeakMap();
  const getKey = identity => Effect.suspend(() => {
    if (keyCache.has(identity)) {
      return Effect.succeed(keyCache.get(identity));
    }
    return Effect.promise(() => crypto.subtle.importKey("raw", toArrayBuffer(Redacted.value(identity.privateKey)), "AES-GCM", true, ["encrypt", "decrypt"])).pipe(Effect.tap(key => Effect.sync(() => {
      keyCache.set(identity, key);
    })));
  });
  return EventLogEncryption.of({
    encrypt: Effect.fnUntraced(function* (identity, entries) {
      const data = yield* Effect.orDie(Entry.encodeArray(entries));
      const key = yield* getKey(identity);
      const iv = crypto.getRandomValues(new Uint8Array(12));
      const encryptedEntries = yield* Effect.promise(() => Promise.all(data.map(entry => crypto.subtle.encrypt({
        name: "AES-GCM",
        iv: toBufferSource(iv),
        tagLength: 128
      }, key, toBufferSource(entry)))));
      return {
        iv,
        encryptedEntries: encryptedEntries.map(entry => new Uint8Array(entry))
      };
    }),
    decrypt: Effect.fnUntraced(function* (identity, entries) {
      const key = yield* getKey(identity);
      const decryptedData = (yield* Effect.promise(() => Promise.all(entries.map(data => crypto.subtle.decrypt({
        name: "AES-GCM",
        iv: toBufferSource(data.iv),
        tagLength: 128
      }, key, toBufferSource(data.encryptedEntry)))))).map(buffer => new Uint8Array(buffer));
      const decoded = yield* Effect.orDie(Entry.decodeArray(decryptedData));
      return decoded.map((entry, index) => new RemoteEntry({
        remoteSequence: entries[index].sequence,
        entry
      }));
    }),
    sha256: data => Effect.promise(() => crypto.subtle.digest("SHA-256", toArrayBuffer(data))).pipe(Effect.map(hash => new Uint8Array(hash))),
    sha256String: data => Effect.map(Effect.promise(() => crypto.subtle.digest("SHA-256", toArrayBuffer(data))), hash => {
      const hashArray = Array.from(new Uint8Array(hash));
      const hashHex = hashArray.map(bytes => bytes.toString(16).padStart(2, "0")).join("");
      return hashHex;
    })
  });
});
/**
 * @since 4.0.0
 * @category encryption
 */
export const layerSubtle = /*#__PURE__*/Layer.effect(EventLogEncryption, /*#__PURE__*/makeEncryptionSubtle(globalThis.crypto));
