// effect/dist/unstable/eventlog/EventLogServer.js
/**
 * @since 4.0.0
 */
import * as Uuid from "uuid";
import * as Effect from "../../Effect.js";
import * as FiberMap from "../../FiberMap.js";
import * as Layer from "../../Layer.js";
import * as PubSub from "../../PubSub.js";
import * as Queue from "../../Queue.js";
import * as RcMap from "../../RcMap.js";
import * as Schema from "../../Schema.js";
import * as ServiceMap from "../../ServiceMap.js";
import * as HttpServerRequest from "../http/HttpServerRequest.js";
import * as HttpServerResponse from "../http/HttpServerResponse.js";
import { EntryId, makeRemoteIdUnsafe } from "./EventJournal.js";
import { Ack, Changes, ChunkedMessage, decodeRequest, encodeResponse, Hello, Pong } from "./EventLogRemote.js";
const constChunkSize = 512_000;
/**
 * @since 4.0.0
 * @category constructors
 */
export const makeHandler = /*#__PURE__*/Effect.gen(function* () {
  const storage = yield* Storage;
  const remoteId = yield* storage.getId;
  let chunkId = 0;
  return Effect.fnUntraced(function* (socket) {
    const subscriptions = yield* FiberMap.make();
    const writeRaw = yield* socket.writer;
    const chunks = new Map();
    let latestSequence = -1;
    const write = Effect.fnUntraced(function* (response) {
      const data = yield* encodeResponse(response);
      if (response._tag !== "Changes" || data.byteLength <= constChunkSize) {
        return yield* writeRaw(data);
      }
      const id = chunkId++;
      for (const part of ChunkedMessage.split(id, data)) {
        yield* writeRaw(yield* encodeResponse(part));
      }
    });
    yield* Effect.forkChild(Effect.orDie(write(new Hello({
      remoteId
    }))));
    const handleRequest = request => {
      switch (request._tag) {
        case "Ping":
          {
            return Effect.orDie(write(new Pong({
              id: request.id
            })));
          }
        case "WriteEntries":
          {
            if (request.encryptedEntries.length === 0) {
              return Effect.orDie(write(new Ack({
                id: request.id,
                sequenceNumbers: []
              })));
            }
            return Effect.gen(function* () {
              const entries = request.encryptedEntries.map(({
                encryptedEntry,
                entryId
              }) => new PersistedEntry({
                entryId,
                iv: request.iv,
                encryptedEntry
              }));
              const encrypted = yield* storage.write(request.publicKey, entries);
              latestSequence = encrypted[encrypted.length - 1].sequence;
              return yield* Effect.orDie(write(new Ack({
                id: request.id,
                sequenceNumbers: encrypted.map(entry => entry.sequence)
              })));
            });
          }
        case "RequestChanges":
          {
            return Effect.gen(function* () {
              const changes = yield* storage.changes(request.publicKey, request.startSequence);
              return yield* Queue.takeAll(changes).pipe(Effect.flatMap(entries => {
                const latestEntries = [];
                for (const entry of entries) {
                  if (entry.sequence <= latestSequence) continue;
                  latestEntries.push(entry);
                  latestSequence = entry.sequence;
                }
                if (latestEntries.length === 0) return Effect.void;
                return Effect.orDie(write(new Changes({
                  publicKey: request.publicKey,
                  entries: latestEntries
                })));
              }), Effect.forever);
            }).pipe(Effect.scoped, FiberMap.run(subscriptions, request.publicKey));
          }
        case "StopChanges":
          {
            return FiberMap.remove(subscriptions, request.publicKey);
          }
        case "ChunkedMessage":
          {
            const data = ChunkedMessage.join(chunks, request);
            if (!data) return Effect.void;
            return Effect.flatMap(Effect.orDie(decodeRequest(data)), handleRequest);
          }
      }
    };
    yield* socket.run(data => Effect.flatMap(Effect.orDie(decodeRequest(data)), handleRequest)).pipe(Effect.catchCause(cause => Effect.logDebug(cause)));
  }, Effect.scoped, Effect.annotateLogs({
    module: "EventLogServer"
  }));
});
/**
 * @since 4.0.0
 * @category websockets
 */
export const makeHandlerHttp = /*#__PURE__*/Effect.gen(function* () {
  const handler = yield* makeHandler;
  // @effect-diagnostics-next-line returnEffectInGen:off
  return Effect.gen(function* () {
    const request = yield* HttpServerRequest.HttpServerRequest;
    const socket = yield* request.upgrade;
    yield* handler(socket);
    return HttpServerResponse.empty();
  }).pipe(Effect.annotateLogs({
    module: "EventLogServer"
  }));
});
/**
 * @since 4.0.0
 * @category storage
 */
export class PersistedEntry extends /*#__PURE__*/Schema.Class("effect/eventlog/EventLogServer/PersistedEntry")({
  entryId: EntryId,
  iv: Schema.Uint8Array,
  encryptedEntry: Schema.Uint8Array
}) {
  /**
   * @since 4.0.0
   */
  get entryIdString() {
    return Uuid.stringify(this.entryId);
  }
}
/**
 * @since 4.0.0
 * @category storage
 */
export class Storage extends /*#__PURE__*/ServiceMap.Service()("effect/eventlog/EventLogServer/Storage") {}
/**
 * @since 4.0.0
 * @category storage
 */
export const makeStorageMemory = /*#__PURE__*/Effect.gen(function* () {
  const knownIds = new Map();
  const journals = new Map();
  const remoteId = makeRemoteIdUnsafe();
  const ensureJournal = publicKey => {
    let journal = journals.get(publicKey);
    if (journal) return journal;
    journal = [];
    journals.set(publicKey, journal);
    return journal;
  };
  const pubsubs = yield* RcMap.make({
    lookup: _publicKey => Effect.acquireRelease(PubSub.unbounded(), PubSub.shutdown),
    idleTimeToLive: 60000
  });
  return Storage.of({
    getId: Effect.succeed(remoteId),
    write: (publicKey, entries) => Effect.gen(function* () {
      const active = yield* RcMap.keys(pubsubs);
      let pubsub;
      for (const key of active) {
        if (key === publicKey) {
          pubsub = yield* RcMap.get(pubsubs, publicKey);
          break;
        }
      }
      const journal = ensureJournal(publicKey);
      const encryptedEntries = [];
      for (const entry of entries) {
        const idString = entry.entryIdString;
        if (knownIds.has(idString)) continue;
        const encrypted = {
          sequence: journal.length,
          entryId: entry.entryId,
          iv: entry.iv,
          encryptedEntry: entry.encryptedEntry
        };
        encryptedEntries.push(encrypted);
        knownIds.set(idString, encrypted.sequence);
        journal.push(encrypted);
        if (pubsub) {
          yield* PubSub.publish(pubsub, encrypted);
        }
      }
      return encryptedEntries;
    }).pipe(Effect.scoped),
    entries: (publicKey, startSequence) => Effect.sync(() => ensureJournal(publicKey).slice(startSequence)),
    changes: (publicKey, startSequence) => Effect.gen(function* () {
      const queue = yield* Queue.make();
      const pubsub = yield* RcMap.get(pubsubs, publicKey);
      const subscription = yield* PubSub.subscribe(pubsub);
      yield* Queue.offerAll(queue, ensureJournal(publicKey).slice(startSequence));
      yield* PubSub.takeAll(subscription).pipe(Effect.flatMap(chunk => Queue.offerAll(queue, chunk)), Effect.forever, Effect.forkScoped);
      yield* Effect.addFinalizer(() => Queue.shutdown(queue));
      return Queue.asDequeue(queue);
    })
  });
});
/**
 * @since 4.0.0
 * @category storage
 */
export const layerStorageMemory = /*#__PURE__*/Layer.effect(Storage)(makeStorageMemory);
