// effect/dist/unstable/eventlog/EventLogRemote.js
/**
 * @since 4.0.0
 */
import * as Data from "../../Data.js";
import * as Deferred from "../../Deferred.js";
import * as Effect from "../../Effect.js";
import * as Exit from "../../Exit.js";
import * as Layer from "../../Layer.js";
import * as Queue from "../../Queue.js";
import * as RcMap from "../../RcMap.js";
import * as Schedule from "../../Schedule.js";
import * as Schema from "../../Schema.js";
import * as ServiceMap from "../../ServiceMap.js";
import * as Msgpack from "../encoding/Msgpack.js";
import * as Socket from "../socket/Socket.js";
import { EntryId, RemoteEntry, RemoteId } from "./EventJournal.js";
import { EncryptedEntry, EncryptedRemoteEntry, EventLogEncryption, layerSubtle } from "./EventLogEncryption.js";
/**
 * @since 4.0.0
 * @category models
 */
export class EventLogRemote extends /*#__PURE__*/ServiceMap.Service()("effect/eventlog/EventLogRemote") {}
/**
 * @since 4.0.0
 * @category protocol
 */
export class Hello extends /*#__PURE__*/Schema.Class("effect/eventlog/EventLogRemote/Hello")({
  _tag: /*#__PURE__*/Schema.tag("Hello"),
  remoteId: RemoteId
}) {}
/**
 * @since 4.0.0
 * @category protocol
 */
export class ChunkedMessage extends /*#__PURE__*/Schema.Class("effect/eventlog/EventLogRemote/ChunkedMessage")({
  _tag: /*#__PURE__*/Schema.tag("ChunkedMessage"),
  id: Schema.Number,
  part: /*#__PURE__*/Schema.Tuple([Schema.Number, Schema.Number]),
  data: Schema.Uint8Array
}) {
  /**
   * @since 4.0.0
   */
  static split(id, data) {
    const parts = Math.ceil(data.byteLength / constChunkSize);
    const result = new Array(parts);
    for (let i = 0; i < parts; i++) {
      const start = i * constChunkSize;
      const end = Math.min((i + 1) * constChunkSize, data.byteLength);
      result[i] = new ChunkedMessage({
        _tag: "ChunkedMessage",
        id,
        part: [i, parts],
        data: data.subarray(start, end)
      });
    }
    return result;
  }
  /**
   * @since 4.0.0
   */
  static join(map, part) {
    const [index, total] = part.part;
    let entry = map.get(part.id);
    if (!entry) {
      entry = {
        parts: new Array(total),
        count: 0,
        bytes: 0
      };
      map.set(part.id, entry);
    }
    entry.parts[index] = part.data;
    entry.count++;
    entry.bytes += part.data.byteLength;
    if (entry.count !== total) {
      return;
    }
    const data = new Uint8Array(entry.bytes);
    let offset = 0;
    for (const part of entry.parts) {
      data.set(part, offset);
      offset += part.byteLength;
    }
    map.delete(part.id);
    return data;
  }
}
/**
 * @since 4.0.0
 * @category protocol
 */
export class WriteEntries extends /*#__PURE__*/Schema.Class("effect/eventlog/EventLogRemote/WriteEntries")({
  _tag: /*#__PURE__*/Schema.tag("WriteEntries"),
  publicKey: Schema.String,
  id: Schema.Number,
  iv: Schema.Uint8Array,
  encryptedEntries: /*#__PURE__*/Schema.Array(EncryptedEntry)
}) {}
/**
 * @since 4.0.0
 * @category protocol
 */
export class Ack extends /*#__PURE__*/Schema.Class("effect/eventlog/EventLogRemote/Ack")({
  _tag: /*#__PURE__*/Schema.tag("Ack"),
  id: Schema.Number,
  sequenceNumbers: /*#__PURE__*/Schema.Array(Schema.Number)
}) {}
/**
 * @since 4.0.0
 * @category protocol
 */
export class RequestChanges extends /*#__PURE__*/Schema.Class("effect/eventlog/EventLogRemote/RequestChanges")({
  _tag: /*#__PURE__*/Schema.tag("RequestChanges"),
  publicKey: Schema.String,
  startSequence: Schema.Number
}) {}
/**
 * @since 4.0.0
 * @category protocol
 */
export class Changes extends /*#__PURE__*/Schema.Class("effect/eventlog/EventLogRemote/Changes")({
  _tag: /*#__PURE__*/Schema.tag("Changes"),
  publicKey: Schema.String,
  entries: /*#__PURE__*/Schema.Array(EncryptedRemoteEntry)
}) {}
/**
 * @since 4.0.0
 * @category protocol
 */
export class StopChanges extends /*#__PURE__*/Schema.Class("effect/eventlog/EventLogRemote/StopChanges")({
  _tag: /*#__PURE__*/Schema.tag("StopChanges"),
  publicKey: Schema.String
}) {}
/**
 * @since 4.0.0
 * @category protocol
 */
export class Ping extends /*#__PURE__*/Schema.Class("effect/eventlog/EventLogRemote/Ping")({
  _tag: /*#__PURE__*/Schema.tag("Ping"),
  id: Schema.Number
}) {}
/**
 * @since 4.0.0
 * @category protocol
 */
export class Pong extends /*#__PURE__*/Schema.Class("effect/eventlog/EventLogRemote/Pong")({
  _tag: /*#__PURE__*/Schema.tag("Pong"),
  id: Schema.Number
}) {}
/**
 * @since 4.0.0
 * @category protocol
 */
export const ProtocolRequest = /*#__PURE__*/Schema.Union([WriteEntries, RequestChanges, StopChanges, ChunkedMessage, Ping]);
/**
 * @since 4.0.0
 * @category protocol
 */
export const ProtocolRequestMsgpack = /*#__PURE__*/Msgpack.schema(ProtocolRequest);
/**
 * @since 4.0.0
 * @category protocol
 */
export const decodeRequest = /*#__PURE__*/Schema.decodeUnknownEffect(ProtocolRequestMsgpack);
/**
 * @since 4.0.0
 * @category protocol
 */
export const encodeRequest = /*#__PURE__*/Schema.encodeUnknownEffect(ProtocolRequestMsgpack);
/**
 * @since 4.0.0
 * @category protocol
 */
export const ProtocolResponse = /*#__PURE__*/Schema.Union([Hello, Ack, Changes, ChunkedMessage, Pong]);
/**
 * @since 4.0.0
 * @category protocol
 */
export const ProtocolResponseMsgpack = /*#__PURE__*/Msgpack.schema(ProtocolResponse);
/**
 * @since 4.0.0
 * @category protocol
 */
export const decodeResponse = /*#__PURE__*/Schema.decodeUnknownEffect(ProtocolResponseMsgpack);
/**
 * @since 4.0.0
 * @category protocol
 */
export const encodeResponse = /*#__PURE__*/Schema.encodeUnknownEffect(ProtocolResponseMsgpack);
/**
 * @since 4.0.0
 * @category change
 */
export class RemoteAdditions extends /*#__PURE__*/Schema.Class("effect/eventlog/EventLogRemote/RemoteAdditions")({
  _tag: /*#__PURE__*/Schema.tag("RemoteAdditions"),
  entries: /*#__PURE__*/Schema.Array(RemoteEntry)
}) {}
const constChunkSize = 512_000;
/**
 * @since 4.0.0
 * @category errors
 */
export class EventLogRemoteError extends /*#__PURE__*/Data.TaggedError("EventLogRemoteError") {}
/**
 * @since 4.0.0
 * @category entry
 */
export const RemoteEntryChange = /*#__PURE__*/Schema.Tuple([RemoteId, /*#__PURE__*/Schema.Array(EntryId)]);
/**
 * @since 4.0.0
 * @category constructors
 */
export const fromSocket = options => Effect.gen(function* () {
  const socket = yield* Socket.Socket;
  const encryption = yield* EventLogEncryption;
  const writeRaw = yield* socket.writer;
  const writeRequest = request => Effect.gen(function* () {
    const data = yield* encodeRequest(request);
    if (request._tag !== "WriteEntries" || data.byteLength <= constChunkSize) {
      return yield* writeRaw(data);
    }
    const id = request.id;
    for (const part of ChunkedMessage.split(id, data)) {
      yield* writeRaw(yield* encodeRequest(part));
    }
  });
  let pendingCounter = 0;
  const pending = new Map();
  const chunks = new Map();
  const subscriptions = yield* RcMap.make({
    lookup: publicKey => Effect.acquireRelease(Queue.make(), queue => Queue.shutdown(queue).pipe(Effect.andThen(Effect.ignore(writeRequest(new StopChanges({
      publicKey
    }))))))
  });
  const identities = new Map();
  const badPing = yield* Deferred.make();
  const remoteId = yield* Deferred.make();
  let latestPing = 0;
  let latestPong = 0;
  if (options?.disablePing !== true) {
    yield* Effect.suspend(() => {
      if (latestPing !== latestPong) {
        return Deferred.fail(badPing, new Error("Ping timeout"));
      }
      return writeRequest(new Ping({
        id: ++latestPing
      }));
    }).pipe(Effect.delay("10 seconds"), Effect.ignore, Effect.forever, Effect.interruptible, Effect.forkScoped);
  }
  const handleMessage = res => {
    switch (res._tag) {
      case "Hello":
        {
          return Deferred.succeed(remoteId, res.remoteId).pipe(Effect.asVoid);
        }
      case "Ack":
        {
          return Effect.gen(function* () {
            const entry = pending.get(res.id);
            if (!entry) return;
            pending.delete(res.id);
            const {
              deferred,
              entries,
              publicKey
            } = entry;
            const remoteEntries = res.sequenceNumbers.map((sequenceNumber, i) => {
              const entry = entries[i];
              return new RemoteEntry({
                remoteSequence: sequenceNumber,
                entry
              });
            });
            const queue = yield* RcMap.get(subscriptions, publicKey);
            yield* Queue.offerAll(queue, remoteEntries);
            yield* Deferred.done(deferred, Exit.void);
          }).pipe(Effect.scoped);
        }
      case "Pong":
        {
          latestPong = res.id;
          if (res.id === latestPing) {
            return Effect.void;
          }
          return Deferred.fail(badPing, new Error("Pong id mismatch")).pipe(Effect.asVoid);
        }
      case "Changes":
        {
          return Effect.gen(function* () {
            const queue = yield* RcMap.get(subscriptions, res.publicKey);
            const identity = identities.get(res.publicKey);
            if (!identity) {
              return;
            }
            const entries = yield* encryption.decrypt(identity, res.entries);
            yield* Queue.offerAll(queue, entries);
          }).pipe(Effect.scoped);
        }
      case "ChunkedMessage":
        {
          const data = ChunkedMessage.join(chunks, res);
          if (!data) return Effect.void;
          return Effect.scoped(Effect.flatMap(decodeResponse(data), handleMessage));
        }
    }
  };
  yield* socket.run(data => Effect.flatMap(decodeResponse(data), handleMessage)).pipe(Effect.raceFirst(Deferred.await(badPing)), Effect.tapCause(Effect.logDebug), Effect.retry({
    schedule: Schedule.exponential(100).pipe(Schedule.either(Schedule.spaced(5000)))
  }), Effect.annotateLogs({
    service: "EventLogRemote",
    method: "fromSocket"
  }), Effect.forkScoped, Effect.interruptible);
  const id = yield* Deferred.await(remoteId);
  return {
    id,
    write: (identity, entries) => Effect.gen(function* () {
      const encrypted = yield* encryption.encrypt(identity, entries);
      const deferred = yield* Deferred.make();
      const id = pendingCounter++;
      pending.set(id, {
        entries,
        deferred,
        publicKey: identity.publicKey
      });
      yield* Effect.orDie(writeRequest(new WriteEntries({
        publicKey: identity.publicKey,
        id,
        iv: encrypted.iv,
        encryptedEntries: encrypted.encryptedEntries.map((encryptedEntry, i) => ({
          entryId: entries[i].id,
          encryptedEntry
        }))
      })));
      yield* Deferred.await(deferred);
    }),
    changes: (identity, startSequence) => Effect.gen(function* () {
      const queue = yield* RcMap.get(subscriptions, identity.publicKey);
      identities.set(identity.publicKey, identity);
      yield* Effect.orDie(writeRequest(new RequestChanges({
        publicKey: identity.publicKey,
        startSequence
      })));
      return queue;
    })
  };
});
/**
 * @since 4.0.0
 * @category constructors
 */
export const fromWebSocket = (url, options) => Effect.gen(function* () {
  const socket = yield* Socket.makeWebSocket(url);
  return yield* fromSocket(options).pipe(Effect.provideService(Socket.Socket, socket));
});
/**
 * @since 4.0.0
 * @category layers
 */
export const layerWebSocket = (url, options) => Layer.effectDiscard(fromWebSocket(url, options));
/**
 * @since 4.0.0
 * @category layers
 */
export const layerWebSocketBrowser = (url, options) => layerWebSocket(url, options).pipe(Layer.provide([layerSubtle, Socket.layerWebSocketConstructorGlobal]));
