// effect/dist/unstable/httpapi/HttpApiError.js
/**
 * @since 4.0.0
 */
import * as ErrorReporter from "../../ErrorReporter.js";
import { identity } from "../../Function.js";
import * as Schema from "../../Schema.js";
import * as Transformation from "../../SchemaTransformation.js";
import * as HttpApiSchema from "./HttpApiSchema.js";
/**
 * @category Built-in errors
 * @since 4.0.0
 */
export class BadRequest extends /*#__PURE__*/Schema.ErrorClass("effect/HttpApiError/BadRequest")({
  _tag: /*#__PURE__*/Schema.tag("BadRequest")
}, {
  description: "BadRequest",
  httpApiStatus: 400
}) {
  [ErrorReporter.ignore] = true;
  static singleton = /*#__PURE__*/new BadRequest();
}
/**
 * @category NoContent errors
 * @since 4.0.0
 */
export const BadRequestNoContent = /*#__PURE__*/BadRequest.pipe(/*#__PURE__*/HttpApiSchema.asNoContent({
  decode: () => new BadRequest({})
}));
/**
 * @category Built-in errors
 * @since 4.0.0
 */
export const BadRequestFromSchemaError = /*#__PURE__*/BadRequest.pipe(Schema.decodeTo(Schema.Union([Schema.declare(Schema.isSchemaError), BadRequest]), Transformation.transform({
  encode: _ => BadRequest.singleton,
  decode: identity
})), HttpApiSchema.asNoContent({
  decode: () => new BadRequest({})
})).annotate({
  httpApiStatus: 400,
  description: "BadRequest"
});
/**
 * @category Built-in errors
 * @since 4.0.0
 */
export class Unauthorized extends /*#__PURE__*/Schema.ErrorClass("effect/HttpApiError/Unauthorized")({
  _tag: /*#__PURE__*/Schema.tag("Unauthorized")
}, {
  description: "Unauthorized",
  httpApiStatus: 401
}) {
  [ErrorReporter.ignore] = true;
}
/**
 * @category NoContent errors
 * @since 4.0.0
 */
export const UnauthorizedNoContent = /*#__PURE__*/Unauthorized.pipe(/*#__PURE__*/HttpApiSchema.asNoContent({
  decode: () => new Unauthorized({})
}));
/**
 * @category Built-in errors
 * @since 4.0.0
 */
export class Forbidden extends /*#__PURE__*/Schema.ErrorClass("effect/HttpApiError/Forbidden")({
  _tag: /*#__PURE__*/Schema.tag("Forbidden")
}, {
  description: "Forbidden",
  httpApiStatus: 403
}) {
  [ErrorReporter.ignore] = true;
}
/**
 * @category NoContent errors
 * @since 4.0.0
 */
export const ForbiddenNoContent = /*#__PURE__*/Forbidden.pipe(/*#__PURE__*/HttpApiSchema.asNoContent({
  decode: () => new Forbidden({})
}));
/**
 * @category Built-in errors
 * @since 4.0.0
 */
export class NotFound extends /*#__PURE__*/Schema.ErrorClass("effect/HttpApiError/NotFound")({
  _tag: /*#__PURE__*/Schema.tag("NotFound")
}, {
  description: "NotFound",
  httpApiStatus: 404
}) {
  [ErrorReporter.ignore] = true;
}
/**
 * @category NoContent errors
 * @since 4.0.0
 */
export const NotFoundNoContent = /*#__PURE__*/NotFound.pipe(/*#__PURE__*/HttpApiSchema.asNoContent({
  decode: () => new NotFound({})
}));
/**
 * @category Built-in errors
 * @since 4.0.0
 */
export class MethodNotAllowed extends /*#__PURE__*/Schema.ErrorClass("effect/HttpApiError/MethodNotAllowed")({
  _tag: /*#__PURE__*/Schema.tag("MethodNotAllowed")
}, {
  description: "MethodNotAllowed",
  httpApiStatus: 405
}) {
  [ErrorReporter.ignore] = true;
}
/**
 * @category NoContent errors
 * @since 4.0.0
 */
export const MethodNotAllowedNoContent = /*#__PURE__*/MethodNotAllowed.pipe(/*#__PURE__*/HttpApiSchema.asNoContent({
  decode: () => new MethodNotAllowed({})
}));
/**
 * @category Built-in errors
 * @since 4.0.0
 */
export class NotAcceptable extends /*#__PURE__*/Schema.ErrorClass("effect/HttpApiError/NotAcceptable")({
  _tag: /*#__PURE__*/Schema.tag("NotAcceptable")
}, {
  description: "NotAcceptable",
  httpApiStatus: 406
}) {
  [ErrorReporter.ignore] = true;
}
/**
 * @category NoContent errors
 * @since 4.0.0
 */
export const NotAcceptableNoContent = /*#__PURE__*/NotAcceptable.pipe(/*#__PURE__*/HttpApiSchema.asNoContent({
  decode: () => new NotAcceptable({})
}));
/**
 * @category Built-in errors
 * @since 4.0.0
 */
export class RequestTimeout extends /*#__PURE__*/Schema.ErrorClass("effect/HttpApiError/RequestTimeout")({
  _tag: /*#__PURE__*/Schema.tag("RequestTimeout")
}, {
  description: "RequestTimeout",
  httpApiStatus: 408
}) {
  [ErrorReporter.ignore] = true;
}
/**
 * @category NoContent errors
 * @since 4.0.0
 */
export const RequestTimeoutNoContent = /*#__PURE__*/RequestTimeout.pipe(/*#__PURE__*/HttpApiSchema.asNoContent({
  decode: () => new RequestTimeout({})
}));
/**
 * @category Built-in errors
 * @since 4.0.0
 */
export class Conflict extends /*#__PURE__*/Schema.ErrorClass("effect/HttpApiError/Conflict")({
  _tag: /*#__PURE__*/Schema.tag("Conflict")
}, {
  description: "Conflict",
  httpApiStatus: 409
}) {
  [ErrorReporter.ignore] = true;
}
/**
 * @since 4.0.0
 * @category NoContent errors
 */
export const ConflictNoContent = /*#__PURE__*/Conflict.pipe(/*#__PURE__*/HttpApiSchema.asNoContent({
  decode: () => new Conflict({})
}));
/**
 * @category Built-in errors
 * @since 4.0.0
 */
export class Gone extends /*#__PURE__*/Schema.ErrorClass("effect/HttpApiError/Gone")({
  _tag: /*#__PURE__*/Schema.tag("Gone")
}, {
  description: "Gone",
  httpApiStatus: 410
}) {
  [ErrorReporter.ignore] = true;
}
/**
 * @category NoContent errors
 * @since 4.0.0
 */
export const GoneNoContent = /*#__PURE__*/Gone.pipe(/*#__PURE__*/HttpApiSchema.asNoContent({
  decode: () => new Gone({})
}));
/**
 * @category Built-in errors
 * @since 4.0.0
 */
export class InternalServerError extends /*#__PURE__*/Schema.ErrorClass("effect/HttpApiError/InternalServerError")({
  _tag: /*#__PURE__*/Schema.tag("InternalServerError")
}, {
  description: "InternalServerError",
  httpApiStatus: 500
}) {}
/**
 * @category NoContent errors
 * @since 4.0.0
 */
export const InternalServerErrorNoContent = /*#__PURE__*/InternalServerError.pipe(/*#__PURE__*/HttpApiSchema.asNoContent({
  decode: () => new InternalServerError({})
}));
/**
 * @category Built-in errors
 * @since 4.0.0
 */
export class NotImplemented extends /*#__PURE__*/Schema.ErrorClass("effect/HttpApiError/NotImplemented")({
  _tag: /*#__PURE__*/Schema.tag("NotImplemented")
}, {
  description: "NotImplemented",
  httpApiStatus: 501
}) {}
/**
 * @category NoContent errors
 * @since 4.0.0
 */
export const NotImplementedNoContent = /*#__PURE__*/NotImplemented.pipe(/*#__PURE__*/HttpApiSchema.asNoContent({
  decode: () => new NotImplemented({})
}));
/**
 * @category Built-in errors
 * @since 4.0.0
 */
export class ServiceUnavailable extends /*#__PURE__*/Schema.ErrorClass("effect/HttpApiError/ServiceUnavailable")({
  _tag: /*#__PURE__*/Schema.tag("ServiceUnavailable")
}, {
  description: "ServiceUnavailable",
  httpApiStatus: 503
}) {}
/**
 * @category NoContent errors
 * @since 4.0.0
 */
export const ServiceUnavailableNoContent = /*#__PURE__*/ServiceUnavailable.pipe(/*#__PURE__*/HttpApiSchema.asNoContent({
  decode: () => new ServiceUnavailable({})
}));
